<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Binned Motif Enrichment Analysis and Visualization • monaLisa</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Binned Motif Enrichment Analysis and Visualization">
<meta name="description" content="Useful functions to work with sequence motifs in the analysis of genomics data. These include methods to annotate genomic regions or sequences with predicted motif hits and to identify motifs that drive observed changes in accessibility or expression. Functions to produce informative visualizations of the obtained results are also provided.">
<meta property="og:description" content="Useful functions to work with sequence motifs in the analysis of genomics data. These include methods to annotate genomic regions or sequences with predicted motif hits and to identify motifs that drive observed changes in accessibility or expression. Functions to produce informative visualizations of the obtained results are also provided.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/monaLisa.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><img src="reference/figures/monaLisa.png" align="right" alt="monaLisa" width="150"></p>
<p><br></p>
<div class="section level1">
<div class="page-header"><h1 id="monalisa-motif-analysis-with-lisa">
<code>monaLisa</code>: MOtif aNAlysis with Lisa<a class="anchor" aria-label="anchor" href="#monalisa-motif-analysis-with-lisa"></a>
</h1></div>
<p><br></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>monaLisa</code> was inspired by her father <a href="http://homer.ucsd.edu/homer/index.html" class="external-link">Homer</a> to look for enriched motifs in sets (bins) of genomic regions, compared to all other regions (“binned motif enrichment analysis”).</p>
<p>It uses known motifs representing transcription factor binding preferences, for example for the <code>JASPAR2020</code> Bioconductor package. The regions are for example promoters or accessible regions, which are grouped into bins according to a numerical value assigned to each region, such as change of expression or accessibility. The goal of the analysis is to identify transcription factors that are associated to that numerical value and thus candidates to be drivers in the underlying biological process.</p>
<p>In addition to the “binned motif enrichment analysis”, <code>monaLisa</code> can also be used to address the above question using stability selection (a form of linear regression), or to look for motif matches in sequences.</p>
<p>Current contributors include:</p>
<ul>
<li><a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a></li>
<li><a href="https://github.com/machlabd" class="external-link">Dania Machlab</a></li>
<li><a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a></li>
<li><a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="news">News<a class="anchor" aria-label="anchor" href="#news"></a>
</h2>
<ul>
<li>information on the latest changes can be found <a href="https://github.com/fmicompbio/monaLisa/blob/master/NEWS.md" class="external-link">here</a>
</li>
<li>a preprint is available on <a href="https://doi.org/10.1101/2021.11.30.470570" class="external-link">bioRxiv</a>
</li>
<li>
<code>monaLisa</code> is available on <a href="https://bioconductor.org/packages/monaLisa" class="external-link">Bioconductor</a>
</li>
<li>
<code>monaLisa</code> is now published in <a href="https://doi.org/10.1093/bioinformatics/btac102" class="external-link">Bioinformatics</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>To cite <code>monaLisa</code> please use the publication found <a href="https://doi.org/10.1093/bioinformatics/btac102" class="external-link">here</a> or see <code>citation("monaLisa")</code>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>monaLisa</code> can be installed from Bioconductor via the <code>BiocManager</code> package:</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"monaLisa"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="functionality">Functionality<a class="anchor" aria-label="anchor" href="#functionality"></a>
</h2>
<p>Here is a minimal example to run a <code>monaLisa</code> analysis:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/monaLisa" class="external-link">monaLisa</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">mcparams</span> <span class="op">&lt;-</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">10L</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calcBinnedMotifEnrR.html">calcBinnedMotifEnrR</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">seqs</span>,   <span class="co"># DNAStringSet (e.g. peak sequences)</span></span>
<span>                          bins <span class="op">=</span> <span class="va">bins</span>,   <span class="co"># factor that groups 'seqs'</span></span>
<span>                          pwmL <span class="op">=</span> <span class="va">pwms</span>, <span class="co"># PWMatrixList (know motifs)</span></span>
<span>                          BPPARAM <span class="op">=</span> <span class="va">mcparams</span>,</span>
<span>                          min.score <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<p>The return value <code>se</code> is a <code>SummarizedExperiment</code> with motifs in rows and bins in columns, and multiple assays with significance and magnitude of the enrichments.</p>
<p>The inputs for <code>calcBinnedMotifEnrR</code> can be easily obtained using other Bioconductor packages:</p>
<pre><code><span><span class="co"># get sequences ('atacPeaks' is a GRanges)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span><span class="op">)</span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span>, <span class="va">atacPeaks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># bin sequences ('atacPeaksChange' is a numerical vector)</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu">monaLisa</span><span class="fu">::</span><span class="fu"><a href="reference/bin.html">bin</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">atacPeaksChange</span>, binmode <span class="op">=</span> <span class="st">"equalN"</span>, nElement <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># obtain known motifs from Jaspar</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span></span>
<span><span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span><span class="va">JASPAR2020</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>matrixtype <span class="op">=</span> <span class="st">"PWM"</span>, tax_group <span class="op">=</span> <span class="st">"vertebrates"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>The results can be conveniently visualized:</p>
<pre><code><span><span class="fu"><a href="reference/plotBinDensity.html">plotBinDensity</a></span><span class="op">(</span><span class="va">atacPeaksChange</span>, <span class="va">bins</span>, legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<p><img src="reference/figures/monaLisa_binning_small.png" align="center" alt="binning" width="412px"></p>
<pre><code><span><span class="fu"><a href="reference/plotMotifHeatmaps.html">plotMotifHeatmaps</a></span><span class="op">(</span><span class="va">se</span>, cluster <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  which.plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"enr"</span>, <span class="st">"FDR"</span><span class="op">)</span>,</span>
<span>                  show_seqlogo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<p><img src="reference/figures/monaLisa_heatmaps_small.png" align="center" alt="heatmaps" width="501px"></p>
<!-- badges: start -->

</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/monaLisa" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/fmicompbio/monaLisa/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/fmicompbio/monaLisa/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing monaLisa</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/machlabd" class="external-link">Dania Machlab</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-2578-6930" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-5596-7567" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3833-2169" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2269-4934" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/fmicompbio/monaLisa/actions" class="external-link"><img src="https://github.com/fmicompbio/monaLisa/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/github/fmicompbio/monaLisa" class="external-link"><img src="https://codecov.io/github/fmicompbio/monaLisa/coverage.svg?branch=master" alt="Codecov.io coverage status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
