<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate k-mer enrichment in bins of sequences. — calcBinnedKmerEnr • monaLisa</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate k-mer enrichment in bins of sequences. — calcBinnedKmerEnr"><meta property="og:description" content="Given a set of sequences and corresponding bins, identify
  enriched k-mers (n-grams) in each bin. The sequences can be given either
  directly or as genomic coordinates."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/monaLisa.html">Binned motif enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fmicompbio/monaLisa">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate k-mer enrichment in bins of sequences.</h1>
    <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/master/R/motif_enrichment_kmers.R"><code>R/motif_enrichment_kmers.R</code></a></small>
    <div class="hidden name"><code>calcBinnedKmerEnr.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a set of sequences and corresponding bins, identify
  enriched k-mers (n-grams) in each bin. The sequences can be given either
  directly or as genomic coordinates.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class="sourceCode r"><code><span class="fu">calcBinnedKmerEnr</span><span class="op">(</span>
  <span class="va">seqs</span>,
  bins <span class="op">=</span> <span class="cn">NULL</span>,
  kmerLen <span class="op">=</span> <span class="fl">5</span>,
  background <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"otherBins"</span>, <span class="st">"allBins"</span>, <span class="st">"zeroBin"</span>, <span class="st">"genome"</span>, <span class="st">"model"</span><span class="op">)</span>,
  MMorder <span class="op">=</span> <span class="fl">1</span>,
  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"fisher"</span>, <span class="st">"binomial"</span><span class="op">)</span>,
  includeRevComp <span class="op">=</span> <span class="cn">TRUE</span>,
  maxFracN <span class="op">=</span> <span class="fl">0.7</span>,
  maxKmerSize <span class="op">=</span> <span class="fl">3L</span>,
  GCbreaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.25</span>, <span class="fl">0.3</span>, <span class="fl">0.35</span>, <span class="fl">0.4</span>, <span class="fl">0.45</span>, <span class="fl">0.5</span>, <span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">0.8</span><span class="op">)</span>,
  pseudocount.kmers <span class="op">=</span> <span class="fl">1</span>,
  pseudocount.log2enr <span class="op">=</span> <span class="fl">8</span>,
  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,
  genome <span class="op">=</span> <span class="cn">NULL</span>,
  genome.regions <span class="op">=</span> <span class="cn">NULL</span>,
  genome.oversample <span class="op">=</span> <span class="fl">2</span>,
  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,
  verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>

    <div id="arguments">
    <h2>Arguments</h2>
    <table class="ref-arguments"><colgroup><col class="name"><col class="desc"></colgroup><tr><th>seqs</th>
      <td><p><code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></code> object with sequences to
test</p></td>
    </tr><tr><th>bins</th>
      <td><p>factor of the same length and order as <code>seqs</code>, indicating
the bin for each sequence. Typically the return value of
<code><a href="bin.html">bin</a></code>. For <code>background = "genome"</code> or
<code>background = "model"</code>, <code>bins</code> can be omitted.</p></td>
    </tr><tr><th>kmerLen</th>
      <td><p>A <code>numeric</code> scalar giving the k-mer length.</p></td>
    </tr><tr><th>background</th>
      <td><p>A <code>character</code> scalar specifying the background 
sequences to use. One of <code>"otherBins"</code> (default), <code>"allBins"</code>, 
<code>"zeroBin"</code>, <code>"genome"</code> or <code>"model"</code> (see "Details").</p></td>
    </tr><tr><th>MMorder</th>
      <td><p>A <code>numeric</code> scalar giving the order of the Markov model
used to calculate the expected frequencies for <code>background = "model"</code>.</p></td>
    </tr><tr><th>test</th>
      <td><p>A <code>character</code> scalar specifying the type of enrichment test
to perform. One of <code>"fisher"</code> (default) or <code>"binomial"</code>. The
enrichment test is one-sided (enriched in foreground).</p></td>
    </tr><tr><th>includeRevComp</th>
      <td><p>A <code>logical</code> scalar. If <code>TRUE</code> (default),
count k-mer occurrences in both <code>seqs</code> and their reverse-complement,
by concatenating <code>seqs</code> and their reverse-complemented versions
before the counting. This is useful if motifs can be expected to occur
on any strand (e.g. DNA sequences of ChIP-seq peaks). If motifs are only
expected on the forward strand (e.g. RNA sequences of CLIP-seq peaks),
<code>includeRevComp = FALSE</code> should be used. Note that <code>bins</code>
will be recycled for the reverse complemented sequences, which means that
each reverse-complemented sequence will be assigned to the same bib as the
corresponding forward sequence.</p></td>
    </tr><tr><th>maxFracN</th>
      <td><p>A numeric scalar with the maximal fraction of N bases allowed
in a sequence (defaults to 0.7). Sequences with higher fractions are
excluded from the analysis.</p></td>
    </tr><tr><th>maxKmerSize</th>
      <td><p>the maximum k-mer size to consider, when adjusting
background sequence weights for k-mer composition compared to the
foreground sequences. The default value (3) will correct for mono-, di-
and tri-mer composition.</p></td>
    </tr><tr><th>GCbreaks</th>
      <td><p>The breaks between GC bins. The default value is based on
the hard-coded bins used in Homer.</p></td>
    </tr><tr><th>pseudocount.kmers</th>
      <td><p>A <code>numeric</code> scalar - will be added to the
observed and expected counts for each k-mer to avoid zero values.</p></td>
    </tr><tr><th>pseudocount.log2enr</th>
      <td><p>A numerical scalar with the pseudocount to add to
foreground and background counts when calculating log2 motif enrichments</p></td>
    </tr><tr><th>p.adjust.method</th>
      <td><p>A character scalar selecting the p value adjustment
method (used in <code><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></code>).</p></td>
    </tr><tr><th>genome</th>
      <td><p>A <code>BSgenome</code> or <code>DNAStringSet</code> object with the
genome sequence. Only used for <code>background = "genome"</code> for extracting
background sequences.</p></td>
    </tr><tr><th>genome.regions</th>
      <td><p>An optional <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></code> object
defining the intervals in <code>genome</code> from which background sequences are
sampled for <code>background = "genome"</code>. If <code>NULL</code>, background
sequences are sampled randomly from <code>genome</code>.</p></td>
    </tr><tr><th>genome.oversample</th>
      <td><p>A <code>numeric</code> scalar of at least 1.0 defining how
many background sequences will be sampled per foreground sequence for
<code>background = "genome"</code>. Larger values will take longer but improve
the sequence composition similarity between foreground and background
(see <code>"Details"</code>).</p></td>
    </tr><tr><th>BPPARAM</th>
      <td><p>An optional <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html">BiocParallelParam</a></code>
instance determining the parallel back-end to be used during evaluation.</p></td>
    </tr><tr><th>verbose</th>
      <td><p>A <code>logical</code> scalar. If <code>TRUE</code>, report on progress.</p></td>
    </tr></table></div>

    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></code> object
  with motifs in rows and bins in columns, containing seven assays:</p><ul><li><p>negLog10P: -log10 P values</p></li>
<li><p>negLog10Padj: -log10 adjusted P values</p></li>
<li><p>pearsonResid: k-mer enrichments as Pearson residuals</p></li>
<li><p>expForegroundWgtWithHits: expected number of foreground 
    sequences with motif hits</p></li>
<li><p>log2enr: k-mer enrichments as log2 ratios</p></li>
<li><p>sumForegroundWgtWithHits: Sum of foreground sequence weights
    in a bin that have k-mer occurrences</p></li>
<li><p>sumBackgroundWgtWithHits: Sum of background sequence weights
    in a bin that have k-mer occurrences</p></li>
</ul><p>#' The <code>rowData</code> of the object contains annotations (name, PFMs, PWMs 
and GC fraction) for the k-mers, while the <code>colData</code> slot contains 
summary information about the bins.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function implements a binned k-mer enrichment analysis. In each
  enrichment analysis, the sequences in a specific bin are used as foreground
  sequences to test for k-mer enrichments comparing to background sequences
  (defined by <code>background</code>, see below), similarly as in done for motifs
  in <code><a href="calcBinnedMotifEnrR.html">calcBinnedMotifEnrR</a></code>. Sequences are weighted to correct for
  GC and shorter k-mer composition differences between fore- and background
  sets.</p>  
<p>The background sequences are defined according to the value of the
  <code>background</code> argument:</p><ul><li><p>otherBins: sequences from all other bins (excluding the current  
      bin)</p></li>
<li><p>allBins: sequences from all bins (including the current bin)</p></li>
<li><p>zeroBin: sequences from the "zero bin", defined by the
      <code>maxAbsX</code> argument of <code><a href="bin.html">bin</a></code>. If <code>bins</code>
      does not define a "zero bin", for example because it was created by
      <code>bin(..., maxAbsX = NULL)</code>, selecting this background definition
      will abort with an error.</p></li>
<li><p>genome: sequences randomly sampled from the genome (or the
      intervals defined in <code>genome.regions</code> if given). For each
      foreground sequence, <code>genome.oversample</code> background sequences
      of the same size are sampled (on average). From these, one per
      foreground sequence is selected trying to match the G+C composition.
      In order to make the sampling deterministic, a seed number needs to be
      provided to the <code>RNGseed</code> parameter in 
      <code><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></code>
      or <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></code> when creating the 
      <code>BiocParallelParam</code> instance in <code>BPPARAM</code>.</p></li>
<li><p>model: a Markov model of the order <code>MMorder</code> is estimated
      from the foreground sequences and used to estimate expected k-mer
      frequencies. K-mer enrichments are then calculated comparing observed
      to these expected frequencies. In order to make the process 
      deterministic, a seed number needs to be provided to the 
      <code>RNGseed</code> parameter in <code><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></code> or 
      <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></code> when creating the 
      <code>BiocParallelParam</code> instance in <code>BPPARAM</code>.</p></li>
</ul><p>For each k-mer, the weights of sequences is multiplied with the number
  of k-mer occurrences in each sequence and summed, separately for foreground
  (<code>sumForegroundWgtWithHits</code>) and background
  (<code>sumBackgroundWgtWithHits</code>) sequences. The function works in ZOOPS
  (Zero-Or-One-Per-Sequence) mode, so at most one occurrence per
  sequence is counted, which helps reduce the impact of sequence repeats.
  The total foreground (<code>totalWgtForeground</code>) and background
  (<code>totalWgtBackground</code>) sum of sequence weights is also calculated. If
  a k-mer has zero <code>sumForegroundWgtWithHits</code> and
  <code>sumBackgroundWgtWithHits</code>, then any values (p-values and enrichment)
  that are calculated using these two numbers are set to NA.</p>
<p>Two statistical tests for the calculation of enrichment log p-value are
  available: <code>test = "fisher"</code> (default) to perform Fisher's exact
  tests, or <code>test = "binomial"</code> to perform binomial tests, using:</p><ul><li><p>fisher: <code>fisher.test(x = tab, alternative =
      "greater")</code>, where <code>tab</code> is the contingency table with the summed
      weights of sequences in foreground or background sets (rows), and with
      or without a occurrences of a particular k-mer (columns).</p></li>
<li><p>binomial: <code>pbinom(q = sumForegroundWgtWithHits - 1, size =
      totalWgtForeground, 
      prob = sumBackgroundWgtWithHits / totalWgtBackground,
      lower.tail = FALSE, log.p = TRUE)</code></p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="getKmerFreq.html">getKmerFreq</a></code> used to calculate k-mer enrichments;
  <code><a href="https://rdrr.io/pkg/BSgenome/man/getSeq-methods.html">getSeq,BSgenome-method</a></code> which is used to extract
  sequences from <code>genomepkg</code> if <code>x</code> is a <code>GRanges</code> object;
  <code><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></code> that is used for parallelization;
  <code><a href="bin.html">bin</a></code> for binning of regions</p></div>
    </div>

    <div id="examples">
    <h2>Examples</h2>
    <div class="ref-examples sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GCATGCATGC"</span>, <span class="st">"CATGCGCATG"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">calcBinnedKmerEnr</span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">seqs</span>, bins <span class="op">=</span> <span class="va">bins</span>, kmerLen <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 64 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(5): bins bins.binmode bins.breaks bins.bin0 param</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(7): negLog10P negLog10Padj ... sumForegroundWgtWithHits</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   sumBackgroundWgtWithHits</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(64): AAA AAC ... TTG TTT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(5): motif.id motif.name motif.pfm motif.pwm</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   motif.percentGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(2): 1 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(6): bin.names bin.lower ... totalWgtForeground</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   totalWgtBackground</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

