<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Prepare and run HOMER motif enrichment analysis. — calcBinnedMotifEnrHomer • monaLisa</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prepare and run HOMER motif enrichment analysis. — calcBinnedMotifEnrHomer"><meta name="description" content="Run complete HOMER motif enrichment analysis, consisting of
    calls to prepareHomer, system2 and
    parseHomerOutput. This function requires HOMER
    to be installed (see http://homer.ucsd.edu/homer/index.html)
    and the path to the tool to be provided (homerfile argument)."><meta property="og:description" content="Run complete HOMER motif enrichment analysis, consisting of
    calls to prepareHomer, system2 and
    parseHomerOutput. This function requires HOMER
    to be installed (see http://homer.ucsd.edu/homer/index.html)
    and the path to the tool to be provided (homerfile argument)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/monaLisa.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Prepare and run HOMER motif enrichment analysis.</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/devel/R/motif_enrichment_HOMER.R" class="external-link"><code>R/motif_enrichment_HOMER.R</code></a></small>
      <div class="d-none name"><code>calcBinnedMotifEnrHomer.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run complete HOMER motif enrichment analysis, consisting of
    calls to <code><a href="prepareHomer.html">prepareHomer</a></code>, <code><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></code> and
    <code><a href="parseHomerOutput.html">parseHomerOutput</a></code>. This function requires <code>HOMER</code>
    to be installed (see <a href="http://homer.ucsd.edu/homer/index.html" class="external-link">http://homer.ucsd.edu/homer/index.html</a>)
    and the path to the tool to be provided (<code>homerfile</code> argument).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calcBinnedMotifEnrHomer</span><span class="op">(</span></span>
<span>  <span class="va">gr</span>,</span>
<span>  <span class="va">b</span>,</span>
<span>  <span class="va">genomedir</span>,</span>
<span>  <span class="va">outdir</span>,</span>
<span>  <span class="va">motifFile</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  regionsize <span class="op">=</span> <span class="st">"given"</span>,</span>
<span>  pseudocount.log2enr <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  Ncpu <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose.Homer <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gr">gr<a class="anchor" aria-label="anchor" href="#arg-gr"></a></dt>
<dd><p>A <code>GRanges</code> object (or an object that can be coerced to one)
with the genomic regions to analyze.</p></dd>


<dt id="arg-b">b<a class="anchor" aria-label="anchor" href="#arg-b"></a></dt>
<dd><p>A vector of the same length as <code>gr</code> that groups its elements
into bins (typically a factor, such as the one returned by
<code><a href="bin.html">bin</a></code>).</p></dd>


<dt id="arg-genomedir">genomedir<a class="anchor" aria-label="anchor" href="#arg-genomedir"></a></dt>
<dd><p>Directory containing sequence files in Fasta format
(one per chromosome).</p></dd>


<dt id="arg-outdir">outdir<a class="anchor" aria-label="anchor" href="#arg-outdir"></a></dt>
<dd><p>A path specifying the folder into which the output files will
be written.</p></dd>


<dt id="arg-motiffile">motifFile<a class="anchor" aria-label="anchor" href="#arg-motiffile"></a></dt>
<dd><p>A file with HOMER formatted PWMs to be used in the
enrichment analysis.</p></dd>


<dt id="arg-homerfile">homerfile<a class="anchor" aria-label="anchor" href="#arg-homerfile"></a></dt>
<dd><p>Path and file name of the <code>findMotifsGenome.pl</code>
HOMER script.</p></dd>


<dt id="arg-regionsize">regionsize<a class="anchor" aria-label="anchor" href="#arg-regionsize"></a></dt>
<dd><p>The peak size to use in HOMER (<code>"given"</code> keeps the
coordinate region, an integer value will keep only that many bases in
the region center).</p></dd>


<dt id="arg-pseudocount-log-enr">pseudocount.log2enr<a class="anchor" aria-label="anchor" href="#arg-pseudocount-log-enr"></a></dt>
<dd><p>A numerical scalar with the pseudocount to add to
foreground and background counts when calculating log2 motif enrichments</p></dd>


<dt id="arg-p-adjust-method">p.adjust.method<a class="anchor" aria-label="anchor" href="#arg-p-adjust-method"></a></dt>
<dd><p>A character scalar selecting the p value adjustment
method (used in <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>).</p></dd>


<dt id="arg-ncpu">Ncpu<a class="anchor" aria-label="anchor" href="#arg-ncpu"></a></dt>
<dd><p>Number of parallel threads that HOMER can use.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>A logical scalar. If <code>TRUE</code>, print progress messages.</p></dd>


<dt id="arg-verbose-homer">verbose.Homer<a class="anchor" aria-label="anchor" href="#arg-verbose-homer"></a></dt>
<dd><p>A logical scalar. If <code>TRUE</code>, print the console
output when running Homer.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>SummarizedExperiment</code> object with motifs in rows and bins
  in columns, containing seven assays:</p><dl><dt>negLog10P</dt>
<dd><p>: -log10 P values</p></dd>

  <dt>negLog10Padj</dt>
<dd><p>: -log10 adjusted P values</p></dd>

  <dt>pearsonResid</dt>
<dd><p>: motif enrichments as Pearson residuals</p></dd>

  <dt>expForegroundWgtWithHits</dt>
<dd><p>: expected number of foreground
    sequences with motif hits</p></dd>

  <dt>log2enr</dt>
<dd><p>: motif enrichments as log2 ratios</p></dd>

  <dt>sumForegroundWgtWithHits</dt>
<dd><p>: Sum of foreground sequence weights
    in a bin that have motif hits</p></dd>

  <dt>sumBackgroundWgtWithHits</dt>
<dd><p>: Sum of background sequence weights
    in a bin that have motif hits</p></dd>


</dl><p>The <code>rowData</code> of the object contains annotations (name, PFMs, PWMs
and GC fraction) for the motifs, while the <code>colData</code> slot contains
summary information about the bins.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>The functions that are wrapped: <code><a href="prepareHomer.html">prepareHomer</a></code>,
    <code><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></code> and <code><a href="parseHomerOutput.html">parseHomerOutput</a></code>,
    <code><a href="bin.html">bin</a></code> for binning of regions</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># genome</span></span></span>
<span class="r-in"><span>  <span class="va">genome</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"exampleGenome.fa"</span>, package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># create motif file for Homer</span></span></span>
<span class="r-in"><span>  <span class="va">motiffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">motifIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MA0139.1"</span>, <span class="st">"MA1102.1"</span>, <span class="st">"MA0740.1"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="dumpJaspar.html">dumpJaspar</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">motiffile</span>, pkg <span class="op">=</span> <span class="st">"JASPAR2020"</span>,</span></span>
<span class="r-in"><span>             opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="va">motifIDs</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># GRanges of regions used in binned motif enrichment analysis</span></span></span>
<span class="r-in"><span>  <span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/tileGenome.html" class="external-link">tileGenome</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      seqlengths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>chr1 <span class="op">=</span> <span class="fl">10000L</span>, chr2 <span class="op">=</span> <span class="fl">10000L</span>, chr3 <span class="op">=</span> <span class="fl">10000L</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      tilewidth <span class="op">=</span> <span class="fl">200</span>, cut.last.tile.in.chrom <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># create bins (motif enrichment analysis will be per bin)</span></span></span>
<span class="r-in"><span>  <span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># run calcBinnedMotifEnrHomer</span></span></span>
<span class="r-in"><span>  <span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">calcBinnedMotifEnrHomer</span><span class="op">(</span>gr <span class="op">=</span> <span class="va">gr</span>, b <span class="op">=</span> <span class="va">bins</span>, genomedir <span class="op">=</span> <span class="va">genome</span>,</span></span>
<span class="r-in"><span>      outdir <span class="op">=</span> <span class="va">outdir</span>, motifFile <span class="op">=</span> <span class="va">motiffile</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

