<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Binned Motif Enrichment Analysis with monaLisa — calcBinnedMotifEnrR • monaLisa</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Binned Motif Enrichment Analysis with monaLisa — calcBinnedMotifEnrR" />
<meta property="og:description" content="This function performs a motif enrichment analysis on bins of
  sequences. For each bin, the sequences in all other bins are used as
  background." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.50</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/monaLisa.html">monaLisa</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/fmicompbio/monaLisa">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Binned Motif Enrichment Analysis with <code>monaLisa</code></h1>
    
    <div class="hidden name"><code>calcBinnedMotifEnrR.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs a motif enrichment analysis on bins of
  sequences. For each bin, the sequences in all other bins are used as
  background.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>calcBinnedMotifEnrR</span><span class='op'>(</span>
  <span class='va'>seqs</span>,
  bins <span class='op'>=</span> <span class='cn'>NULL</span>,
  pwmL <span class='op'>=</span> <span class='cn'>NULL</span>,
  background <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"otherBins"</span>, <span class='st'>"allBins"</span>, <span class='st'>"zeroBin"</span>, <span class='st'>"genome"</span><span class='op'>)</span>,
  test <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"fisher"</span>, <span class='st'>"binomial"</span><span class='op'>)</span>,
  maxFracN <span class='op'>=</span> <span class='fl'>0.7</span>,
  maxKmerSize <span class='op'>=</span> <span class='fl'>3L</span>,
  min.score <span class='op'>=</span> <span class='fl'>10</span>,
  matchMethod <span class='op'>=</span> <span class='st'>"matchPWM"</span>,
  GCbreaks <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.25</span>, <span class='fl'>0.3</span>, <span class='fl'>0.35</span>, <span class='fl'>0.4</span>, <span class='fl'>0.45</span>, <span class='fl'>0.5</span>, <span class='fl'>0.6</span>, <span class='fl'>0.7</span>, <span class='fl'>0.8</span><span class='op'>)</span>,
  pseudocount.log2enr <span class='op'>=</span> <span class='fl'>8</span>,
  pseudocount.pearsonResid <span class='op'>=</span> <span class='fl'>0.001</span>,
  p.adjust.method <span class='op'>=</span> <span class='st'>"BH"</span>,
  genome <span class='op'>=</span> <span class='cn'>NULL</span>,
  genome.regions <span class='op'>=</span> <span class='cn'>NULL</span>,
  genome.oversample <span class='op'>=</span> <span class='fl'>2</span>,
  genome.seed <span class='op'>=</span> <span class='fl'>42L</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>SerialParam</span><span class='op'>(</span><span class='op'>)</span>,
  verbose <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>seqs</th>
      <td><p>DNAStringSet object with sequences to test</p></td>
    </tr>
    <tr>
      <th>bins</th>
      <td><p>factor of the same length and order as <code>seqs</code>, indicating
the bin for each sequence. Typically the return value of
<code><a href='bin.html'>bin</a></code>. For <code>background = "genome"</code>, <code>bins</code>
can be omitted.</p></td>
    </tr>
    <tr>
      <th>pwmL</th>
      <td><p>PWMatrixList with motifs for which to calculate enrichments.</p></td>
    </tr>
    <tr>
      <th>background</th>
      <td><p>A <code>character</code> scalar specifying the background sequences
to use. One of <code>"otherBins"</code> (default), <code>"allBins"</code>, <code>"zeroBin"</code>
or <code>"genome"</code> (see "Details").</p></td>
    </tr>
    <tr>
      <th>test</th>
      <td><p>A <code>character</code> scalar specifying the type of enrichment test
to perform. One of <code>"fisher"</code> (default) or <code>"binomial"</code>. The
enrichment test is one-sided (enriched in foreground).</p></td>
    </tr>
    <tr>
      <th>maxFracN</th>
      <td><p>A numeric scalar with the maximal fraction of N bases allowed
in a sequence (defaults to 0.7). Sequences with higher fractions are
excluded from the analysis.</p></td>
    </tr>
    <tr>
      <th>maxKmerSize</th>
      <td><p>the maximum k-mer size to consider, when adjusting
background sequence weights for k-mer composition compared to the
foreground sequences. The default value (3) will correct for mono-, di-
and tri-mer composition.</p></td>
    </tr>
    <tr>
      <th>min.score</th>
      <td><p>the minimal score for motif hits, used in
<code><a href='findMotifHits-methods.html'>findMotifHits</a></code>.</p></td>
    </tr>
    <tr>
      <th>matchMethod</th>
      <td><p>the method used to scan for motif hits, passed to the
<code>method</code> parameter in <code><a href='findMotifHits-methods.html'>findMotifHits</a></code>.</p></td>
    </tr>
    <tr>
      <th>GCbreaks</th>
      <td><p>The breaks between GC bins. The default value is based on
the hard-coded bins used in Homer.</p></td>
    </tr>
    <tr>
      <th>pseudocount.log2enr</th>
      <td><p>A numerical scalar with the pseudocount to add to
foreground and background counts when calculating log2 motif enrichments</p></td>
    </tr>
    <tr>
      <th>pseudocount.pearsonResid</th>
      <td><p>A numerical scalar with the pseudocount to add
to foreground and background frequencies when calculating expected counts
and Pearson residuals.</p></td>
    </tr>
    <tr>
      <th>p.adjust.method</th>
      <td><p>A character scalar selecting the p value adjustment
method (used in <code><a href='https://rdrr.io/r/stats/p.adjust.html'>p.adjust</a></code>).</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>A <code>BSgenome</code> or <code>DNAStringSet</code> object with the
genome sequence. Only used for <code>background = "genome"</code> for extracting
background sequences.</p></td>
    </tr>
    <tr>
      <th>genome.regions</th>
      <td><p>An optional <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> object
defining the intervals in <code>genome</code> from which background sequences are
sampled for <code>background = "genome"</code>. If <code>NULL</code>, background
sequences are sampled randomly from <code>genome</code>.</p></td>
    </tr>
    <tr>
      <th>genome.oversample</th>
      <td><p>A <code>numeric</code> scalar of at least 1.0 defining how
many background sequences will be sampled per foreground sequence for
<code>background = "genome"</code>. Larger values will take longer but improve
the sequence composition similarity between foreground and background
(see <code>"Details"</code>).</p></td>
    </tr>
    <tr>
      <th>genome.seed</th>
      <td><p>An <code>integer</code> scalar used to seed the random number
generator before sampling regions.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>An optional <code><a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html'>BiocParallelParam</a></code>
instance determining the parallel back-end to be used during evaluation.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>A logical scalar. If <code>TRUE</code>, print progress messages.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments for  <code><a href='findMotifHits-methods.html'>findMotifHits</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>SummarizedExperiment</code> object with motifs in rows and bins
  in columns, containing six assays:</p><ul>
<li><p>negLog10P: -log10 P values</p></li>
<li><p>negLog10Padj: -log10 adjusted P values</p></li>
<li><p>pearsonResid: motif enrichments as Pearson residuals</p></li>
<li><p>log2enr: motif enrichments as log2 ratios</p></li>
<li><p>sumForegroundWgtWithHits: Sum of foreground sequence weights
    in a bin that have motif hits</p></li>
<li><p>sumBackgroundWgtWithHits: Sum of background sequence weights
    in a bin that have motif hits</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function implements a binned motif enrichment analysis. In each
  enrichment analysis, the sequences in a specific bin are used as foreground
  sequences to test for motif enrichment comparing to background sequences
  (defined by <code>background</code>, see below). The logic follows the
  <code>findMotifsGenome.pl</code> tool from <code>Homer</code> version 4.11, with
  <code>-size given -nomotif -mknown</code> and additionally <code>-h</code> if using 
  <code>test = "fisher"</code>, and gives very similar results.
  As in the <code>Homer</code> tool, sequences are weighted to correct for GC and
  k-mer composition differences between fore- and background sets.</p>  
<p>The background sequences are defined according to the value of the
  <code>background</code> argument:</p><ul>
<li><p>otherBins: sequences from all other bins (excluding the current bin)</p></li>
<li><p>allBins: sequences from all bins (including the current bin)</p></li>
<li><p>zeroBin: sequences from the "zero bin", defined by the
      <code>maxAbsX</code> argument of <code><a href='bin.html'>bin</a></code>. If <code>bins</code>
      does not define a "zero bin", for example because it was created by
      <code>bin(..., maxAbsX = NULL)</code>, selecting this background definition
      will abort with an error.</p></li>
<li><p>genome: sequences randomly sampled from the genome (or the
      intervals defined in <code>genome.regions</code> if given). For each
      foreground sequence, <code>genome.oversample</code> background sequences
      of the same size are sampled (on average). From these, one per
      foreground sequence is selected trying to match the G+C composition.
      In order to make the sampling deterministic, the random number
      generator is seeded using <code>genome.seed</code>.</p></li>
</ul>

<p>Motif hits are predicted using <code><a href='findMotifHits-methods.html'>findMotifHits</a></code> and
  multiple hits per sequence are counted as just one hit (ZOOPS mode). For
  each motif, the weights of sequences that have a hit are summed separately
  for foreground (<code>sumForegroundWgtWithHits</code>) and background
  (<code>sumBackgroundWgtWithHits</code>). The total foreground 
  (<code>totalWgtForeground</code>) and background (<code>totalWgtBackground</code>)
  sum of sequence weights is also calculated. If a motif has zero 
  (<code>sumForegroundWgtWithHits</code>) and (<code>sumBackgroundWgtWithHits</code>), 
  then any values (p-values and enrichment) that are calculated using 
  these two numbers are set to NA.</p>
<p>Two statistical tests for the calculation of enrichment log p-value are
  available: <code>test = "fisher"</code> (default) to perform Fisher's exact
  tests, or <code>test = "binomial"</code> to perform binomial tests
  (default in <code>Homer</code>), using:</p><ul>
<li><p>fisher: <code>fisher.test(x = tab, alternative =
      "greater")</code>, where <code>tab</code> is the contingency table with the summed
      weights of sequences in foreground or background sets (rows), and with
      or without a hit for a particular motif (columns).</p></li>
<li><p>binomial: <code>pbinom(q = sumForegroundWgtWithHits - 1, size =
      totalWgtForeground, prob = sumBackgroundWgtWithHits / totalWgtBackground,
      lower.tail = FALSE, log.p = TRUE)</code></p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


