<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Binned Motif Enrichment Analysis with monaLisa — calcBinnedMotifEnrR • monaLisa</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Binned Motif Enrichment Analysis with monaLisa — calcBinnedMotifEnrR"><meta property="og:description" content="This function performs a motif enrichment analysis on bins of
  sequences. For each bin, the sequences in all other bins are used as
  background."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Regression Based Approach for Motif Selection</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fmicompbio/monaLisa" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Binned Motif Enrichment Analysis with <code>monaLisa</code></h1>
    <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/HEAD/R/motif_enrichment_monaLisa.R" class="external-link"><code>R/motif_enrichment_monaLisa.R</code></a></small>
    <div class="hidden name"><code>calcBinnedMotifEnrR.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs a motif enrichment analysis on bins of
  sequences. For each bin, the sequences in all other bins are used as
  background.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calcBinnedMotifEnrR</span><span class="op">(</span></span>
<span>  <span class="va">seqs</span>,</span>
<span>  bins <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pwmL <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  background <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"otherBins"</span>, <span class="st">"allBins"</span>, <span class="st">"zeroBin"</span>, <span class="st">"genome"</span><span class="op">)</span>,</span>
<span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fisher"</span>, <span class="st">"binomial"</span><span class="op">)</span>,</span>
<span>  maxFracN <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  maxKmerSize <span class="op">=</span> <span class="fl">3L</span>,</span>
<span>  min.score <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  matchMethod <span class="op">=</span> <span class="st">"matchPWM"</span>,</span>
<span>  GCbreaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.25</span>, <span class="fl">0.3</span>, <span class="fl">0.35</span>, <span class="fl">0.4</span>, <span class="fl">0.45</span>, <span class="fl">0.5</span>, <span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>  pseudocount.log2enr <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome.regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome.oversample <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt id="arg-seqs">seqs<a class="anchor" aria-label="anchor" href="#arg-seqs"></a></dt>
<dd><p><code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></code> object with sequences to
test</p></dd>


<dt id="arg-bins">bins<a class="anchor" aria-label="anchor" href="#arg-bins"></a></dt>
<dd><p>factor of the same length and order as <code>seqs</code>, indicating
the bin for each sequence. Typically the return value of
<code><a href="bin.html">bin</a></code>. For <code>background = "genome"</code>, <code>bins</code>
can be omitted.</p></dd>


<dt id="arg-pwml">pwmL<a class="anchor" aria-label="anchor" href="#arg-pwml"></a></dt>
<dd><p>PWMatrixList with motifs for which to calculate enrichments.</p></dd>


<dt id="arg-background">background<a class="anchor" aria-label="anchor" href="#arg-background"></a></dt>
<dd><p>A <code>character</code> scalar specifying the background
sequences to use. One of <code>"otherBins"</code> (default), <code>"allBins"</code>,
<code>"zeroBin"</code> or <code>"genome"</code> (see "Details").</p></dd>


<dt id="arg-test">test<a class="anchor" aria-label="anchor" href="#arg-test"></a></dt>
<dd><p>A <code>character</code> scalar specifying the type of enrichment test
to perform. One of <code>"fisher"</code> (default) or <code>"binomial"</code>. The
enrichment test is one-sided (enriched in foreground).</p></dd>


<dt id="arg-maxfracn">maxFracN<a class="anchor" aria-label="anchor" href="#arg-maxfracn"></a></dt>
<dd><p>A numeric scalar with the maximal fraction of N bases allowed
in a sequence (defaults to 0.7). Sequences with higher fractions are
excluded from the analysis.</p></dd>


<dt id="arg-maxkmersize">maxKmerSize<a class="anchor" aria-label="anchor" href="#arg-maxkmersize"></a></dt>
<dd><p>the maximum k-mer size to consider, when adjusting
background sequence weights for k-mer composition compared to the
foreground sequences. The default value (3) will correct for mono-, di-
and tri-mer composition.</p></dd>


<dt id="arg-min-score">min.score<a class="anchor" aria-label="anchor" href="#arg-min-score"></a></dt>
<dd><p>the minimal score for motif hits, used in
<code><a href="findMotifHits-methods.html">findMotifHits</a></code>.</p></dd>


<dt id="arg-matchmethod">matchMethod<a class="anchor" aria-label="anchor" href="#arg-matchmethod"></a></dt>
<dd><p>the method used to scan for motif hits, passed to the
<code>method</code> parameter in <code><a href="findMotifHits-methods.html">findMotifHits</a></code>.</p></dd>


<dt id="arg-gcbreaks">GCbreaks<a class="anchor" aria-label="anchor" href="#arg-gcbreaks"></a></dt>
<dd><p>The breaks between GC bins. The default value is based on
the hard-coded bins used in Homer.</p></dd>


<dt id="arg-pseudocount-log-enr">pseudocount.log2enr<a class="anchor" aria-label="anchor" href="#arg-pseudocount-log-enr"></a></dt>
<dd><p>A numerical scalar with the pseudocount to add to
foreground and background counts when calculating log2 motif enrichments</p></dd>


<dt id="arg-p-adjust-method">p.adjust.method<a class="anchor" aria-label="anchor" href="#arg-p-adjust-method"></a></dt>
<dd><p>A character scalar selecting the p value adjustment
method (used in <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>).</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>A <code>BSgenome</code> or <code>DNAStringSet</code> object with the
genome sequence. Only used for <code>background = "genome"</code> for extracting
background sequences.</p></dd>


<dt id="arg-genome-regions">genome.regions<a class="anchor" aria-label="anchor" href="#arg-genome-regions"></a></dt>
<dd><p>An optional <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object
defining the intervals in <code>genome</code> from which background sequences are
sampled for <code>background = "genome"</code>. If <code>NULL</code>, background
sequences are sampled randomly from <code>genome</code>.</p></dd>


<dt id="arg-genome-oversample">genome.oversample<a class="anchor" aria-label="anchor" href="#arg-genome-oversample"></a></dt>
<dd><p>A <code>numeric</code> scalar of at least 1.0 defining how
many background sequences will be sampled per foreground sequence for
<code>background = "genome"</code>. Larger values will take longer but improve
the sequence composition similarity between foreground and background
(see <code>"Details"</code>).</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>An optional <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallelParam</a></code>
instance determining the parallel back-end to be used during evaluation.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>A logical scalar. If <code>TRUE</code>, print progress messages.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments for  <code><a href="findMotifHits-methods.html">findMotifHits</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object
  with motifs in rows and bins in columns, containing seven assays:</p><dl><dt>negLog10P</dt>
<dd><p>: -log10 P values</p></dd>

  <dt>negLog10Padj</dt>
<dd><p>: -log10 adjusted P values</p></dd>

  <dt>pearsonResid</dt>
<dd><p>: motif enrichments as Pearson residuals</p></dd>

  <dt>expForegroundWgtWithHits</dt>
<dd><p>: expected number of foreground
    sequences with motif hits</p></dd>

  <dt>log2enr</dt>
<dd><p>: motif enrichments as log2 ratios</p></dd>

  <dt>sumForegroundWgtWithHits</dt>
<dd><p>: Sum of foreground sequence weights
    in a bin that have motif hits</p></dd>

  <dt>sumBackgroundWgtWithHits</dt>
<dd><p>: Sum of background sequence weights
    in a bin that have motif hits</p></dd>


</dl><p>The <code>rowData</code> of the object contains annotations (name, PFMs, PWMs
and GC fraction) for the motifs, while the <code>colData</code> slot contains
summary information about the bins.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function implements a binned motif enrichment analysis. In each
  enrichment analysis, the sequences in a specific bin are used as foreground
  sequences to test for motif enrichments comparing to background sequences
  (defined by <code>background</code>, see below). The logic follows the
  <code>findMotifsGenome.pl</code> tool from <code>Homer</code> version 4.11, with
  <code>-size given -nomotif -mknown</code> and additionally <code>-h</code> if using
  <code>test = "fisher"</code>, and gives very similar results. As in the
  <code>Homer</code> tool, sequences are weighted to correct for GC and k-mer
  composition differences between fore- and background sets.</p>
<p>The background sequences are defined according to the value of the
  <code>background</code> argument:</p><dl><dt>otherBins</dt>
<dd><p>: sequences from all other bins (excluding the current
      bin)</p></dd>

    <dt>allBins</dt>
<dd><p>: sequences from all bins (including the current bin)</p></dd>

    <dt>zeroBin</dt>
<dd><p>: sequences from the "zero bin", defined by the
      <code>maxAbsX</code> argument of <code><a href="bin.html">bin</a></code>. If <code>bins</code>
      does not define a "zero bin", for example because it was created by
      <code>bin(..., maxAbsX = NULL)</code>, selecting this background definition
      will abort with an error.</p></dd>

    <dt>genome</dt>
<dd><p>: sequences randomly sampled from the genome (or the
      intervals defined in <code>genome.regions</code> if given). For each
      foreground sequence, <code>genome.oversample</code> background sequences
      of the same size are sampled (on average). From these, one per
      foreground sequence is selected trying to match the G+C composition.
      In order to make the sampling deterministic, a seed number needs to be
      provided to the <code>RNGseed</code> parameter in
      <code><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></code>
      or <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></code> when creating the
      <code>BiocParallelParam</code> instance in <code>BPPARAM</code>.</p></dd>

  
</dl><p>Motif hits are predicted using <code><a href="findMotifHits-methods.html">findMotifHits</a></code> and
  multiple hits per sequence are counted as just one hit (ZOOPS mode). For
  each motif, the weights of sequences that have a hit are summed separately
  for foreground (<code>sumForegroundWgtWithHits</code>) and background
  (<code>sumBackgroundWgtWithHits</code>). The total foreground
  (<code>totalWgtForeground</code>) and background (<code>totalWgtBackground</code>)
  sum of sequence weights is also calculated. If a motif has zero
  <code>sumForegroundWgtWithHits</code> and <code>sumBackgroundWgtWithHits</code>,
  then any values (p-values and enrichment) that are calculated using
  these two numbers are set to NA.</p>
<p>Two statistical tests for the calculation of enrichment log p-value are
  available: <code>test = "fisher"</code> (default) to perform Fisher's exact
  tests, or <code>test = "binomial"</code> to perform binomial tests
  (default in <code>Homer</code>), using:</p><dl><dt>fisher</dt>
<dd><p>: <code>fisher.test(x = tab, alternative =
      "greater")</code>, where <code>tab</code> is the contingency table with the summed
      weights of sequences in foreground or background sets (rows), and with
      or without a hit for a particular motif (columns).</p></dd>

    <dt>binomial</dt>
<dd><p>: <code>pbinom(q = sumForegroundWgtWithHits - 1, size =
      totalWgtForeground,
      prob = sumBackgroundWgtWithHits / totalWgtBackground,
      lower.tail = FALSE, log.p = TRUE)</code></p></dd>

  
</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GTCAGTCGATC"</span>, <span class="st">"CAGTCTAGCTG"</span>,</span></span>
<span class="r-in"><span>                                   <span class="st">"CGATCGTCAGT"</span>, <span class="st">"AGCTGCAGTCT"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           C <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           G <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">PWMatrixList</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">PWMatrix</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="st">"m1"</span>, profileMatrix <span class="op">=</span> <span class="va">m</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">PWMatrix</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="st">"m2"</span>, profileMatrix <span class="op">=</span> <span class="va">m</span><span class="op">[</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">calcBinnedMotifEnrR</span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">seqs</span>, bins <span class="op">=</span> <span class="va">bins</span>, pwmL <span class="op">=</span> <span class="va">pwms</span>,</span></span>
<span class="r-in"><span>                    min.score <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 2 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(5): bins bins.binmode bins.breaks bins.bin0 param</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(7): negLog10P negLog10Padj ... sumForegroundWgtWithHits</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   sumBackgroundWgtWithHits</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(2): m1 m2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(5): motif.id motif.name motif.pfm motif.pwm</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   motif.percentGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(2): 1 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(6): bin.names bin.lower ... totalWgtForeground</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   totalWgtBackground</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

