<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Prepare and run HOMER motif enrichment analysis. — calcBinnedMotifEnrHomer • monaLisa</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Prepare and run HOMER motif enrichment analysis. — calcBinnedMotifEnrHomer"><meta property="og:description" content="Run complete HOMER motif enrichment analysis, consisting of
    calls to prepareHomer, system2 and
    parseHomerOutput. This function requires HOMER
    to be installed (see http://homer.ucsd.edu/homer/index.html)
    and the path to the tool to be provided (homerfile argument)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/monaLisa.html">Binned motif enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fmicompbio/monaLisa" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Prepare and run HOMER motif enrichment analysis.</h1>
    <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/HEAD/R/motif_enrichment_HOMER.R" class="external-link"><code>R/motif_enrichment_HOMER.R</code></a></small>
    <div class="hidden name"><code>calcBinnedMotifEnrHomer.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run complete HOMER motif enrichment analysis, consisting of
    calls to <code><a href="prepareHomer.html">prepareHomer</a></code>, <code><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></code> and
    <code><a href="parseHomerOutput.html">parseHomerOutput</a></code>. This function requires <code>HOMER</code>
    to be installed (see <a href="http://homer.ucsd.edu/homer/index.html" class="external-link">http://homer.ucsd.edu/homer/index.html</a>)
    and the path to the tool to be provided (<code>homerfile</code> argument).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">calcBinnedMotifEnrHomer</span><span class="op">(</span>
  <span class="va">gr</span>,
  <span class="va">b</span>,
  <span class="va">genomedir</span>,
  <span class="va">outdir</span>,
  <span class="va">motifFile</span>,
  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="op">)</span>,
  regionsize <span class="op">=</span> <span class="st">"given"</span>,
  pseudocount.log2enr <span class="op">=</span> <span class="fl">8</span>,
  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,
  Ncpu <span class="op">=</span> <span class="fl">2L</span>,
  verbose <span class="op">=</span> <span class="cn">FALSE</span>,
  verbose.Homer <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gr</dt>
<dd><p>A <code>GRanges</code> object (or an object that can be coerced to one)
with the genomic regions to analyze.</p></dd>
<dt>b</dt>
<dd><p>A vector of the same length as <code>gr</code> that groups its elements
into bins (typically a factor, such as the one returned by 
<code><a href="bin.html">bin</a></code>).</p></dd>
<dt>genomedir</dt>
<dd><p>Directory containing sequence files in Fasta format 
(one per chromosome).</p></dd>
<dt>outdir</dt>
<dd><p>A path specifying the folder into which the output files will 
be written.</p></dd>
<dt>motifFile</dt>
<dd><p>A file with HOMER formatted PWMs to be used in the 
enrichment analysis.</p></dd>
<dt>homerfile</dt>
<dd><p>Path and file name of the <code>findMotifsGenome.pl</code> 
HOMER script.</p></dd>
<dt>regionsize</dt>
<dd><p>The peak size to use in HOMER (<code>"given"</code> keeps the 
coordinate region, an integer value will keep only that many bases in 
the region center).</p></dd>
<dt>pseudocount.log2enr</dt>
<dd><p>A numerical scalar with the pseudocount to add to
foreground and background counts when calculating log2 motif enrichments</p></dd>
<dt>p.adjust.method</dt>
<dd><p>A character scalar selecting the p value adjustment
method (used in <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>).</p></dd>
<dt>Ncpu</dt>
<dd><p>Number of parallel threads that HOMER can use.</p></dd>
<dt>verbose</dt>
<dd><p>A logical scalar. If <code>TRUE</code>, print progress messages.</p></dd>
<dt>verbose.Homer</dt>
<dd><p>A logical scalar. If <code>TRUE</code>, print the console
output when running Homer.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>SummarizedExperiment</code> object with motifs in rows and bins
  in columns, containing seven assays:</p><ul><li><p>negLog10P: -log10 P values</p></li>
<li><p>negLog10Padj: -log10 adjusted P values</p></li>
<li><p>pearsonResid: motif enrichments as Pearson residuals</p></li>
<li><p>expForegroundWgtWithHits: expected number of foreground 
    sequences with motif hits</p></li>
<li><p>log2enr: motif enrichments as log2 ratios</p></li>
<li><p>sumForegroundWgtWithHits: Sum of foreground sequence weights
    in a bin that have motif hits</p></li>
<li><p>sumBackgroundWgtWithHits: Sum of background sequence weights
    in a bin that have motif hits</p></li>
</ul><p>The <code>rowData</code> of the object contains annotations (name, PFMs, PWMs 
and GC fraction) for the motifs, while the <code>colData</code> slot contains 
summary information about the bins.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>The functions that are wrapped: <code><a href="prepareHomer.html">prepareHomer</a></code>,
    <code><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></code> and <code><a href="parseHomerOutput.html">parseHomerOutput</a></code>,
    <code><a href="bin.html">bin</a></code> for binning of regions</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span class="r-in"></span>
<span class="r-in">  <span class="co"># genome</span></span>
<span class="r-in">  <span class="va">genome</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"exampleGenome.fa"</span>, package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span></span>
<span class="r-in">  </span>
<span class="r-in">  <span class="co"># create motif file for Homer</span></span>
<span class="r-in">  <span class="va">motiffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">motifIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MA0139.1"</span>, <span class="st">"MA1102.1"</span>, <span class="st">"MA0740.1"</span><span class="op">)</span></span>
<span class="r-in">  <span class="fu"><a href="dumpJaspar.html">dumpJaspar</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">motiffile</span>, pkg <span class="op">=</span> <span class="st">"JASPAR2020"</span>, </span>
<span class="r-in">             opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="va">motifIDs</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">             </span>
<span class="r-in">  <span class="co"># GRanges of regions used in binned motif enrichment analysis </span></span>
<span class="r-in">  <span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/tileGenome.html" class="external-link">tileGenome</a></span><span class="op">(</span></span>
<span class="r-in">      seqlengths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>chr1 <span class="op">=</span> <span class="fl">10000L</span>, chr2 <span class="op">=</span> <span class="fl">10000L</span>, chr3 <span class="op">=</span> <span class="fl">10000L</span><span class="op">)</span>,</span>
<span class="r-in">      tilewidth <span class="op">=</span> <span class="fl">200</span>, cut.last.tile.in.chrom <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in">      </span>
<span class="r-in">  <span class="co"># create bins (motif enrichment analysis will be per bin)</span></span>
<span class="r-in">  <span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span></span>
<span class="r-in">  </span>
<span class="r-in">  <span class="co"># run calcBinnedMotifEnrHomer</span></span>
<span class="r-in">  <span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">calcBinnedMotifEnrHomer</span><span class="op">(</span>gr <span class="op">=</span> <span class="va">gr</span>, b <span class="op">=</span> <span class="va">bins</span>, genomedir <span class="op">=</span> <span class="va">genome</span>,</span>
<span class="r-in">      outdir <span class="op">=</span> <span class="va">outdir</span>, motifFile <span class="op">=</span> <span class="va">motiffile</span><span class="op">)</span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span class="r-in">  </span>
<span class="r-in">  <span class="op">}</span></span>
<span class="r-in">  </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

