<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Adjust for k-mer composition (multiple iterations) — .iterativeNormForKmers • monaLisa</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Adjust for k-mer composition (multiple iterations) — .iterativeNormForKmers"><meta property="og:description" content="Here we run `.normForKmers` multiple times to converge to
  the final weights that will be used to correct the background
  sequences for k-mer composition differences compared to the foreground. We
  closely follow HOMER's normalizeSequence() function found in
  Motif2.cpp. Note that HOMER runs the
  normalizeSequence() one last time after going through all iterations
  or reaching a low error, which we do not do here."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/monaLisa.html">Binned motif enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fmicompbio/monaLisa">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Adjust for k-mer composition (multiple iterations)</h1>
    <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/master/R/utils_enrichment.R"><code>R/utils_enrichment.R</code></a></small>
    <div class="hidden name"><code>dot-iterativeNormForKmers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Here we run `.normForKmers` multiple times to converge to
  the final weights that will be used to correct the background
  sequences for k-mer composition differences compared to the foreground. We
  closely follow <code>HOMER</code>'s <code>normalizeSequence()</code> function found in
  <code>Motif2.cpp</code>. Note that <code>HOMER</code> runs the
  <code>normalizeSequence()</code> one last time after going through all iterations
  or reaching a low error, which we do not do here.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class="sourceCode r"><code><span class="fu">.iterativeNormForKmers</span><span class="op">(</span>
  <span class="va">df</span>,
  maxKmerSize <span class="op">=</span> <span class="fl">3L</span>,
  minSeqWgt <span class="op">=</span> <span class="fl">0.001</span>,
  maxIter <span class="op">=</span> <span class="fl">160L</span>,
  verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments"><colgroup><col class="name"><col class="desc"></colgroup><tr><th>df</th>
      <td><p>a <code>DataFrame</code> with sequence information as returned by
<code>.calculateGCweight</code>.</p></td>
    </tr><tr><th>maxKmerSize</th>
      <td><p>Integer scalar giving the maximum k-mer size to
consider. The default is set to 3 (like in <code>HOMER</code>), meaning that
k-mers of size 1, 2 and 3 are considered.</p></td>
    </tr><tr><th>minSeqWgt</th>
      <td><p>Numeric scalar greater than zero giving the
minimal weight of a sequence. The default value (0.001) was also used by
<code>HOMER</code> (HOMER_MINIMUM_SEQ_WEIGHT  constant in Motif2.h).</p></td>
    </tr><tr><th>maxIter</th>
      <td><p>An integer scalar giving the maximum number if
times to run <code>.normForKmers</code>. the default is set to 160 (as in
<code>HOMER</code>).</p></td>
    </tr><tr><th>verbose</th>
      <td><p>A logical scalar. If <code>TRUE</code>, report on k-mer composition
adjustment.</p></td>
    </tr></table><h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a DataFrame containing:</p><ul><li><p>sequenceWeights: a
  <code>dataframe</code> containing the sequence GC content, GC bins they were
  assigned to, the weight to correct for GC differences between foreGround
  and background sequences, the weight to adjust for kmer composition, and
  the the error term</p></li>
<li><p>sequenceNucleotides: a <code>DNAStringSet</code>
  object containing the raw sequences</p></li>
</ul></div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

