<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Adjust for k-mer composition (multiple iterations) — .iterativeNormForKmers • monaLisa</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Adjust for k-mer composition (multiple iterations) — .iterativeNormForKmers"><meta name="description" content="Here we run `.normForKmers` multiple times to converge to
  the final weights that will be used to correct the background
  sequences for k-mer composition differences compared to the foreground. We
  closely follow HOMER's normalizeSequence() function found in
  Motif2.cpp. Note that HOMER runs the
  normalizeSequence() one last time after going through all iterations
  or reaching a low error, which we do not do here."><meta property="og:description" content="Here we run `.normForKmers` multiple times to converge to
  the final weights that will be used to correct the background
  sequences for k-mer composition differences compared to the foreground. We
  closely follow HOMER's normalizeSequence() function found in
  Motif2.cpp. Note that HOMER runs the
  normalizeSequence() one last time after going through all iterations
  or reaching a low error, which we do not do here."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/monaLisa.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Adjust for k-mer composition (multiple iterations)</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/devel/R/utils_enrichment.R" class="external-link"><code>R/utils_enrichment.R</code></a></small>
      <div class="d-none name"><code>dot-iterativeNormForKmers.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Here we run `.normForKmers` multiple times to converge to
  the final weights that will be used to correct the background
  sequences for k-mer composition differences compared to the foreground. We
  closely follow <code>HOMER</code>'s <code>normalizeSequence()</code> function found in
  <code>Motif2.cpp</code>. Note that <code>HOMER</code> runs the
  <code>normalizeSequence()</code> one last time after going through all iterations
  or reaching a low error, which we do not do here.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">.iterativeNormForKmers</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  maxKmerSize <span class="op">=</span> <span class="fl">3L</span>,</span>
<span>  minSeqWgt <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  maxIter <span class="op">=</span> <span class="fl">160L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-df">df<a class="anchor" aria-label="anchor" href="#arg-df"></a></dt>
<dd><p>a <code>DataFrame</code> with sequence information as returned by
<code>.calculateGCweight</code>.</p></dd>


<dt id="arg-maxkmersize">maxKmerSize<a class="anchor" aria-label="anchor" href="#arg-maxkmersize"></a></dt>
<dd><p>Integer scalar giving the maximum k-mer size to
consider. The default is set to 3 (like in <code>HOMER</code>), meaning that
k-mers of size 1, 2 and 3 are considered.</p></dd>


<dt id="arg-minseqwgt">minSeqWgt<a class="anchor" aria-label="anchor" href="#arg-minseqwgt"></a></dt>
<dd><p>Numeric scalar greater than zero giving the
minimal weight of a sequence. The default value (0.001) was also used by
<code>HOMER</code> (HOMER_MINIMUM_SEQ_WEIGHT  constant in Motif2.h).</p></dd>


<dt id="arg-maxiter">maxIter<a class="anchor" aria-label="anchor" href="#arg-maxiter"></a></dt>
<dd><p>An integer scalar giving the maximum number if
times to run <code>.normForKmers</code>. the default is set to 160 (as in
<code>HOMER</code>).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>A logical scalar. If <code>TRUE</code>, report on k-mer composition
adjustment.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a DataFrame containing:</p><dl><dt>sequenceWeights</dt>
<dd><p>: a
  <code>dataframe</code> containing the sequence GC content, GC bins they were
  assigned to, the weight to correct for GC differences between foreGround
  and background sequences, the weight to adjust for kmer composition, and
  the the error term</p></dd>
 <dt>sequenceNucleotides</dt>
<dd><p>: a <code>DNAStringSet</code>
  object containing the raw sequences</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

