<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Find motif matches in sequences. — findMotifHits • monaLisa</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find motif matches in sequences. — findMotifHits"><meta name="description" content="findMotifHits scans sequences (either provided
as a file, an R object or genomic coordinates)  for matches to
positional weight matrices (provided as a file or as R objects)"><meta property="og:description" content="findMotifHits scans sequences (either provided
as a file, an R object or genomic coordinates)  for matches to
positional weight matrices (provided as a file or as R objects)"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/monaLisa.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Find motif matches in sequences.</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/devel/R/motif_finding.R" class="external-link"><code>R/motif_finding.R</code></a></small>
      <div class="d-none name"><code>findMotifHits-methods.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>findMotifHits</code> scans sequences (either provided
as a file, an R object or genomic coordinates)  for matches to
positional weight matrices (provided as a file or as R objects)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'character,character'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'character,DNAString'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'character,DNAStringSet'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrix,character'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrix,DNAString'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrix,DNAStringSet'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrixList,character'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrixList,DNAString'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrixList,DNAStringSet'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrix,GRanges'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'PWMatrixList,GRanges'</span></span>
<span><span class="fu">findMotifHits</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  <span class="va">min.score</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"matchPWM"</span>, <span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  homerfile <span class="op">=</span> <span class="fu"><a href="findHomer.html">findHomer</a></span><span class="op">(</span><span class="st">"homer2"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-query">query<a class="anchor" aria-label="anchor" href="#arg-query"></a></dt>
<dd><p>The motifs to search for, either a</p><dl><dt><code>character(1)</code></dt>
<dd><p>with the path and file name of a motif
          file with PWM in HOMER format (currently only
          supported for <code>method="homer2"</code>)</p></dd>

    <dt><code>PWMatrix</code></dt>
<dd><p>with a single PWM</p></dd>

    <dt><code>PWMatrixList</code></dt>
<dd><p>with several PWMs to search for.</p></dd>


</dl></dd>


<dt id="arg-subject">subject<a class="anchor" aria-label="anchor" href="#arg-subject"></a></dt>
<dd><p>The sequences to be searched, either a</p><dl><dt><code>character</code></dt>
<dd><p>with the path and file name of a sequence
          file with DNA sequences in FASTA format</p></dd>

    <dt><code>DNAString</code></dt>
<dd><p>with a single sequence</p></dd>

    <dt><code>DNAStringSet</code></dt>
<dd><p>with several sequences</p></dd>

    <dt><code>GRanges</code></dt>
<dd><p>object with the genomic coordinates
        of the sequences to be searched.</p></dd>


</dl></dd>


<dt id="arg-min-score">min.score<a class="anchor" aria-label="anchor" href="#arg-min-score"></a></dt>
<dd><p>The minimum score for counting a match. Can be given as
a character string containing a percentage (e.g. "85<!-- %") of  of the -->
highest possible score or as a single number.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>The internal method to use for motif searching. One of</p><dl><dt><code>"matchPWM"</code></dt>
<dd><p>using Biostrings::matchPWM (optimized)</p></dd>

    <dt><code>"homer2"</code></dt>
<dd><p>call to the homer2 binary</p></dd>


</dl><p>Please note that the two methods might give slightly different results
(see details).</p></dd>


<dt id="arg-homerfile">homerfile<a class="anchor" aria-label="anchor" href="#arg-homerfile"></a></dt>
<dd><p>Path and file name of the <code>homer2</code> binary.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>An optional <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallelParam</a></code>
instance determining the parallel back-end to be used during evaluation.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p><code>BSgenome</code> object that is the reference genome of the
subject. This argument is set to NULL by default and only used by the
function when the subject is a <code>GRanges</code> object. It is then
necessary to specify the genome so that the function can internally
convert the genomic regions into a <code>DNAStringSet</code> object.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>GRanges</code> object with the matches to <code>query</code> in
    <code>subject</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The implemented methods (<code>matchPWM</code> and <code>homer2</code>) are
    there for convenience (<code>method="matchPWM"</code> calls
    <code><a href="https://rdrr.io/pkg/Biostrings/man/matchPWM.html" class="external-link">Biostrings::matchPWM</a></code> internally in an optimized fashion, and
    <code>method = "homer2"</code> calls the command line tool from Homer and
    therefore requires an installation of Homer).</p>
<p>In general, running <code>findMotifHits</code> with the same parameters using
    any of the methods generates identical results. Some minor differences
    could occur that result from rounding errors during the necessary
    conversion of PWMs (log2-odd scores) to the probability matrices needed
    by Homer, and the conversion of scores from and to the natural log scale
    used by Homer. These conversions are implemented transparently for the
    user, so that the arguments of <code>findMotifHits</code> do not have to be
    adjusted (e.g. the PWMs should always contain log2-odd scores, and
    <code>min.score</code> is always on the log2 scale).</p>
<p>If there are bases with frequencies of less than 0.001 in a motif, Homer
    will set them to 0.001 and adjust the other frequencies at that motif
    position accordingly so that they sum to 1.0. This may differ from the
    adjustment used when scanning a PWM with <code>matchPWM</code> (e.g. the
    <code>pseudocounts</code> argument in the <code><a href="https://rdrr.io/pkg/TFBSTools/man/toPWM-methods.html" class="external-link">toPWM</a></code>
    function), and thus can give rise to differences in reported motif hits
    and hit scores (typically only low-scoring hits).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>s1 <span class="op">=</span> <span class="st">"GTCAGTCGATC"</span>, s2 <span class="op">=</span> <span class="st">"CAGTCTAGCTG"</span>,</span></span>
<span class="r-in"><span>                                   s3 <span class="op">=</span> <span class="st">"CGATCGTCAGT"</span>, s4 <span class="op">=</span> <span class="st">"AGCTGCAGTCT"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           C <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           G <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">PWMatrixList</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">PWMatrix</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="st">"m1"</span>, profileMatrix <span class="op">=</span> <span class="va">m</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">PWMatrix</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="st">"m2"</span>, profileMatrix <span class="op">=</span> <span class="va">m</span><span class="op">[</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">findMotifHits</span><span class="op">(</span><span class="va">pwms</span>, <span class="va">seqs</span>, min.score <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 6 ranges and 4 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |     matchedSeq pwmid pwmname     score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;DNAStringSet&gt; &lt;Rle&gt;   &lt;Rle&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]       s1       4-6      + |            AGT    m1 Unknown         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]       s1       2-4      - |            TGA    m2 Unknown         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]       s2       2-4      + |            AGT    m1 Unknown         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]       s3      9-11      + |            AGT    m1 Unknown         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]       s3       7-9      - |            TGA    m2 Unknown         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]       s4       7-9      + |            AGT    m1 Unknown         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 4 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

