<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate observed and expected k-mer frequencies — getKmerFreq • monaLisa</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate observed and expected k-mer frequencies — getKmerFreq"><meta name="description" content="Given a set of sequences, calculate observed and expected k-mer
  frequencies. Expected frequencies are based on a Markov model of order
  MMorder."><meta property="og:description" content="Given a set of sequences, calculate observed and expected k-mer
  frequencies. Expected frequencies are based on a Markov model of order
  MMorder."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/monaLisa.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate observed and expected k-mer frequencies</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/devel/R/motif_enrichment_kmers.R" class="external-link"><code>R/motif_enrichment_kmers.R</code></a></small>
      <div class="d-none name"><code>getKmerFreq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given a set of sequences, calculate observed and expected k-mer
  frequencies. Expected frequencies are based on a Markov model of order
  <code>MMorder</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getKmerFreq</span><span class="op">(</span></span>
<span>  <span class="va">seqs</span>,</span>
<span>  kmerLen <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  MMorder <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  pseudocount <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  zoops <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  strata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1L</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  includeRevComp <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seqs">seqs<a class="anchor" aria-label="anchor" href="#arg-seqs"></a></dt>
<dd><p>Set of sequences, either a <code>character</code> vector or a
<code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></code>.</p></dd>


<dt id="arg-kmerlen">kmerLen<a class="anchor" aria-label="anchor" href="#arg-kmerlen"></a></dt>
<dd><p>A <code>numeric</code> scalar giving the k-mer length.</p></dd>


<dt id="arg-mmorder">MMorder<a class="anchor" aria-label="anchor" href="#arg-mmorder"></a></dt>
<dd><p>A <code>numeric</code> scalar giving the order of the Markov model
used to calculate the expected frequencies.</p></dd>


<dt id="arg-pseudocount">pseudocount<a class="anchor" aria-label="anchor" href="#arg-pseudocount"></a></dt>
<dd><p>A <code>numeric</code> scalar - will be added to the observed
counts for each k-mer to avoid zero values.</p></dd>


<dt id="arg-zoops">zoops<a class="anchor" aria-label="anchor" href="#arg-zoops"></a></dt>
<dd><p>A <code>logical</code> scalar. If <code>TRUE</code> (the default), only one
or zero occurences of a k-mer are considered per sequence.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>A <code>factor</code> or a <code>numeric</code> scalar defining the strata
of sequences. A separate Markov model and expected k-mer frequencies are
estimated for the set of sequences in each stratum (level in a
<code>strata</code> factor). If <code>strata</code> is a scalar value, it will be
interpreted as the number of strata to split the sequences into according
to their CpG observed-over-expected counts using
<code>kmeans(CpGoe, centers = strata)</code>.</p></dd>


<dt id="arg-p-adjust-method">p.adjust.method<a class="anchor" aria-label="anchor" href="#arg-p-adjust-method"></a></dt>
<dd><p>A character scalar selecting the p value adjustment
method (used in <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>).</p></dd>


<dt id="arg-includerevcomp">includeRevComp<a class="anchor" aria-label="anchor" href="#arg-includerevcomp"></a></dt>
<dd><p>A <code>logical</code> scalar. If <code>TRUE</code> (default),
count k-mer occurrences in both <code>seqs</code> and their reverse-complement,
by concatenating <code>seqs</code> and their reverse-complemented versions
before the counting. This is useful if motifs can be expected to occur
on any strand (e.g. DNA sequences of ChIP-seq peaks). If motifs are only
expected on the forward strand (e.g. RNA sequences of CLIP-seq peaks),
<code>includeRevComp = FALSE</code> should be used. Note that if <code>strata</code>
is a vector of the same length as <code>seqs</code>, each reverse-complemented
sequence will be assigned to the same stratum as the forward sequence.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>list</code> with observed and expected k-mer frequencies
  (<code>freq.obs</code> and <code>freq.exp</code>, respectively), and enrichment
  statistics for each k-mer.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">getKmerFreq</span><span class="op">(</span>seqs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AAAAATT"</span>, <span class="st">"AAATTTT"</span><span class="op">)</span>, kmerLen <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "freq.obs"    "freq.exp"    "log2enr"     "sqrtDelta"   "z"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "p"           "padj"        "strata"      "freq.strata" "CpGoe"      </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">freq.obs</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAA AAC AAG AAT ACA ACC </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3   0   0   4   0   0 </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">freq.exp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       AAA       AAC       AAG       AAT       ACA       ACC </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1.8295455 0.1663223 0.1663223 1.4969008 0.3326446 0.3326446 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

