<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate similarities between motifs and k-mers. — motifKmerSimilarity • monaLisa</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate similarities between motifs and k-mers. — motifKmerSimilarity"><meta name="description" content="For each motif, calculate it's similarity to all k-mers of
  length kmerLen, defined as the maximal probability of observing the
  k-mer given the base frequencies of the motif (the maximum is taken over
  for all possible ungapped alignments between motif and k-mer). If necessary
  matrices are padded on the sides with background base frequencies (assuming
  all bases to have a frequency of 0.25)."><meta property="og:description" content="For each motif, calculate it's similarity to all k-mers of
  length kmerLen, defined as the maximal probability of observing the
  k-mer given the base frequencies of the motif (the maximum is taken over
  for all possible ungapped alignments between motif and k-mer). If necessary
  matrices are padded on the sides with background base frequencies (assuming
  all bases to have a frequency of 0.25)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/monaLisa.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate similarities between motifs and k-mers.</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/devel/R/pfm_comparison.R" class="external-link"><code>R/pfm_comparison.R</code></a></small>
      <div class="d-none name"><code>motifKmerSimilarity.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For each motif, calculate it's similarity to all k-mers of
  length <code>kmerLen</code>, defined as the maximal probability of observing the
  k-mer given the base frequencies of the motif (the maximum is taken over
  for all possible ungapped alignments between motif and k-mer). If necessary
  matrices are padded on the sides with background base frequencies (assuming
  all bases to have a frequency of 0.25).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">motifKmerSimilarity</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  kmerLen <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  kmers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  includeRevComp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Either a <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">PFMatrixList</a></code>, or a character
scalar with a file containing motifs in HOMER format (used directly
<code>method = "HOMER"</code>, loaded into a
<code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">PFMatrixList</a></code> by <code><a href="homerToPFMatrixList.html">homerToPFMatrixList</a></code>
for <code>method = "R"</code>).</p></dd>


<dt id="arg-kmerlen">kmerLen<a class="anchor" aria-label="anchor" href="#arg-kmerlen"></a></dt>
<dd><p>A <code>numeric</code> scalar giving the k-mer length.</p></dd>


<dt id="arg-kmers">kmers<a class="anchor" aria-label="anchor" href="#arg-kmers"></a></dt>
<dd><p>Either a character vector of k-mers for which to calculate
the similarity to each motif, or <code>NULL</code>, in which case all k-mers
of length <code>kmerLen</code> are used.</p></dd>


<dt id="arg-includerevcomp">includeRevComp<a class="anchor" aria-label="anchor" href="#arg-includerevcomp"></a></dt>
<dd><p>A <code>logical</code> scalar. If set to <code>TRUE</code>, each
k-mer as well as its reverse complement is compared to each motif, and the
larger of the two similarities is returned.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>An optional <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallelParam</a></code>
instance determining the parallel back-end to be used during evaluation.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>A logical scalar. If <code>TRUE</code>, report on progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix of probabilties for each motif - k-mer pair.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html" class="external-link">bplapply</a></code> used for parallelization.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">12</span>,  <span class="fl">0</span>,  <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           C <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">3</span>,  <span class="fl">2</span>,  <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           G <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">0</span>, <span class="fl">14</span>,  <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">0</span>,  <span class="fl">0</span>, <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pfms</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">PFMatrixList</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">PFMatrix</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"m1"</span>, profileMatrix <span class="op">=</span> <span class="va">m</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">PFMatrix</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"m2"</span>, profileMatrix <span class="op">=</span> <span class="va">m</span><span class="op">[</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">motifKmerSimilarity</span><span class="op">(</span><span class="va">pfms</span>, kmerLen <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AGT"</span>, <span class="st">"TGA"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       AGT    TGA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> m1 0.7000 0.0625</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> m2 0.0625 0.7000</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

