<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Heatmap of motif enrichments. — plotMotifHeatmaps • monaLisa</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap of motif enrichments. — plotMotifHeatmaps"><meta name="description" content="Plot motif enrichments (e.g. significance or magnitude) as a
    heatmap."><meta property="og:description" content="Plot motif enrichments (e.g. significance or magnitude) as a
    heatmap."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">monaLisa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/monaLisa.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/monaLisa.html">Binned motif enrichment analysis</a></li>
    <li><a class="dropdown-item" href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/monaLisa"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Heatmap of motif enrichments.</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/devel/R/plotting.R" class="external-link"><code>R/plotting.R</code></a></small>
      <div class="d-none name"><code>plotMotifHeatmaps.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot motif enrichments (e.g. significance or magnitude) as a
    heatmap.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotMotifHeatmaps</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  which.plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"negLog10P"</span>, <span class="st">"pearsonResid"</span>, <span class="st">"negLog10Padj"</span>, <span class="st">"log2enr"</span><span class="op">)</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  col.enr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#053061"</span>, <span class="st">"#2166AC"</span>, <span class="st">"#4393C3"</span>, <span class="st">"#92C5DE"</span>, <span class="st">"#D1E5F0"</span>, <span class="st">"#F7F7F7"</span>,</span>
<span>    <span class="st">"#FDDBC7"</span>, <span class="st">"#F4A582"</span>, <span class="st">"#D6604D"</span>, <span class="st">"#B2182B"</span>, <span class="st">"#67001F"</span><span class="op">)</span>,</span>
<span>  col.sig <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F0F0F0"</span>, <span class="st">"#D9D9D9"</span>, <span class="st">"#BDBDBD"</span>, <span class="st">"#969696"</span>, <span class="st">"#737373"</span>, <span class="st">"#525252"</span>,</span>
<span>    <span class="st">"#252525"</span>, <span class="st">"#000000"</span><span class="op">)</span>,</span>
<span>  col.gc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F7FCF5"</span>, <span class="st">"#E5F5E0"</span>, <span class="st">"#C7E9C0"</span>, <span class="st">"#A1D99B"</span>, <span class="st">"#74C476"</span>, <span class="st">"#41AB5D"</span>, <span class="st">"#238B45"</span>,</span>
<span>    <span class="st">"#006D2C"</span>, <span class="st">"#00441B"</span><span class="op">)</span>,</span>
<span>  maxEnr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxSig <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_dendrogram <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_motif_GC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_seqlogo <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_bin_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  width.seqlogo <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  use_raster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  na_col <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  doPlot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> with
numerical matrices (motifs-by-bins) in its <code>assays()</code>, typically
the return value of <code><a href="calcBinnedMotifEnrR.html">calcBinnedMotifEnrR</a></code> or
<code><a href="calcBinnedMotifEnrHomer.html">calcBinnedMotifEnrHomer</a></code>.</p></dd>


<dt id="arg-which-plots">which.plots<a class="anchor" aria-label="anchor" href="#arg-which-plots"></a></dt>
<dd><p>Selects which heatmaps to plot (one or several from
<code>"negLog10P"</code>, <code>"negLog10Padj"</code>, <code>"pearsonResid"</code> and
<code>"log2enr"</code>).</p></dd>


<dt id="arg-width">width<a class="anchor" aria-label="anchor" href="#arg-width"></a></dt>
<dd><p>The width (in inches) of each individual heatmap, without
legend.</p></dd>


<dt id="arg-col-enr">col.enr<a class="anchor" aria-label="anchor" href="#arg-col-enr"></a></dt>
<dd><p>Colors used for enrichment heatmap ("pearsonResid" and
"log2enr").</p></dd>


<dt id="arg-col-sig">col.sig<a class="anchor" aria-label="anchor" href="#arg-col-sig"></a></dt>
<dd><p>Colors used for significance hetmaps ("negLog10P" and
"negLog10Padj").</p></dd>


<dt id="arg-col-gc">col.gc<a class="anchor" aria-label="anchor" href="#arg-col-gc"></a></dt>
<dd><p>Colors used for motif GC content (for
<code>show_motif_GC = TRUE</code>).</p></dd>


<dt id="arg-maxenr">maxEnr<a class="anchor" aria-label="anchor" href="#arg-maxenr"></a></dt>
<dd><p>Cap color mapping at enrichment = <code>maxEnr</code>
(default: 99.5th percentile).</p></dd>


<dt id="arg-maxsig">maxSig<a class="anchor" aria-label="anchor" href="#arg-maxsig"></a></dt>
<dd><p>Cap color mapping at -log10 P value or -log10 FDR =
<code>maxSig</code> (default: 99.5th percentile).</p></dd>


<dt id="arg-highlight">highlight<a class="anchor" aria-label="anchor" href="#arg-highlight"></a></dt>
<dd><p>A logical vector indicating motifs to be highlighted.</p></dd>


<dt id="arg-cluster">cluster<a class="anchor" aria-label="anchor" href="#arg-cluster"></a></dt>
<dd><p>If <code>TRUE</code>, the order of transcription factors will be
determined by hierarchical clustering of the <code>"pearsonResid"</code>
component. Alternatively, an <code>hclust</code>-object can be supplied which
will determine the motif ordering.
No reordering is done for <code>cluster = FALSE</code>.</p></dd>


<dt id="arg-show-dendrogram">show_dendrogram<a class="anchor" aria-label="anchor" href="#arg-show-dendrogram"></a></dt>
<dd><p>If <code>cluster != FALSE</code>, controls whether to show
a row dendrogram for the clustering of motifs. Ignored for
<code>cluster = FALSE</code>.</p></dd>


<dt id="arg-show-motif-gc">show_motif_GC<a class="anchor" aria-label="anchor" href="#arg-show-motif-gc"></a></dt>
<dd><p>If <code>TRUE</code>, show a column with the percent G+C of
the motif as part of the heatmap.</p></dd>


<dt id="arg-show-seqlogo">show_seqlogo<a class="anchor" aria-label="anchor" href="#arg-show-seqlogo"></a></dt>
<dd><p>If <code>TRUE</code>, show a sequence logo next to each motif
label. This will likely only make sense for a heatmap with a low number
of motifs.</p></dd>


<dt id="arg-show-bin-legend">show_bin_legend<a class="anchor" aria-label="anchor" href="#arg-show-bin-legend"></a></dt>
<dd><p>If <code>TRUE</code>, show a legend for the bin labels.
If FALSE (default), the bin legend will be hidden.</p></dd>


<dt id="arg-width-seqlogo">width.seqlogo<a class="anchor" aria-label="anchor" href="#arg-width-seqlogo"></a></dt>
<dd><p>The width (in inches) for the longest sequence logo
(shorter logos are drawn to scale).</p></dd>


<dt id="arg-use-raster">use_raster<a class="anchor" aria-label="anchor" href="#arg-use-raster"></a></dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> (default). Passed to
<code>use_raster</code> of <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code>.</p></dd>


<dt id="arg-na-col">na_col<a class="anchor" aria-label="anchor" href="#arg-na-col"></a></dt>
<dd><p>"white" (default). Passed to <code>na_col</code> of
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code>.</p></dd>


<dt id="arg-doplot">doPlot<a class="anchor" aria-label="anchor" href="#arg-doplot"></a></dt>
<dd><p>If <code>TRUE</code> (default), plot the generated heatmap(s)
using <code>Reduce(ComplexHeatmap::add_heatmap, heatmapList)</code>. If
<code>FALSE</code>, just return the list of heatmap(s) (<code>heatmapList</code>) in
example before), allowing to modify them further before plotting.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code>
when creating the main heatmaps selected by <code>which.plots</code>. For
example, the following will set the font size of the motif names:
<code>plotMotifHeatmaps(..., row_names_gp = gpar(fontsize = 12))</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap</a></code> objects.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The heatmaps are created using the <span class="pkg">ComplexHeatmap</span> package
    and plotted side-by-side.</p>
<p>Each heatmap will be <code>width</code> inches wide, so the total plot needs a
    graphics device with a width of at least
    <code>length(which.plots) * width</code> plus the space used for motif names
    and legend. The height will be auto-adjusted to the graphics device.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Gu, Z. Complex heatmaps reveal patterns and correlations in
    multidimensional genomic data. Bioinformatics 2016.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="bin.html">bin</a></code>, <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, </span></span>
<span class="r-in"><span>                          <span class="st">"results.binned_motif_enrichment_LMRs.rds"</span>, </span></span>
<span class="r-in"><span>                          package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">8</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotMotifHeatmaps</span><span class="op">(</span><span class="va">se</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span>, which.plots <span class="op">=</span> <span class="st">"pearsonResid"</span>,</span></span>
<span class="r-in"><span>                  width <span class="op">=</span> <span class="fl">2</span>, show_seqlogo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotMotifHeatmaps-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

