<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>monaLisa - MOtif aNAlysis with Lisa • monaLisa</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="monaLisa - MOtif aNAlysis with Lisa">
<meta property="og:description" content="monaLisa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/monaLisa.html">Binned motif enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fmicompbio/monaLisa" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="monaLisa_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>monaLisa - MOtif aNAlysis with Lisa</h1>
                        <h4 data-toc-skip class="author">Dania Machlab, Lukas Burger, Charlotte Soneson, Michael Stadler</h4>
            
            <h4 data-toc-skip class="date">2021-10-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/master/vignettes/monaLisa.Rmd" class="external-link"><code>vignettes/monaLisa.Rmd</code></a></small>
      <div class="hidden name"><code>monaLisa.Rmd</code></div>

    </div>

    
    
<p><img src="monaLisa_logo_v1.png" style="display: block; margin-left: auto;
margin-right: auto; width: 50%; border: 0"></p>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p><em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> is a collection of functions for working with biological sequences and motifs that represent the binding preferences of transcription factors or nucleic acid binding proteins.</p>
<p>For example, <em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> can be used to conveniently find motif hits in sequences (see section @ref(findhits)), or to identify motifs that are likely associated with observed experimental data. Such analyses are supposed to provide potential answers to the question “Which transcription factors are the drivers of my observed changes in expression/methylation/accessibility?”.</p>
<p>Several other approaches have been described that also address this problem, among them REDUCE <span class="citation">(Roven and Bussemaker 2003)</span> and ISMARA <span class="citation">(Balwierz et al. 2014)</span>. In <em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em>, we aim to provide a flexible implementation that integrates well with other Bioconductor resources, makes use of the sequence composition correction developed for Homer <span class="citation">(Heinz et al. 2010)</span> or stability selection <span class="citation">(Meinshausen and Bühlmann 2010)</span> and provides several alternative ways to study the relationship between experimental measurements and sequence motifs.</p>
<p>You can use known motifs from collections of transcription factor binding specificities such as <em><a href="https://bioconductor.org/packages/3.14/JASPAR2020" class="external-link">JASPAR2020</a></em>, also available from Bioconductor. Genomic regions could be for example promoters, enhancers or accessible regions for which experimental data is available.</p>
<p>Two independent approaches are implemented to identify interesting motifs:</p>
<ul>
<li><p>In <strong>binned motif enrichment analysis</strong> (<code><a href="../reference/calcBinnedMotifEnrR.html">monaLisa::calcBinnedMotifEnrR</a></code>, see section @ref(binned)), genomic regions are grouped into bins according to a numerical value assigned to each region, such as the change in expression, accessibility or methylation. Motif enrichments are then calculated for each bin, normalizing for differences in sequence composition in a very similar way as originally done by <a href="http://homer.ucsd.edu/homer/index.html" class="external-link">Homer</a> <span class="citation">(Heinz et al. 2010)</span>. The binned motif enrichment approach was first introduced in <span class="citation">Ginno et al. (2018)</span> and subsequently applied in e.g. <span class="citation">Barisic et al. (2019)</span>. To see more details on how <code>calcBinnedMotifEnrR</code> resembles <code>Homer</code>, check the function help page. We recommend using this function to do the binned motif enrichment analysis, since it corrects for sequence composition differences similarly to <code>Homer</code>, but is implemented more efficiently. <code><a href="../reference/calcBinnedMotifEnrHomer.html">monaLisa::calcBinnedMotifEnrHomer</a></code> implements the same analysis using Homer and therefore requires a local <a href="http://homer.ucsd.edu/homer/introduction/install.html" class="external-link">installation of Homer</a>, and <code><a href="../reference/calcBinnedKmerEnr.html">monaLisa::calcBinnedKmerEnr</a></code>(see section @ref(binnedkmers)) implements the analysis for k-mers instead of motifs, to study sequence enrichments without the requirement of known motifs.</p></li>
<li><p><strong>Randomized Lasso stability selection</strong> (<code><a href="../reference/randLassoStabSel.html">monaLisa::randLassoStabSel</a></code>, see <a href="https://bioconductor.org/packages/3.14/monaLisa/vignettes/selecting_motifs_with_randLassoStabSel.html" class="external-link">the stability selection vignette</a>, <em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> uses a robust regression approach (stability selection, <span class="citation">Meinshausen and Bühlmann (2010)</span>) to predict what transcription factors can explain experimental measurements, for example changes in chromatin accessibility between two conditions. Also this approach allows to correct for sequence composition. In addition, similar motifs have to “compete” with each other to be selected.</p></li>
</ul>
<p>For both approaches, functions that allow visualization of obtained results are provided.</p>
<p>If you prefer to jump right in, you can continue with section @ref(quick) that shows a quick hypothetical example of how to run a binned motif enrichment analysis. If you prefer to actually compute enrichments on real data, you can find below a detailed example for a binned motif enrichment analysis (section @ref(binned)).</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p><em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> can be installed from Bioconductor via the <em><a href="https://CRAN.R-project.org/package=BiocManager" class="external-link">BiocManager</a></em> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"monaLisa"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="quick">Quick example: Identify enriched motifs in bins<a class="anchor" aria-label="anchor" href="#quick"></a>
</h1>
<p>The quick example below, which we do not run, illustrates how a binned motif enrichment analysis can be performed in <code>monaLisa</code>. We assume that you already have extracted a set of peaks. The sequences of the peak regions are stored in a <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">Biostrings::DNAStringSet</a></code> (<code>peak_seqs</code>), and additionally each peak is associated with a numeric value (e.g., the change of methylation between two conditions, stored in the <code>peak_change</code> vector), that will be used to bin the regions before finding motifs enriched in each bin.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/monaLisa" class="external-link">monaLisa</a></span><span class="op">)</span>

<span class="co"># bin regions</span>
<span class="co"># (peak_change is a numerical vector)</span>
<span class="co"># (peak_change needs to be created by the user to run this code)</span>
<span class="va">peak_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bin.html">bin</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">peak_change</span>, binmode <span class="op">=</span> <span class="st">"equalN"</span>, nElement <span class="op">=</span> <span class="fl">400</span><span class="op">)</span>

<span class="co"># calculate motif enrichments</span>
<span class="co"># (peak_seqs is a DNAStringSet, pwms is a PWMatrixList)</span>
<span class="co"># (peak_seqs and pwms need to be created by the user to run this code)</span>
<span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcBinnedMotifEnrR.html">calcBinnedMotifEnrR</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">peak_seqs</span>,
                          bins <span class="op">=</span> <span class="va">peak_bins</span>,
                          pwmL <span class="op">=</span> <span class="va">pwms</span><span class="op">)</span></code></pre></div>
<p>The returned <code>se</code> is a <code>SummarizedExperiment</code> with <code>assays</code> <em>negLog10P</em>, <em>negLog10Padj</em>, <em>expForegroundWgt</em>, <em>pearsonResid</em>, <em>log2enr</em>, <em>sumForegroundWgtWithHits</em> and <em>sumBackgroundWgtWithHits</em>, each containing a matrix with motifs (rows) by bins (columns). The values are:</p>
<ul>
<li>
<em>negLog10P</em>: the raw P value (<span class="math inline">\(-\log_{10} p\)</span>) of a given motif enrichment in a given bin. Each P value results from an enrichment calculation comparing occurrences of each motif in the bin to its occurrences in background sequences, defined by the <code>background</code> argument (by default: sequences in all other bins).<br>
</li>
<li>
<em>negLog10Padj</em>: Same as <em>negLog10P</em> but adjusted for multiple testing using the method of provided in the <code>p.adjust.method</code> argument, by default: Benjamini and Hochberg, 1995 (<code>p.adjust(..., method="fdr")</code>).<br>
</li>
<li>
<em>expForegroundWgt</em>: The expected number of regions in the bin containing a given motif.</li>
<li>
<em>pearsonResid</em>: Standardized Pearson residuals, a measure of motif enrichment akin to a z-score for the number of regions in the bin containing the motif. The standardized Pearson residuals are given by<br><span class="math inline">\(resid = (o - \mu)/\sigma\)</span>, where <span class="math inline">\(\mu\)</span> is the expected count and <span class="math inline">\(\sigma\)</span> the standard deviation of the expression in the numerator, under the null hypothesis that the probability of containing a motif is independent of whether the sequence is in the foreground or the background (see e.g.  <span class="citation">Agresti (2007)</span>, section 2.4.5).</li>
<li>
<em>log2enr</em>: Motif enrichments, calculated as: <span class="math inline">\(log2enr = log2((o + c)/(e + c))\)</span>, where <span class="math inline">\(o\)</span> and <span class="math inline">\(e\)</span> are the observed and expected numbers of regions in the bin containing a given motif, and <span class="math inline">\(c\)</span> is a pseudocount defined by the <code>pseudocount.log2enr</code> argument.<br>
</li>
<li>
<em>sumForegroundWgtWithHits</em> and <em>sumBackgroundWgtWithHits</em> are the sum of foreground and background sequences that have at least one occurrence of the motif, respectively. The background sequences are weighted in order to adjust for differences in sequence composition between foreground and background.</li>
</ul>
<p>In addition, <code>rowData(se)</code> and <code>colData(se)</code> give information about the used motifs and bins, respectively. In <code>metadata(se)</code> you can find information about parameter values.</p>
</div>
<div class="section level1">
<h1 id="binned">Binned motif enrichment analysis: Finding TFs enriched in differentially methylated regions<a class="anchor" aria-label="anchor" href="#binned"></a>
</h1>
<p>This example is based on an <em>in vitro</em> differentiation system, in which mouse embryonic stem (ES) cells are differentiated into neuronal progenitors (NP). In an earlier study <span class="citation">(Stadler et al. 2011)</span>, we have analyzed the genome-wide CpG methylation patterns in these cell types and identified so called <em>low methylated regions</em> (LMRs), that have reduced methylation levels and correspond to regions bound by transcription factors.</p>
<p>We also developed a tool that systematically identifies such regions from genome-wide methylation data <span class="citation">(Burger et al. 2013)</span>. Interestingly, a change in methylation of LMRs is indicative of altered transcription factor binding. We will therefore use these regions to identify transcription factor motifs that are enriched or depleted in LMRs that change their methylation between ES and NP cell states.</p>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<p>We start by loading the needed packages:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/monaLisa" class="external-link">monaLisa</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="genomic-regions-of-interest">Genomic regions of interest<a class="anchor" aria-label="anchor" href="#genomic-regions-of-interest"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> contains a file with genomic coordinates (mouse mm10 assembly) of LMRs, with the respective changes of methylation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lmrfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"LMRsESNPmerged.gr.rds"</span>, 
                       package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span>
<span class="va">lmr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">lmrfile</span><span class="op">)</span>
<span class="va">lmr</span>
<span class="co">#&gt; GRanges object with 45414 ranges and 1 metadata column:</span>
<span class="co">#&gt;           seqnames          ranges strand |    deltaMeth</span>
<span class="co">#&gt;              &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;numeric&gt;</span>
<span class="co">#&gt;       [1]     chr1 3549153-3550201      * |    0.3190299</span>
<span class="co">#&gt;       [2]     chr1 3680914-3682164      * |    0.0657352</span>
<span class="co">#&gt;       [3]     chr1 3913315-3914523      * |    0.4803313</span>
<span class="co">#&gt;       [4]     chr1 3953500-3954157      * |    0.4504727</span>
<span class="co">#&gt;       [5]     chr1 4150457-4151567      * |    0.5014768</span>
<span class="co">#&gt;       ...      ...             ...    ... .          ...</span>
<span class="co">#&gt;   [45410]     chrY 4196254-4196510      * | -0.020020382</span>
<span class="co">#&gt;   [45411]     chrY 4193654-4194152      * | -0.102559935</span>
<span class="co">#&gt;   [45412]     chrY 4190208-4192766      * | -0.031668206</span>
<span class="co">#&gt;   [45413]     chrY 4188072-4188924      * |  0.130623049</span>
<span class="co">#&gt;   [45414]     chrY 4181867-4182624      * |  0.000494588</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 21 sequences from an unspecified genome</span></code></pre></div>
<p>We can see there are 45414 LMRs, most of which gain methylation between ES and NP stages:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">lmr</span><span class="op">$</span><span class="va">deltaMeth</span>, <span class="fl">100</span>, col <span class="op">=</span> <span class="st">"gray"</span>, main <span class="op">=</span> <span class="st">""</span>,
     xlab <span class="op">=</span> <span class="st">"Change of methylation (NP - ES)"</span>, ylab <span class="op">=</span> <span class="st">"Number of LMRs"</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/deltameth-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>In order to keep the computation time reasonable, we’ll select 10,000 of the LMRs randomly:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">lmrsel</span> <span class="op">&lt;-</span> <span class="va">lmr</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lmr</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">10000</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">]</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="bin-genomic-regions">Bin genomic regions<a class="anchor" aria-label="anchor" href="#bin-genomic-regions"></a>
</h2>
<p>Now let’s bin our LMRs by how much they change methylation, using the <code>bin</code> function. We are not interested in small changes of methylation, say less than 0.3, so we’ll use the <code>minAbsX</code> argument to create a <em>no-change</em> bin in [-0.3, 0.3). The remaining LMRs are put into bins of 800 each:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bin.html">bin</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">lmrsel</span><span class="op">$</span><span class="va">deltaMeth</span>, binmode <span class="op">=</span> <span class="st">"equalN"</span>, nElement <span class="op">=</span> <span class="fl">800</span>, 
            minAbsX <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span>
<span class="co">#&gt; bins</span>
<span class="co">#&gt; [-0.935,-0.242]  (-0.242,0.327]   (0.327,0.388]   (0.388,0.443]   (0.443,0.491] </span>
<span class="co">#&gt;             800            4400             800             800             800 </span>
<span class="co">#&gt;   (0.491,0.536]   (0.536,0.585]   (0.585,0.862] </span>
<span class="co">#&gt;             800             800             800</span></code></pre></div>
<p>We can see which bin has been set to be the zero bin using <code><a href="../reference/getSetZeroBin.html">monaLisa::getZeroBin</a></code>, or set it to a different bin using <code><a href="../reference/getSetZeroBin.html">monaLisa::setZeroBin</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># find the index of the level representing the zero bin </span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span>
<span class="co">#&gt; [1] "[-0.935,-0.242]" "(-0.242,0.327]"  "(0.327,0.388]"   "(0.388,0.443]"  </span>
<span class="co">#&gt; [5] "(0.443,0.491]"   "(0.491,0.536]"   "(0.536,0.585]"   "(0.585,0.862]"</span>
<span class="fu"><a href="../reference/getSetZeroBin.html">getZeroBin</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span>
<span class="co">#&gt; [1] 2</span></code></pre></div>
<p>Because of the asymmetry of methylation changes, there is only a single bin with LMRs that lost methylation and many that gained:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBinDensity.html">plotBinDensity</a></span><span class="op">(</span><span class="va">lmrsel</span><span class="op">$</span><span class="va">deltaMeth</span>, <span class="va">bins</span>, legend <span class="op">=</span> <span class="st">"topleft"</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/plotbins-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>Note that the bin breaks around the <em>no-change</em> bin are not exactly -0.3 to 0.3. They have been adjusted to have the required 800 LMRs per bin.</p>
</div>
<div class="section level2">
<h2 id="prepare-motif-enrichment-analysis">Prepare motif enrichment analysis<a class="anchor" aria-label="anchor" href="#prepare-motif-enrichment-analysis"></a>
</h2>
<p>Next we prepare the motif enrichment analysis. We first need known motifs representing transcription factor binding site preferences. We extract all vertebrate motifs from the <em><a href="https://bioconductor.org/packages/3.14/JASPAR2020" class="external-link">JASPAR2020</a></em> package as positional weight matrices (PWMs):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span><span class="va">JASPAR2020</span>,
                     opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>matrixtype <span class="op">=</span> <span class="st">"PWM"</span>,
                                 tax_group <span class="op">=</span> <span class="st">"vertebrates"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Furthermore, we need the sequences corresponding to our LMRs. As sequences in one bin are compared to the sequences in other bins, we would not want differences of sequence lengths or composition between bins that might bias our motif enrichment results.</p>
<p>In general, we would recommend to use regions of similar or even equal lengths to avoid a length bias, for example by using a fixed-size region around the midpoint of each region of interest using <code><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html" class="external-link">GenomicRanges::resize</a></code>. In addition, the resized regions may have to be constrained to the chromosome boundaries using trim:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">lmrsel</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;     9.0   213.0   401.0   512.9   676.0  5973.0</span>
<span class="va">lmrsel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html" class="external-link">trim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span><span class="va">lmrsel</span>, width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">lmrsel</span><span class="op">)</span><span class="op">)</span>, fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">lmrsel</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;     401     401     401     401     401     401</span></code></pre></div>
<p>We can now directly extract the corresponding sequences from the <em><a href="https://bioconductor.org/packages/3.14/BSgenome.Mmusculus.UCSC.mm10" class="external-link">BSgenome.Mmusculus.UCSC.mm10</a></em> package</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lmrseqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/getSeq-methods.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span>, <span class="va">lmrsel</span><span class="op">)</span></code></pre></div>
<p>and check for differences in sequence composition between bins using the <code><a href="../reference/plotBinDiagnostics.html">monaLisa::plotBinDiagnostics</a></code> function. “GCfrac” will plot the distributions of the fraction of G+C bases, and “dinucfreq” creates a heatmap of average di-nucleotide frequencies in each bin, relative to the overall average.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBinDiagnostics.html">plotBinDiagnostics</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">lmrseqs</span>, bins <span class="op">=</span> <span class="va">bins</span>, aspect <span class="op">=</span> <span class="st">"GCfrac"</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/bindiag-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBinDiagnostics.html">plotBinDiagnostics</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">lmrseqs</span>, bins <span class="op">=</span> <span class="va">bins</span>, aspect <span class="op">=</span> <span class="st">"dinucfreq"</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/bindiag-2.png" width="80%" style="display: block; margin: auto;"></p>
<p>From these plot, we can see that LMRs with lower methylation in NP cells compared to ES cells (bin [-0.935,-0.242]) tend to be GC-poorer than LMRs in other bins. A strong bias of this kind can give rise to false positives in that bin, e.g. enrichments of AT-rich motifs. In this example, the bias is however not very strong and the built-in sequence composition correction in <code><a href="../reference/calcBinnedMotifEnrR.html">monaLisa::calcBinnedMotifEnrR</a></code> should be able to handle it.</p>
</div>
<div class="section level2">
<h2 id="run-motif-enrichment-analysis">Run motif enrichment analysis<a class="anchor" aria-label="anchor" href="#run-motif-enrichment-analysis"></a>
</h2>
<p>Finally, we run the binned motif enrichment analysis.</p>
<p>This step will take a while, and typically you would use the <code>BPPARAM</code> argument to run it with parallelization using <code>n</code> cores as follows: <code>calcBinnedMotifEnrR(..., BPPARAM = BiocParallel::MulticoreParam(n)))</code>. For this example however, you can just load the pre-computed results as shown below.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcBinnedMotifEnrR.html">calcBinnedMotifEnrR</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">lmrseqs</span>, bins <span class="op">=</span> <span class="va">bins</span>, pwmL <span class="op">=</span> <span class="va">pwms</span><span class="op">)</span></code></pre></div>
<p>In case you did not run the above code, let’s now read in the results:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"results.binned_motif_enrichment_LMRs.rds"</span>,
                          package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><code>se</code> is a <code>SummarizedExperiment</code> object which nicely keeps motifs, bins and corresponding metadata together:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summary</span>
<span class="va">se</span>
<span class="co">#&gt; class: SummarizedExperiment </span>
<span class="co">#&gt; dim: 746 8 </span>
<span class="co">#&gt; metadata(5): bins bins.binmode bins.breaks bins.bin0 param</span>
<span class="co">#&gt; assays(6): negLog10P negLog10Padj ... sumForegroundWgtWithHits</span>
<span class="co">#&gt;   sumBackgroundWgtWithHits</span>
<span class="co">#&gt; rownames(746): MA0004.1 MA0006.1 ... MA0528.2 MA0609.2</span>
<span class="co">#&gt; rowData names(5): motif.id motif.name motif.pfm motif.pwm</span>
<span class="co">#&gt;   motif.percentGC</span>
<span class="co">#&gt; colnames(8): [-0.935,-0.242] (-0.242,0.327] ... (0.536,0.585]</span>
<span class="co">#&gt;   (0.585,0.862]</span>
<span class="co">#&gt; colData names(6): bin.names bin.lower ... totalWgtForeground</span>
<span class="co">#&gt;   totalWgtBackground</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="co"># motifs-by-bins</span>
<span class="co">#&gt; [1] 746   8</span>

<span class="co"># motif info</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 746 rows and 5 columns</span>
<span class="co">#&gt;             motif.id   motif.name                       motif.pfm</span>
<span class="co">#&gt;          &lt;character&gt;  &lt;character&gt;                  &lt;PFMatrixList&gt;</span>
<span class="co">#&gt; MA0004.1    MA0004.1         Arnt         MA0004.1; Arnt; Unknown</span>
<span class="co">#&gt; MA0006.1    MA0006.1    Ahr::Arnt    MA0006.1; Ahr::Arnt; Unknown</span>
<span class="co">#&gt; MA0019.1    MA0019.1 Ddit3::Cebpa MA0019.1; Ddit3::Cebpa; Unknown</span>
<span class="co">#&gt; MA0029.1    MA0029.1        Mecom        MA0029.1; Mecom; Unknown</span>
<span class="co">#&gt; MA0030.1    MA0030.1        FOXF2        MA0030.1; FOXF2; Unknown</span>
<span class="co">#&gt; ...              ...          ...                             ...</span>
<span class="co">#&gt; MA0093.3    MA0093.3         USF1         MA0093.3; USF1; Unknown</span>
<span class="co">#&gt; MA0526.3    MA0526.3         USF2         MA0526.3; USF2; Unknown</span>
<span class="co">#&gt; MA0748.2    MA0748.2          YY2          MA0748.2; YY2; Unknown</span>
<span class="co">#&gt; MA0528.2    MA0528.2       ZNF263       MA0528.2; ZNF263; Unknown</span>
<span class="co">#&gt; MA0609.2    MA0609.2         CREM         MA0609.2; CREM; Unknown</span>
<span class="co">#&gt;                                                            motif.pwm</span>
<span class="co">#&gt;                                                       &lt;PWMatrixList&gt;</span>
<span class="co">#&gt; MA0004.1       MA0004.1; Arnt; Basic helix-loop-helix factors (bHLH)</span>
<span class="co">#&gt; MA0006.1  MA0006.1; Ahr::Arnt; Basic helix-loop-helix factors (bHLH)</span>
<span class="co">#&gt; MA0019.1 MA0019.1; Ddit3::Cebpa; Basic leucine zipper factors (bZIP)</span>
<span class="co">#&gt; MA0029.1                   MA0029.1; Mecom; C2H2 zinc finger factors</span>
<span class="co">#&gt; MA0030.1           MA0030.1; FOXF2; Fork head / winged helix factors</span>
<span class="co">#&gt; ...                                                              ...</span>
<span class="co">#&gt; MA0093.3       MA0093.3; USF1; Basic helix-loop-helix factors (bHLH)</span>
<span class="co">#&gt; MA0526.3       MA0526.3; USF2; Basic helix-loop-helix factors (bHLH)</span>
<span class="co">#&gt; MA0748.2                     MA0748.2; YY2; C2H2 zinc finger factors</span>
<span class="co">#&gt; MA0528.2                  MA0528.2; ZNF263; C2H2 zinc finger factors</span>
<span class="co">#&gt; MA0609.2         MA0609.2; CREM; Basic leucine zipper factors (bZIP)</span>
<span class="co">#&gt;          motif.percentGC</span>
<span class="co">#&gt;                &lt;numeric&gt;</span>
<span class="co">#&gt; MA0004.1         64.0893</span>
<span class="co">#&gt; MA0006.1         71.5266</span>
<span class="co">#&gt; MA0019.1         48.3898</span>
<span class="co">#&gt; MA0029.1         28.0907</span>
<span class="co">#&gt; MA0030.1         34.2125</span>
<span class="co">#&gt; ...                  ...</span>
<span class="co">#&gt; MA0093.3         51.0234</span>
<span class="co">#&gt; MA0526.3         51.4931</span>
<span class="co">#&gt; MA0748.2         67.2542</span>
<span class="co">#&gt; MA0528.2         67.4339</span>
<span class="co">#&gt; MA0609.2         53.2402</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "MA0004.1" "MA0006.1" "MA0019.1" "MA0029.1" "MA0030.1" "MA0031.1"</span>

<span class="co"># bin info</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 8 rows and 6 columns</span>
<span class="co">#&gt;                       bin.names bin.lower bin.upper bin.nochange</span>
<span class="co">#&gt;                     &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;logical&gt;</span>
<span class="co">#&gt; [-0.935,-0.242] [-0.935,-0.242] -0.935484 -0.242127        FALSE</span>
<span class="co">#&gt; (-0.242,0.327]   (-0.242,0.327] -0.242127  0.327369         TRUE</span>
<span class="co">#&gt; (0.327,0.388]     (0.327,0.388]  0.327369  0.387698        FALSE</span>
<span class="co">#&gt; (0.388,0.443]     (0.388,0.443]  0.387698  0.443079        FALSE</span>
<span class="co">#&gt; (0.443,0.491]     (0.443,0.491]  0.443079  0.490691        FALSE</span>
<span class="co">#&gt; (0.491,0.536]     (0.491,0.536]  0.490691  0.535714        FALSE</span>
<span class="co">#&gt; (0.536,0.585]     (0.536,0.585]  0.535714  0.584707        FALSE</span>
<span class="co">#&gt; (0.585,0.862]     (0.585,0.862]  0.584707  0.862443        FALSE</span>
<span class="co">#&gt;                 totalWgtForeground totalWgtBackground</span>
<span class="co">#&gt;                          &lt;numeric&gt;          &lt;numeric&gt;</span>
<span class="co">#&gt; [-0.935,-0.242]                800            8628.40</span>
<span class="co">#&gt; (-0.242,0.327]                4400            5576.92</span>
<span class="co">#&gt; (0.327,0.388]                  800            9186.26</span>
<span class="co">#&gt; (0.388,0.443]                  800            9186.58</span>
<span class="co">#&gt; (0.443,0.491]                  800            9195.14</span>
<span class="co">#&gt; (0.491,0.536]                  800            9157.61</span>
<span class="co">#&gt; (0.536,0.585]                  800            9163.05</span>
<span class="co">#&gt; (0.585,0.862]                  800            9137.44</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "[-0.935,-0.242]" "(-0.242,0.327]"  "(0.327,0.388]"   "(0.388,0.443]"  </span>
<span class="co">#&gt; [5] "(0.443,0.491]"   "(0.491,0.536]"</span>

<span class="co"># assays: the motif enrichment results</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="co">#&gt; [1] "negLog10P"                "negLog10Padj"            </span>
<span class="co">#&gt; [3] "pearsonResid"             "log2enr"                 </span>
<span class="co">#&gt; [5] "sumForegroundWgtWithHits" "sumBackgroundWgtWithHits"</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"log2enr"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">#&gt;          [-0.935,-0.242] (-0.242,0.327] (0.327,0.388]</span>
<span class="co">#&gt; MA0004.1      -0.4332719    -0.16418567   0.047435758</span>
<span class="co">#&gt; MA0006.1       0.2407477    -0.11995829  -0.005914484</span>
<span class="co">#&gt; MA0019.1      -0.6736372     0.26842621   0.030973190</span>
<span class="co">#&gt; MA0029.1      -0.1475501    -0.12750322   0.088480526</span>
<span class="co">#&gt; MA0030.1      -0.4021844     0.06710565   0.152049687</span></code></pre></div>
<p>We can plot the results using the <code><a href="../reference/plotMotifHeatmaps.html">monaLisa::plotMotifHeatmaps</a></code> function, e.g. selecting all transcription factor motifs that have an <span class="math inline">\(-log_{10} FDR\)</span> of at least 4.0 in any bin (corresponding to an <span class="math inline">\(FDR &lt; 10^{-4}\)</span>). FDR values are stored in the <code>negLog10Padj</code> assay:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># select strongly enriched motifs</span>
<span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span>, <span class="fl">1</span>, 
             <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">0</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">4.0</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sel</span><span class="op">)</span>
<span class="co">#&gt; [1] 59</span>
<span class="va">seSel</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="va">sel</span>, <span class="op">]</span>

<span class="co"># plot</span>
<span class="fu"><a href="../reference/plotMotifHeatmaps.html">plotMotifHeatmaps</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seSel</span>, which.plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log2enr"</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span>, 
                  width <span class="op">=</span> <span class="fl">2.0</span>, cluster <span class="op">=</span> <span class="cn">TRUE</span>, maxEnr <span class="op">=</span> <span class="fl">2</span>, maxSig <span class="op">=</span> <span class="fl">10</span>, 
                  show_motif_GC <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/plottfs-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>In order to select only motifs with significant enrichments in a specific bin, or in any bin except the “zero” bin, you could use:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># significantly enriched in bin 8</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span><span class="op">[</span><span class="fl">8</span><span class="op">]</span>
<span class="co">#&gt; [1] "(0.585,0.862]"</span>
<span class="va">sel.bin8</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">8</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">4.0</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sel.bin8</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; [1] 10</span>

<span class="co"># significantly enriched in any "non-zero" bin</span>
<span class="fu"><a href="../reference/getSetZeroBin.html">getZeroBin</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span>
<span class="co">#&gt; [1] 2</span>
<span class="va">sel.nonZero</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="../reference/getSetZeroBin.html">getZeroBin</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, <span class="fl">1</span>,
    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">0</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">4.0</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sel.nonZero</span><span class="op">)</span>
<span class="co">#&gt; [1] 55</span></code></pre></div>
<p>Setting <code>cluster = TRUE</code> in <code><a href="../reference/plotMotifHeatmaps.html">monaLisa::plotMotifHeatmaps</a></code> has re-ordered the rows using hierarchical clustering of the <code>"pearsonResid"</code> assay. As many transcription factor binding motifs are similar to each other, it is also helpful to show the enrichment heatmap clustered by motif similarity. To this end, we first calculate all pairwise motif similarities (measured as maximum Pearson correlation of all possible shifted alignments). This can be quickly calculated for the few selected motifs using the <code><a href="../reference/motifSimilarity.html">monaLisa::motifSimilarity</a></code> function. For many motifs, this step may take a while, and it may be useful to parallelize it using the <code>BPPARAM</code> argument (e.g. to run on <code>n</code> parallel threads using the multi-core backend, you can use: <code>motifSimilarity(..., BPPARAM = BiocParallel::MulticoreParam(n))</code>).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SimMatSel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/motifSimilarity.html">motifSimilarity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">seSel</span><span class="op">)</span><span class="op">$</span><span class="va">motif.pfm</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html" class="external-link">range</a></span><span class="op">(</span><span class="va">SimMatSel</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.05339967 1.00000000</span></code></pre></div>
<p>The order of the TFs in the resulting matrix is consistent with the elements of <code>seSel</code>, and the maximal similarity between any pair of motifs is 1.0. By subtracting these similarities from 1.0, we obtain distances that we use to perform a hierarchical clustering with the <code><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">stats::hclust</a></code> function. The returned object is then passed to the <code>cluster</code> argument to define the order of the rows in the heatmap. The plotting of the dendrogram is controlled by the argument <code>show_dendrogram</code>, and we also display the motifs as sequence logos using <code>show_seqlogo</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create hclust object, similarity defined by 1 - Pearson correlation</span>
<span class="va">hcl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">SimMatSel</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotMotifHeatmaps.html">plotMotifHeatmaps</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seSel</span>, which.plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log2enr"</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span>, 
                  width <span class="op">=</span> <span class="fl">1.8</span>, cluster <span class="op">=</span> <span class="va">hcl</span>, maxEnr <span class="op">=</span> <span class="fl">2</span>, maxSig <span class="op">=</span> <span class="fl">10</span>,
                  show_dendrogram <span class="op">=</span> <span class="cn">TRUE</span>, show_seqlogo <span class="op">=</span> <span class="cn">TRUE</span>,
                  width.seqlogo <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/plottfsclustered-1.png" width="80%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="convert-between-motif-text-file-for-homer-and-motif-objects-in-r">Convert between motif text file for <code>Homer</code> and motif objects in <code>R</code><a class="anchor" aria-label="anchor" href="#convert-between-motif-text-file-for-homer-and-motif-objects-in-r"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> provides two functions for performing binned motif enrichment analysis (<code><a href="../reference/calcBinnedMotifEnrR.html">monaLisa::calcBinnedMotifEnrR</a></code> and <code><a href="../reference/calcBinnedMotifEnrHomer.html">monaLisa::calcBinnedMotifEnrHomer</a></code>). <code><a href="../reference/calcBinnedMotifEnrR.html">monaLisa::calcBinnedMotifEnrR</a></code> implements the binned motif enrichment analysis in <code>R</code>, similarly to <code>Homer</code> and does not require the user to have the <code>Homer</code> tool pre-installed. For more information on that function and how it resembles the <code>Homer</code> tool see the function documentation.</p>
<p>A simple way to represent a DNA sequence motif that assumes independence of positions in the motif is a matrix with four rows (for the bases A, C, G and T) and <code>n</code> columns for the <code>n</code> positions in the motif. The values in that matrix can represent the sequence preferences of a binding protein in several different ways:</p>
<ul>
<li>
<strong>Position frequency matrices (PFM)</strong> contain values that correspond to the number of times (frequency) that a given base has been observed in at a given position of the motif. It is usually obtained from a set of known, aligned binding site sequences, and depending on the number of sequences, the values will be lower or higher. In <code>R</code>, PFMs are often represented using <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">TFBSTools::PFMatrix</a></code> (single motif) or <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">TFBSTools::PFMatrixList</a></code> objects (set of motifs). This is the rawest way to represent a sequence motif and can be converted into any other representation.<br>
</li>
<li>
<strong>Position probability matrices (PPM)</strong> are obtained by dividing the counts in each column of a PFM by their sum. The values now give a probability of observing a given base at that position of the motif and sum up to one in each column. This is the representation used in motif text files for <code>Homer</code>. A PPM can only be converted back to a PFM by knowing or assuming how many binding site sequences were observed (see argument <code>n</code> in <code>homerToPFMatrixList</code>).<br>
</li>
<li>
<strong>Position weight matrices (PWM)</strong> (also known as position specific scoring matrices, PSSM) are obtained by comparing the base probabilities in a PPM to the probabilities of observing each base outside of a binding site (background base probabilities), for example by calculating log-odds scores (see <code><a href="https://rdrr.io/pkg/TFBSTools/man/toPWM-methods.html" class="external-link">TFBSTools::toPWM</a></code> for details). This is a useful representation for scanning sequences for motif matches. In <code>R</code>, PWMs are often represented using <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">TFBSTools::PWMatrix</a></code> (single motif) or <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">TFBSTools::PWMatrixList</a></code> (set of motifs).</li>
</ul>
<p><code><a href="../reference/calcBinnedMotifEnrR.html">monaLisa::calcBinnedMotifEnrR</a></code> takes PWMs as a <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">TFBSTools::PWMatrixList</a></code> object to scan for motif hits. <code><a href="../reference/calcBinnedMotifEnrHomer.html">monaLisa::calcBinnedMotifEnrHomer</a></code> on the other hand takes a motif text file with PPMs, and requires the user to have <code>Homer</code> installed to use it for the binned motif enrichment analysis. Here, we show how one can get motif PFMs from <em><a href="https://bioconductor.org/packages/3.14/JASPAR2020" class="external-link">JASPAR2020</a></em> and convert them to a <code>Homer</code>-compatible text file with PPMs (<code><a href="../reference/dumpJaspar.html">monaLisa::dumpJaspar</a></code>) and vice versa (<code><a href="../reference/homerToPFMatrixList.html">monaLisa::homerToPFMatrixList</a></code> ), and how to convert a <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">TFBSTools::PFMatrixList</a></code> to a <code><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrixList-class.html" class="external-link">TFBSTools::PWMatrixList</a></code> for use with <code><a href="../reference/calcBinnedMotifEnrR.html">monaLisa::calcBinnedMotifEnrR</a></code> or <code><a href="../reference/findMotifHits-methods.html">monaLisa::findMotifHits</a></code>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get PFMs from JASPAR2020 package (vertebrate subset)</span>
<span class="va">pfms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span><span class="va">JASPAR2020</span>,
                     opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>matrixtype <span class="op">=</span> <span class="st">"PFM"</span>,
                                 tax_group <span class="op">=</span> <span class="st">"vertebrates"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># convert PFMs to PWMs</span>
<span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/toPWM-methods.html" class="external-link">toPWM</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span>

<span class="co"># convert JASPAR2020 PFMs (vertebrate subset) to Homer motif file</span>
<span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">convert</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dumpJaspar.html">dumpJaspar</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">tmp</span>,
                      pkg <span class="op">=</span> <span class="st">"JASPAR2020"</span>,
                      pseudocount <span class="op">=</span> <span class="fl">0</span>,
                      opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>tax_group <span class="op">=</span> <span class="st">"vertebrates"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># convert Homer motif file to PFMatrixList</span>
<span class="va">pfms_ret</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/homerToPFMatrixList.html">homerToPFMatrixList</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">tmp</span>, n <span class="op">=</span> <span class="fl">100L</span><span class="op">)</span>

<span class="co"># compare the first PFM</span>
<span class="co"># - notice the different magnitude of counts (controlled by `n`)</span>
<span class="co"># - notice that with the default (recommended) value of `pseudocount = 1.0`,</span>
<span class="co">#   there would be no zero values in pfms_ret matrices, making</span>
<span class="co">#   pfms and pfms_ret even more different</span>
<span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;   [,1] [,2] [,3] [,4] [,5] [,6]</span>
<span class="co">#&gt; A    4   19    0    0    0    0</span>
<span class="co">#&gt; C   16    0   20    0    0    0</span>
<span class="co">#&gt; G    0    1    0   20    0   20</span>
<span class="co">#&gt; T    0    0    0    0   20    0</span>
<span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pfms_ret</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;   [,1] [,2] [,3] [,4] [,5] [,6]</span>
<span class="co">#&gt; A   20   95    0    0    0    0</span>
<span class="co">#&gt; C   80    0  100    0    0    0</span>
<span class="co">#&gt; G    0    5    0  100    0  100</span>
<span class="co">#&gt; T    0    0    0    0  100    0</span>

<span class="co"># compare position probability matrices with the original PFM </span>
<span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, 
            STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;   [,1] [,2] [,3] [,4] [,5] [,6]</span>
<span class="co">#&gt; A  0.2 0.95    0    0    0    0</span>
<span class="co">#&gt; C  0.8 0.00    1    0    0    0</span>
<span class="co">#&gt; G  0.0 0.05    0    1    0    1</span>
<span class="co">#&gt; T  0.0 0.00    0    0    1    0</span>
<span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pfms_ret</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, 
            STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pfms_ret</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;   [,1] [,2] [,3] [,4] [,5] [,6]</span>
<span class="co">#&gt; A  0.2 0.95    0    0    0    0</span>
<span class="co">#&gt; C  0.8 0.00    1    0    0    0</span>
<span class="co">#&gt; G  0.0 0.05    0    1    0    1</span>
<span class="co">#&gt; T  0.0 0.00    0    0    1    0</span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="binnedkmers">Binned k-mer enrichment analysis<a class="anchor" aria-label="anchor" href="#binnedkmers"></a>
</h1>
<p>In some situations it may be beneficial to perform the enrichment analysis in a more ‘unbiased’ way, using k-mers rather than annotated motifs. Here, we will illustrate the process using the same LMR data set as used for the motif enrichment analysis above. Similarly to the motif enrichment, this step takes a while to perform, and we can also load the processed object directly.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sekm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcBinnedKmerEnr.html">calcBinnedKmerEnr</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">lmrseqs</span>, bins <span class="op">=</span> <span class="va">bins</span>, kmerLen <span class="op">=</span> <span class="fl">6</span>, 
                          includeRevComp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sekm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>, <span class="st">"results.binned_6mer_enrichment_LMRs.rds"</span>,
    package <span class="op">=</span> <span class="st">"monaLisa"</span>
<span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Just as for the motif enrichment analysis, the return value is a <code>SummarizedExperiment</code> object, with the same set of assays and annotations.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sekm</span>
<span class="co">#&gt; class: SummarizedExperiment </span>
<span class="co">#&gt; dim: 4096 8 </span>
<span class="co">#&gt; metadata(5): bins bins.binmode bins.breaks bins.bin0 param</span>
<span class="co">#&gt; assays(6): negLog10P negLog10Padj ... sumForegroundWgtWithHits</span>
<span class="co">#&gt;   sumBackgroundWgtWithHits</span>
<span class="co">#&gt; rownames(4096): AAAAAA AAAAAC ... TTTTTG TTTTTT</span>
<span class="co">#&gt; rowData names(5): motif.id motif.name motif.pfm motif.pwm</span>
<span class="co">#&gt;   motif.percentGC</span>
<span class="co">#&gt; colnames(8): [-0.935,-0.242] (-0.242,0.327] ... (0.536,0.585]</span>
<span class="co">#&gt;   (0.585,0.862]</span>
<span class="co">#&gt; colData names(6): bin.names bin.lower ... totalWgtForeground</span>
<span class="co">#&gt;   totalWgtBackground</span></code></pre></div>
<p>As for the motif enrichment, we can extract any k-mer that is enriched in any of the bins.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">selkm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sekm</span>, <span class="st">"negLog10Padj"</span><span class="op">)</span>, <span class="fl">1</span>, 
               <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">0</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">4</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">selkm</span><span class="op">)</span>
<span class="co">#&gt; [1] 85</span>
<span class="va">sekmSel</span> <span class="op">&lt;-</span> <span class="va">sekm</span><span class="op">[</span><span class="va">selkm</span>, <span class="op">]</span></code></pre></div>
<p>Next, let’s compare the enriched k-mers to the motifs that were found earlier. This can be done using the <code><a href="../reference/motifKmerSimilarity.html">monaLisa::motifKmerSimilarity</a></code> function. By showing the similarity between the enriched k-mers and motifs, we can see whether, e.g., strongly enriched k-mers do not seem to correspond to an annotated motif.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pfmSel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">seSel</span><span class="op">)</span><span class="op">$</span><span class="va">motif.pfm</span>
<span class="va">sims</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/motifKmerSimilarity.html">motifKmerSimilarity</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pfmSel</span>,
                            kmers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sekmSel</span><span class="op">)</span>,
                            includeRevComp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sims</span><span class="op">)</span>
<span class="co">#&gt; [1] 59 85</span>

<span class="va">maxwidth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="va">pfmSel</span><span class="op">)</span>, <span class="va">ncol</span><span class="op">)</span><span class="op">)</span>
<span class="va">seqlogoGrobs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pfmSel</span>, <span class="va">seqLogoGrob</span>, xmax <span class="op">=</span> <span class="va">maxwidth</span><span class="op">)</span>
<span class="va">hmSeqlogo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html" class="external-link">rowAnnotation</a></span><span class="op">(</span>logo <span class="op">=</span> <span class="fu"><a href="../reference/annoSeqlogo.html">annoSeqlogo</a></span><span class="op">(</span><span class="va">seqlogoGrobs</span>, which <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span>,
                           annotation_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="st">"inch"</span><span class="op">)</span>, 
                           show_annotation_name <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">sims</span>, 
        show_row_names <span class="op">=</span> <span class="cn">TRUE</span>, row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,
        show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, column_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,
        name <span class="op">=</span> <span class="st">"Similarity"</span>, column_title <span class="op">=</span> <span class="st">"Selected TFs and enriched k-mers"</span>,
        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>, 
        right_annotation <span class="op">=</span> <span class="va">hmSeqlogo</span><span class="op">)</span></code></pre></div>
<p><img src="monaLisa_files/figure-html/unnamed-chunk-6-1.png" width="80%" style="display: block; margin: auto;"></p>
</div>
<div class="section level1">
<h1 id="findhits">Use <em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> to annotate genomic regions with predicted motifs<a class="anchor" aria-label="anchor" href="#findhits"></a>
</h1>
<p>As mentioned <em><a href="https://bioconductor.org/packages/3.14/monaLisa" class="external-link">monaLisa</a></em> can also be used to scan sequences for motifs. Here is an example (just on a few sequences/motifs for illustration):</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get sequences of promoters as a DNAStringSet</span>
<span class="co"># (could also be a single DNAString, or the name of a fasta file)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Mmusculus.UCSC.mm10.knownGene</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: GenomicFeatures</span>
<span class="co">#&gt; Loading required package: AnnotationDbi</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html" class="external-link">trim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">promoters</a></span><span class="op">(</span><span class="va">TxDb.Mmusculus.UCSC.mm10.knownGene</span>,
                     upstream <span class="op">=</span> <span class="fl">1000</span>, downstream <span class="op">=</span> <span class="fl">500</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">10</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span><span class="op">)</span>
<span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/getSeq-methods.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span>, <span class="va">gr</span><span class="op">)</span>
<span class="va">seqs</span>
<span class="co">#&gt; DNAStringSet object of length 4:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]  1500 <span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">TTTT</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span>...<span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span> ENSMUST00000193812.1</span>
<span class="co">#&gt; [2]  1500 <span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span>...<span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span> ENSMUST00000161581.1</span>
<span class="co">#&gt; [3]  1500 <span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span>...<span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span> ENSMUST00000192183.1</span>
<span class="co">#&gt; [4]  1500 <span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CCCC</span><span style="color: #000000; background-color: #FFD7AF;">T</span>...<span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span> ENSMUST00000195361.1</span>

<span class="co"># get motifs as a PWMatrixList</span>
<span class="co"># (could also be a single PWMatrix, or the name of a motif file)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span>
<span class="va">pfms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixByID-methods.html" class="external-link">getMatrixByID</a></span><span class="op">(</span><span class="va">JASPAR2020</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MA0885.1"</span>, <span class="st">"MA0099.3"</span>, <span class="st">"MA0033.2"</span>, 
                                    <span class="st">"MA0037.3"</span>, <span class="st">"MA0158.1"</span><span class="op">)</span><span class="op">)</span>
<span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/toPWM-methods.html" class="external-link">toPWM</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span>
<span class="va">pwms</span>
<span class="co">#&gt; PWMatrixList of length 5</span>
<span class="co">#&gt; names(5): MA0885.1 MA0099.3 MA0033.2 MA0037.3 MA0158.1</span>
<span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">name</a></span><span class="op">(</span><span class="va">pwms</span><span class="op">)</span>
<span class="co">#&gt;   MA0885.1   MA0099.3   MA0033.2   MA0037.3   MA0158.1 </span>
<span class="co">#&gt;     "Dlx2" "FOS::JUN"    "FOXL1"    "GATA3"    "HOXA5"</span>

<span class="co"># predict hits in sequences</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findMotifHits-methods.html">findMotifHits</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">pwms</span>,
                     subject <span class="op">=</span> <span class="va">seqs</span>,
                     min.score <span class="op">=</span> <span class="fl">6.0</span>,
                     method <span class="op">=</span> <span class="st">"matchPWM"</span>,
                     BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span>
<span class="co">#&gt; GRanges object with 115 ranges and 4 metadata columns:</span>
<span class="co">#&gt;                     seqnames    ranges strand |     matchedSeq    pwmid pwmname</span>
<span class="co">#&gt;                        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;DNAStringSet&gt;    &lt;Rle&gt;   &lt;Rle&gt;</span>
<span class="co">#&gt;     [1] ENSMUST00000193812.1    93-100      + |       CTCTTATG MA0158.1   HOXA5</span>
<span class="co">#&gt;     [2] ENSMUST00000193812.1   103-110      + |       AGCTAATT MA0158.1   HOXA5</span>
<span class="co">#&gt;     [3] ENSMUST00000193812.1   252-259      + |       GTCATTAT MA0885.1    Dlx2</span>
<span class="co">#&gt;     [4] ENSMUST00000193812.1   355-362      + |       TGATAAAT MA0037.3   GATA3</span>
<span class="co">#&gt;     [5] ENSMUST00000193812.1   358-365      + |       TAAATTAT MA0885.1    Dlx2</span>
<span class="co">#&gt;     ...                  ...       ...    ... .            ...      ...     ...</span>
<span class="co">#&gt;   [111] ENSMUST00000195361.1   742-749      - |       ATGAAATT MA0158.1   HOXA5</span>
<span class="co">#&gt;   [112] ENSMUST00000195361.1   833-840      - |       ACAATTAT MA0885.1    Dlx2</span>
<span class="co">#&gt;   [113] ENSMUST00000195361.1   842-849      - |       GTAATTAA MA0885.1    Dlx2</span>
<span class="co">#&gt;   [114] ENSMUST00000195361.1   844-851      - |       AAGTAATT MA0158.1   HOXA5</span>
<span class="co">#&gt;   [115] ENSMUST00000195361.1   865-872      - |       ACCATTAT MA0885.1    Dlx2</span>
<span class="co">#&gt;             score</span>
<span class="co">#&gt;         &lt;numeric&gt;</span>
<span class="co">#&gt;     [1]   6.98342</span>
<span class="co">#&gt;     [2]   7.96626</span>
<span class="co">#&gt;     [3]   6.64334</span>
<span class="co">#&gt;     [4]   6.76273</span>
<span class="co">#&gt;     [5]   6.36851</span>
<span class="co">#&gt;     ...       ...</span>
<span class="co">#&gt;   [111]   6.61929</span>
<span class="co">#&gt;   [112]  10.61685</span>
<span class="co">#&gt;   [113]  10.97719</span>
<span class="co">#&gt;   [114]   7.96626</span>
<span class="co">#&gt;   [115]   6.28806</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 4 sequences from an unspecified genome</span>

<span class="co"># create hit matrix:</span>
<span class="co"># number of sites of each motif per sequence</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span><span class="op">)</span>,
           <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pwmname</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">name</a></span><span class="op">(</span><span class="va">pwms</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">m</span>
<span class="co">#&gt;                       </span>
<span class="co">#&gt;                        Dlx2 FOS::JUN FOXL1 GATA3 HOXA5</span>
<span class="co">#&gt;   ENSMUST00000193812.1    4        2    12     7    10</span>
<span class="co">#&gt;   ENSMUST00000161581.1   10        1     3     5    10</span>
<span class="co">#&gt;   ENSMUST00000192183.1    4        2     2     3    13</span>
<span class="co">#&gt;   ENSMUST00000195361.1   11        1     5     0    10</span></code></pre></div>
<p>The transformation of sequence and PWM names to factors with defined levels in the creation of the hit matrix above is not strictly needed, but it ensures that even sequences or motifs without any hits are reported in the matrix, and that the order of sequences (rows) and motifs (columns) is identical to the order in <code>seqs</code> and <code>pwms</code>.</p>
</div>
<div class="section level1">
<h1 id="session-info-and-logo">Session info and logo<a class="anchor" aria-label="anchor" href="#session-info-and-logo"></a>
</h1>
<p>The monaLisa logo uses a drawing that was obtained from <a href="http://vectorish.com/lisa-simpson.html" class="external-link uri">http://vectorish.com/lisa-simpson.html</a> under the Creative Commons attribution - non-commercial 3.0 license: <a href="https://creativecommons.org/licenses/by-nc/3.0/" class="external-link uri">https://creativecommons.org/licenses/by-nc/3.0/</a>.</p>
<p>This vignette was built using:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.1 (2021-08-10)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0</span>
<span class="co">#&gt;  [2] GenomicFeatures_1.45.2                   </span>
<span class="co">#&gt;  [3] AnnotationDbi_1.55.1                     </span>
<span class="co">#&gt;  [4] circlize_0.4.13                          </span>
<span class="co">#&gt;  [5] ComplexHeatmap_2.9.4                     </span>
<span class="co">#&gt;  [6] monaLisa_0.99.5                          </span>
<span class="co">#&gt;  [7] BSgenome.Mmusculus.UCSC.mm10_1.4.3       </span>
<span class="co">#&gt;  [8] BSgenome_1.61.0                          </span>
<span class="co">#&gt;  [9] rtracklayer_1.53.1                       </span>
<span class="co">#&gt; [10] Biostrings_2.61.2                        </span>
<span class="co">#&gt; [11] XVector_0.33.0                           </span>
<span class="co">#&gt; [12] TFBSTools_1.31.2                         </span>
<span class="co">#&gt; [13] JASPAR2020_0.99.10                       </span>
<span class="co">#&gt; [14] SummarizedExperiment_1.23.5              </span>
<span class="co">#&gt; [15] Biobase_2.53.0                           </span>
<span class="co">#&gt; [16] MatrixGenerics_1.5.4                     </span>
<span class="co">#&gt; [17] matrixStats_0.61.0                       </span>
<span class="co">#&gt; [18] GenomicRanges_1.45.0                     </span>
<span class="co">#&gt; [19] GenomeInfoDb_1.29.10                     </span>
<span class="co">#&gt; [20] IRanges_2.27.2                           </span>
<span class="co">#&gt; [21] S4Vectors_0.31.5                         </span>
<span class="co">#&gt; [22] BiocGenerics_0.39.2                      </span>
<span class="co">#&gt; [23] BiocStyle_2.21.4                         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] BiocFileCache_2.1.1         sm_2.2-5.7                 </span>
<span class="co">#&gt;   [3] systemfonts_1.0.3           plyr_1.8.6                 </span>
<span class="co">#&gt;   [5] splines_4.1.1               BiocParallel_1.27.17       </span>
<span class="co">#&gt;   [7] ggplot2_3.3.5               digest_0.6.28              </span>
<span class="co">#&gt;   [9] foreach_1.5.1               htmltools_0.5.2            </span>
<span class="co">#&gt;  [11] GO.db_3.14.0                fansi_0.5.0                </span>
<span class="co">#&gt;  [13] magrittr_2.0.1              memoise_2.0.0              </span>
<span class="co">#&gt;  [15] cluster_2.1.2               doParallel_1.0.16          </span>
<span class="co">#&gt;  [17] tzdb_0.1.2                  readr_2.0.2                </span>
<span class="co">#&gt;  [19] annotate_1.71.0             R.utils_2.11.0             </span>
<span class="co">#&gt;  [21] pkgdown_1.9000.9000.9000    prettyunits_1.1.1          </span>
<span class="co">#&gt;  [23] colorspace_2.0-2            rappdirs_0.3.3             </span>
<span class="co">#&gt;  [25] blob_1.2.2                  textshaping_0.3.6          </span>
<span class="co">#&gt;  [27] xfun_0.27                   dplyr_1.0.7                </span>
<span class="co">#&gt;  [29] crayon_1.4.1                RCurl_1.98-1.5             </span>
<span class="co">#&gt;  [31] jsonlite_1.7.2              TFMPvalue_0.0.8            </span>
<span class="co">#&gt;  [33] survival_3.2-13             zoo_1.8-9                  </span>
<span class="co">#&gt;  [35] iterators_1.0.13            glue_1.4.2                 </span>
<span class="co">#&gt;  [37] gtable_0.3.0                zlibbioc_1.39.0            </span>
<span class="co">#&gt;  [39] GetoptLong_1.0.5            DelayedArray_0.19.4        </span>
<span class="co">#&gt;  [41] shape_1.4.6                 scales_1.1.1               </span>
<span class="co">#&gt;  [43] DBI_1.1.1                   Rcpp_1.0.7                 </span>
<span class="co">#&gt;  [45] xtable_1.8-4                progress_1.2.2             </span>
<span class="co">#&gt;  [47] clue_0.3-59                 bit_4.0.4                  </span>
<span class="co">#&gt;  [49] stabs_0.6-4                 glmnet_4.1-2               </span>
<span class="co">#&gt;  [51] httr_1.4.2                  RColorBrewer_1.1-2         </span>
<span class="co">#&gt;  [53] ellipsis_0.3.2              pkgconfig_2.0.3            </span>
<span class="co">#&gt;  [55] XML_3.99-0.8                R.methodsS3_1.8.1          </span>
<span class="co">#&gt;  [57] sass_0.4.0                  dbplyr_2.1.1               </span>
<span class="co">#&gt;  [59] utf8_1.2.2                  tidyselect_1.1.1           </span>
<span class="co">#&gt;  [61] rlang_0.99.0.9000           reshape2_1.4.4             </span>
<span class="co">#&gt;  [63] munsell_0.5.0               tools_4.1.1                </span>
<span class="co">#&gt;  [65] cachem_1.0.6                cli_3.0.1                  </span>
<span class="co">#&gt;  [67] DirichletMultinomial_1.35.0 generics_0.1.0             </span>
<span class="co">#&gt;  [69] RSQLite_2.2.8               evaluate_0.14              </span>
<span class="co">#&gt;  [71] stringr_1.4.0               fastmap_1.1.0              </span>
<span class="co">#&gt;  [73] yaml_2.2.1                  ragg_1.1.3                 </span>
<span class="co">#&gt;  [75] knitr_1.36                  bit64_4.0.5                </span>
<span class="co">#&gt;  [77] fs_1.5.0                    caTools_1.18.2             </span>
<span class="co">#&gt;  [79] purrr_0.3.4                 KEGGREST_1.33.0            </span>
<span class="co">#&gt;  [81] R.oo_1.24.0                 poweRlaw_0.70.6            </span>
<span class="co">#&gt;  [83] xml2_1.3.2                  pracma_2.3.3               </span>
<span class="co">#&gt;  [85] biomaRt_2.49.4              compiler_4.1.1             </span>
<span class="co">#&gt;  [87] filelock_1.0.2              curl_4.3.2                 </span>
<span class="co">#&gt;  [89] png_0.1-7                   tibble_3.1.5               </span>
<span class="co">#&gt;  [91] bslib_0.3.1                 stringi_1.7.5              </span>
<span class="co">#&gt;  [93] highr_0.9                   desc_1.4.0                 </span>
<span class="co">#&gt;  [95] lattice_0.20-45             CNEr_1.29.0                </span>
<span class="co">#&gt;  [97] Matrix_1.3-4                vctrs_0.3.8                </span>
<span class="co">#&gt;  [99] pillar_1.6.4                lifecycle_1.0.1            </span>
<span class="co">#&gt; [101] BiocManager_1.30.16         jquerylib_0.1.4            </span>
<span class="co">#&gt; [103] GlobalOptions_0.1.2         bitops_1.0-7               </span>
<span class="co">#&gt; [105] R6_2.5.1                    BiocIO_1.3.0               </span>
<span class="co">#&gt; [107] bookdown_0.24               codetools_0.2-18           </span>
<span class="co">#&gt; [109] gtools_3.9.2                assertthat_0.2.1           </span>
<span class="co">#&gt; [111] seqLogo_1.59.0              rprojroot_2.0.2            </span>
<span class="co">#&gt; [113] rjson_0.2.20                GenomicAlignments_1.29.0   </span>
<span class="co">#&gt; [115] Rsamtools_2.9.1             GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt; [117] parallel_4.1.1              hms_1.1.1                  </span>
<span class="co">#&gt; [119] rmarkdown_2.11              vioplot_0.3.7              </span>
<span class="co">#&gt; [121] restfulr_0.0.13</span></code></pre></div>
</div>
<div class="section level1 unnumbered">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references">
<div id="ref-agresti2007">
<p>Agresti, A. 2007. <em>An Introduction to Categorical Data Analysis</em>. Wiley-Blackwell.</p>
</div>
<div id="ref-ismara">
<p>Balwierz, P. J., M. Pachkov, P. Arnold, A. J. Gruber, M. Zavolan, and E. van Nimwegen. 2014. “ISMARA: automated modeling of genomic signals as a democracy of regulatory motifs.” <em>Genome Res</em> 24 (5): 869–84.</p>
</div>
<div id="ref-barisic2019">
<p>Barisic, D., M. B. Stadler, M. Iurlaro, and D. Schübeler. 2019. “Mammalian ISWI and SWI/SNF selectively mediate binding of distinct transcription factors.” <em>Nature</em> 569 (7754): 136–40.</p>
</div>
<div id="ref-MethylSeekR">
<p>Burger, Lukas, Dimos Gaidatzis, Dirk Schuebeler, and Michael B. Stadler. 2013. “Identification of Active Regulatory Regions from DNA Methylation Data.” <em>Nucleic Acids Research</em> 41: e155. <a href="https://doi.org/doi:10.1093/nar/gkt599" class="external-link">https://doi.org/doi:10.1093/nar/gkt599</a>.</p>
</div>
<div id="ref-ginno2018">
<p>Ginno, P. A., L. Burger, J. Seebacher, V. Iesmantavicius, and D. Schübeler. 2018. “Cell cycle-resolved chromatin proteomics reveals the extent of mitotic preservation of the genomic regulatory landscape.” <em>Nat Commun</em> 9 (1): 4048.</p>
</div>
<div id="ref-homer">
<p>Heinz, S., C. Benner, N. Spann, E. Bertolino, Y. C. Lin, P. Laslo, J. X. Cheng, C. Murre, H. Singh, and C. K. Glass. 2010. “Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities.” <em>Mol Cell</em> 38 (4): 576–89.</p>
</div>
<div id="ref-StabSel">
<p>Meinshausen, Nicolai, and Peter Bühlmann. 2010. “Stability Selection.” <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 72 (4): 417–73. <a href="https://doi.org/doi:10.1111/j.1467-9868.2010.00740.x" class="external-link">https://doi.org/doi:10.1111/j.1467-9868.2010.00740.x</a>.</p>
</div>
<div id="ref-reduce">
<p>Roven, C., and H. J. Bussemaker. 2003. “REDUCE: An online tool for inferring cis-regulatory elements and transcriptional module activities from microarray data.” <em>Nucleic Acids Res</em> 31 (13): 3487–90.</p>
</div>
<div id="ref-LMRs">
<p>Stadler, Michael B., Rabih Murr, Lukas Burger, Robert Ivanek, Florian Lienert, Anne Schöler, Erik van Nimwegen, et al. 2011. “DNA-Binding Factors Shape the Mouse Methylome at Distal Regulatory Regions.” <em>Nature</em> 480: 490–95. <a href="https://doi.org/doi:10.1038/nature10716" class="external-link">https://doi.org/doi:10.1038/nature10716</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
