<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Regression Based Approach for Motif Selection • monaLisa</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Regression Based Approach for Motif Selection">
<meta property="og:description" content="monaLisa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/monaLisa.html">Binned motif enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Identifying motifs by stability selection</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fmicompbio/monaLisa" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="selecting_motifs_with_randLassoStabSel_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Regression Based Approach for Motif Selection</h1>
                        <h4 data-toc-skip class="author">Dania Machlab, Lukas Burger, Charlotte Soneson, Michael Stadler</h4>
            
            <h4 data-toc-skip class="date">2021-11-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/HEAD/vignettes/selecting_motifs_with_randLassoStabSel.Rmd" class="external-link"><code>vignettes/selecting_motifs_with_randLassoStabSel.Rmd</code></a></small>
      <div class="hidden name"><code>selecting_motifs_with_randLassoStabSel.Rmd</code></div>

    </div>

    
    
<p><img src="monaLisa_logo_v1.png" style="display: block; margin-left: auto;
margin-right: auto; width: 50%; border: 0"></p>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>Identifying important transcription factor (TF) motifs, as shown in <a href="https://bioconductor.org/packages/3.15/monaLisa/vignettes/monaLisa.html" class="external-link">the main vignette</a>, could also be done using a regression based approach, where motifs are selected and have to compete against each other for selection. In this framework, the response vector can be the observed experimental measure of interest, e.g. log-fold changes of accessibility for a set of regions, and the predictors consist of the TF motif hits across those regions. In <em><a href="https://bioconductor.org/packages/3.15/monaLisa" class="external-link">monaLisa</a></em>, we implement the randomized lasso stability selection proposed by <span class="citation">Meinshausen and Bühlmann (2010)</span> with the improved error bounds introduced by <span class="citation">Shah and Samworth (2013)</span>. We have modified the <code><a href="https://rdrr.io/pkg/stabs/man/fitfuns.html" class="external-link">stabs::glmnet.lasso</a></code> function used by <code><a href="https://rdrr.io/pkg/stabs/man/stabsel.html" class="external-link">stabs::stabsel</a></code> from the <em><a href="https://CRAN.R-project.org/package=stabs" class="external-link">stabs</a></em> package to implement the randomized lasso.</p>
<p>Lasso stability selection implements the lasso regression multiple times on subsamples of the data, and returns a selection probability for each predictor (number of times selected divided by number of regressions done). With the randomized lasso, a weakness parameter is additionally used to vary the lasso penalty term <span class="math inline">\(\lambda\)</span> to a randomly chosen value between [<span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\lambda\)</span>/weakness] for each predictor. Although the main appeal of stability selection is in cases where the number of predictors exceeds the number of observations, it also performs better variable selection with noisy data <span class="citation">(Meinshausen and Bühlmann 2010)</span>.</p>
<p>With this approach, TF motifs compete against each other to explain the response vector, and we can also include additional predictors like GC content to compete against the TF motifs for selection. This is especially useful if the response is biased by sequence composition, for example if regions with higher GC content tend to have higher response values.</p>
</div>
<div class="section level1">
<h1 id="stabsel">Motif selection with Randomized Lasso Stability Selection<a class="anchor" aria-label="anchor" href="#stabsel"></a>
</h1>
<p>In the example below, we select for TF motifs explaining log-fold changes in chromatin accessibility (ATAC-seq) across the enhancers between mouse liver and lung tissue at P0, but this can be applied to other data types as well (ChIP-seq, RNA-seq, methylation etc.). Positive log2-fold changes indicate more accessibility in the liver tissue, whereas negative values indicate more accessibility in the lung tissue.</p>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<p>We start by loading the needed packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/monaLisa" class="external-link">monaLisa</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-dataset">Load dataset<a class="anchor" aria-label="anchor" href="#load-dataset"></a>
</h2>
<p>In our example dataset, we have quantified ATAC-seq reads on enhancers in mouse P0 lung and liver tissues. The log2-fold change (our response vector in this example) is for liver vs lung chromatin accessibility. We are using a set of 10,000 randomly sampled enhancers to illustrate how randomized lasso stability selection can be used to select TFs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load GRanges object with logFC and peaks</span>
<span class="va">gr_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"atac_liver_vs_lung.rds"</span>, 
                       package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">gr_path</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-tfbs-per-motif-and-peak">Get TFBS per motif and peak<a class="anchor" aria-label="anchor" href="#get-tfbs-per-motif-and-peak"></a>
</h2>
<p>We will now construct the transcription factor binding site (TFBS) matrix for known motifs (from a database like JASPAR) in peak regions. We use the <code><a href="../reference/findMotifHits-methods.html">monaLisa::findMotifHits</a></code> function to scan for TF motif hits. This matrix will be the predictor matrix in our regression. This step may take a while, and it may be useful to parallelize it using the <code>BPPARAM</code> argument (e.g. to run on <code>n</code> parallel threads using the multi-core backend, you can use: <code>findMotifHits(..., BPPARAM = BiocParallel::MulticoreParam(n))</code>).</p>
<p>As mentioned, this framework offers the flexibility to add additional predictors to compete against the TF motifs for selection. Here, we add the fraction of G+C and CpG observed/expected ratio as predictors, to ensure that selected TF motifs are not just detecting a simple trend in GC or CpG composition.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get PFMs (vertebrate TFs from Jaspar)</span>
<span class="va">pfms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span><span class="va">JASPAR2020</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>matrixtype <span class="op">=</span> <span class="st">"PFM"</span>, 
                                      tax_group <span class="op">=</span> <span class="st">"vertebrates"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># randomly sample 300 PFMs for illustration purposes (for quick runtime)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">4563</span><span class="op">)</span>
<span class="va">pfms</span> <span class="op">&lt;-</span> <span class="va">pfms</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">300</span><span class="op">)</span><span class="op">]</span>

<span class="co"># convert PFMs to PWMs</span>
<span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/toPWM-methods.html" class="external-link">toPWM</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span>

<span class="co"># get TFBS on given GRanges (peaks)</span>
<span class="co"># suppress warnings generated by matchPWM due to the presence of Ns </span>
<span class="co"># in the sequences</span>
<span class="va">peakSeq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span>, <span class="va">gr</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span>
  <span class="va">hits</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findMotifHits-methods.html">findMotifHits</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">pwms</span>, subject <span class="op">=</span> <span class="va">peakSeq</span>, min.score <span class="op">=</span> <span class="fl">10.0</span>,
                        BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>

<span class="co"># get TFBS matrix</span>
<span class="va">TFBSmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">unclass</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">hits</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">hits</span><span class="op">)</span><span class="op">)</span>,
                            <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">hits</span><span class="op">$</span><span class="va">pwmname</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">name</a></span><span class="op">(</span><span class="va">pwms</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">TFBSmatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="co">#&gt;              </span>
<span class="co">#&gt;               NR3C2 Arnt LHX1 SNAI1 MAFG ZSCAN4</span>
<span class="co">#&gt;   peak_51663      0    0    0     0    0      0</span>
<span class="co">#&gt;   peak_57870      0    0    0     0    0      0</span>
<span class="co">#&gt;   peak_2986       2    0    0     0    0      0</span>
<span class="co">#&gt;   peak_124022     0    0    0     0    0      0</span>
<span class="co">#&gt;   peak_29925      0    2    0     1    0      0</span>
<span class="co">#&gt;   peak_95246      0    0    0     0    0      0</span>

<span class="co"># remove TF motifs with 0 binding sites in all regions</span>
<span class="va">zero_TF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">TFBSmatrix</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">zero_TF</span><span class="op">)</span>
<span class="co">#&gt; [1] 2</span>
<span class="va">TFBSmatrix</span> <span class="op">&lt;-</span> <span class="va">TFBSmatrix</span><span class="op">[</span>, <span class="op">!</span><span class="va">zero_TF</span><span class="op">]</span>

<span class="co"># calculate G+C and CpG obs/expected</span>
<span class="va">fMono</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/nucleotideFrequency.html" class="external-link">oligonucleotideFrequency</a></span><span class="op">(</span><span class="va">peakSeq</span>, width <span class="op">=</span> <span class="fl">1L</span>, as.prob <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fDi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/nucleotideFrequency.html" class="external-link">oligonucleotideFrequency</a></span><span class="op">(</span><span class="va">peakSeq</span>, width <span class="op">=</span> <span class="fl">2L</span>, as.prob <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fracGC</span> <span class="op">&lt;-</span> <span class="va">fMono</span><span class="op">[</span>, <span class="st">"G"</span><span class="op">]</span> <span class="op">+</span> <span class="va">fMono</span><span class="op">[</span>, <span class="st">"C"</span><span class="op">]</span>
<span class="va">oeCpG</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">fDi</span><span class="op">[</span>, <span class="st">"CG"</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">fMono</span><span class="op">[</span>, <span class="st">"G"</span><span class="op">]</span> <span class="op">*</span> <span class="va">fMono</span><span class="op">[</span>, <span class="st">"C"</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.01</span><span class="op">)</span>

<span class="co"># add GC and oeCpG to predictor matrix</span>
<span class="va">TFBSmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">fracGC</span>, <span class="va">oeCpG</span>, <span class="va">TFBSmatrix</span><span class="op">)</span>
<span class="va">TFBSmatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="co">#&gt;                fracGC     oeCpG NR3C2 Arnt LHX1 SNAI1</span>
<span class="co">#&gt; peak_51663  0.5155709 0.4079115     0    0    0     0</span>
<span class="co">#&gt; peak_57870  0.4963235 0.3048298     0    0    0     0</span>
<span class="co">#&gt; peak_2986   0.4008264 0.3103806     2    0    0     0</span>
<span class="co">#&gt; peak_124022 0.4572650 0.4429813     0    0    0     0</span>
<span class="co">#&gt; peak_29925  0.4675000 0.3495939     0    2    0     1</span>
<span class="co">#&gt; peak_95246  0.5144509 0.4020976     0    0    0     0</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identify-important-tfs">Identify important TFs<a class="anchor" aria-label="anchor" href="#identify-important-tfs"></a>
</h2>
<p>We can now run randomized lasso stability selection to identify TFs that are likely to explain the log-fold changes in accessibility.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># # randLassoStabSel() is stochastic, so we set a seed to reproduce </span>
<span class="co"># # ... a parallel run</span>
<span class="co"># RNGkind("L'Ecuyer-CMRG")</span>
<span class="co"># set.seed(123)</span>
<span class="co"># se &lt;- randLassoStabSel(x = TFBSmatrix, y = gr$logFC_liver_vs_lung, </span>
<span class="co">#                        cutoff = 0.8, mc.preschedule = TRUE, </span>
<span class="co">#                        mc.set.seed = TRUE, mc.cores = 2L)</span>

<span class="co"># if not running in parallel mode, it is enough to use set.seed() before </span>
<span class="co"># ... using the function to ensure reproducibility (with 1 core)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/randLassoStabSel.html">randLassoStabSel</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TFBSmatrix</span>, y <span class="op">=</span> <span class="va">gr</span><span class="op">$</span><span class="va">logFC_liver_vs_lung</span>, 
                       cutoff <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span>
<span class="va">se</span>
<span class="co">#&gt; class: SummarizedExperiment </span>
<span class="co">#&gt; dim: 10000 300 </span>
<span class="co">#&gt; metadata(12): stabsel.params.cutoff stabsel.params.selected ...</span>
<span class="co">#&gt;   stabsel.params.call randStabsel.params.weakness</span>
<span class="co">#&gt; assays(1): x</span>
<span class="co">#&gt; rownames(10000): peak_51663 peak_57870 ... peak_98880 peak_67984</span>
<span class="co">#&gt; rowData names(1): y</span>
<span class="co">#&gt; colnames(300): fracGC oeCpG ... CLOCK OLIG2</span>
<span class="co">#&gt; colData names(20): selProb selected ... regStep16 regStep17</span>

<span class="co"># selected TFs</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="va">se</span><span class="op">$</span><span class="va">selected</span><span class="op">]</span>
<span class="co">#&gt;  [1] "NKX2-5"       "GATA1::TAL1"  "HNF1B"        "HNF4A(var.2)" "Nr2f6"       </span>
<span class="co">#&gt;  [6] "ONECUT3"      "MYF5"         "THRB"         "ISL2"         "NR2C2"       </span>
<span class="co">#&gt; [11] "TEAD3"        "TEAD4"        "GATA3"        "RORA"         "NFIC"        </span>
<span class="co">#&gt; [16] "ZEB1"</span></code></pre></div>
<p>The stability paths visualize how predictors get selected, decreasing regularization stringency (from left to right):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotStabilityPaths.html">plotStabilityPaths</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/plotStabilityPaths-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>Each line corresponds to a predictor, and we can see the selection probabilities as a function of the regularization steps, corresponding to decreasing values for the lambda regularization parameter in lasso. The predictor (TF motif) selection happens at the last step, given the specified minimum probability.</p>
<p>We can also visualize the selection probabilities of the selected TF motifs, optionally multiplied by the sign of the correlation to the response vector, to know how the TF relates to the change of accessibility (<code>directional</code> parameter). Note that although one can vary the <code>selProbMinPlot</code> argument which sets the selection probability cutoff, it is best to re-run randomized lasso stability selection with the new cutoff, as this influences other parameter choices the model uses internally. See <span class="citation">Meinshausen and Bühlmann (2010)</span> for more details.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSelectionProb.html">plotSelectionProb</a></span><span class="op">(</span><span class="va">se</span>, directional <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/plotSelProbs-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>Next, we visualize the correlation structure of the TFBS matrix and selected TF motifs. While the colinearity of predictors is generally an issue in regression-based approaches where variables are selected, randomized lasso stability selection generally does better at co-selecting correlated predictors. In practice, we still see it select predictors with correlations as high as 0.9. However, it is good to keep in mind that this can be an issue, and that predictors that are extremely correlated with each other might not end up being co-selected. In a bad case example, if we have two identical predictors, neither of them could end up being selected if half of the times the regressions are done, one would be randomly selected, and the other one in the other half, giving each a selection probability of 0.5. If many cases of high correlations exist or this is a concern, one may consider selecting a representative set of predictors to use.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># subset the selected TFs</span>
<span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="va">se</span><span class="op">$</span><span class="va">selected</span><span class="op">]</span>
<span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="va">sel</span><span class="op">]</span>

<span class="co"># exclude oeCpG and fracGC</span>
<span class="va">excl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"oeCpG"</span>, <span class="st">"fracGC"</span><span class="op">)</span>
<span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se_sub</span><span class="op">[</span>, <span class="op">!</span><span class="va">excl</span><span class="op">]</span>

<span class="co"># correlation matrix </span>
<span class="va">TFBSmatrixCorSel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">TFBSmatrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span>

<span class="co"># heatmap</span>
<span class="va">pfmsSel</span> <span class="op">&lt;-</span> <span class="va">pfms</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">TFBSmatrixCorSel</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">name</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="va">maxwidth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="va">pfmsSel</span><span class="op">)</span>, <span class="va">ncol</span><span class="op">)</span><span class="op">)</span>
<span class="va">seqlogoGrobs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pfmsSel</span>, <span class="va">seqLogoGrob</span>, xmax <span class="op">=</span> <span class="va">maxwidth</span><span class="op">)</span>

<span class="va">hmSeqlogo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html" class="external-link">rowAnnotation</a></span><span class="op">(</span>logo <span class="op">=</span> <span class="fu"><a href="../reference/annoSeqlogo.html">annoSeqlogo</a></span><span class="op">(</span><span class="va">seqlogoGrobs</span>, which <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span>,
                           annotation_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"inch"</span><span class="op">)</span>, 
                           show_annotation_name <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="va">colAnn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>AUC <span class="op">=</span> <span class="va">se_sub</span><span class="op">$</span><span class="va">selAUC</span>, selProb <span class="op">=</span> <span class="va">se_sub</span><span class="op">$</span><span class="va">selProb</span>,
                            show_legend <span class="op">=</span> <span class="cn">TRUE</span>, 
                            show_annotation_name <span class="op">=</span> <span class="cn">TRUE</span>,
                            col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
                              AUC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, 
                                               <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"brown"</span><span class="op">)</span><span class="op">)</span>,
                              selProb <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, 
                                                   <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"steelblue"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">TFBSmatrixCorSel</span>, 
        show_row_names <span class="op">=</span> <span class="cn">TRUE</span>, 
        show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, 
        name <span class="op">=</span> <span class="st">"Pear. Cor."</span>, column_title <span class="op">=</span> <span class="st">"Selected TFs"</span>,
        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>, 
        right_annotation <span class="op">=</span> <span class="va">hmSeqlogo</span>,
        top_annotation <span class="op">=</span> <span class="va">colAnn</span><span class="op">)</span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/TFBScor_selected-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># exclude oeCpG and fracGC</span>
<span class="va">excl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"oeCpG"</span>, <span class="st">"fracGC"</span><span class="op">)</span>
<span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="op">!</span><span class="va">excl</span><span class="op">]</span>

<span class="co"># correlation matrix </span>
<span class="va">TFBSmatrixCorSel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">TFBSmatrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span>

<span class="co"># heatmap annotations</span>
<span class="va">ann</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>
    selected <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span>, 
    AUC <span class="op">=</span> <span class="va">se_sub</span><span class="op">$</span><span class="va">selAUC</span>, 
    selProb <span class="op">=</span> <span class="va">se_sub</span><span class="op">$</span><span class="va">selProb</span>, 
    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>selected <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span><span class="op">=</span><span class="st">"forestgreen"</span>, 
                            <span class="st">"FALSE"</span><span class="op">=</span><span class="st">"grey"</span><span class="op">)</span>, 
               AUC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, 
                                <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"brown"</span><span class="op">)</span><span class="op">)</span>, 
               selProb <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, 
                                    <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"steelblue"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># heatmap</span>
<span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">TFBSmatrixCorSel</span>, show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, 
        name <span class="op">=</span> <span class="st">"Pear. Cor."</span>, column_title <span class="op">=</span> <span class="st">"All TFs"</span>,
        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>, 
        top_annotation <span class="op">=</span> <span class="va">ann</span><span class="op">)</span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/TFBScor_all-1.png" width="80%" style="display: block; margin: auto;"></p>
<p> </p>
<p>We can examine the peaks that have hits for a selected TF motif of interest, ordered by absolute accessibility changes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TF</span> <span class="op">&lt;-</span> <span class="va">sel</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>
<span class="va">TF</span>
<span class="co">#&gt; [1] "GATA1::TAL1"</span>

<span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"x"</span><span class="op">)</span><span class="op">[</span>, <span class="va">TF</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># peaks that contain TF hits...</span>
<span class="va">nm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">gr</span><span class="op">$</span><span class="va">logFC_liver_vs_lung</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, 
                 decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="co"># ... order by |logFC|</span>

<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gr</span><span class="op">[</span><span class="va">nm</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 6 ranges and 1 metadata column:</span>
<span class="co">#&gt;               seqnames            ranges strand | logFC_liver_vs_lung</span>
<span class="co">#&gt;                  &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;numeric&gt;</span>
<span class="co">#&gt;    peak_59642     chr8 80499941-80501128      * |             5.31505</span>
<span class="co">#&gt;    peak_58095     chr8 40658907-40659590      * |             5.13536</span>
<span class="co">#&gt;   peak_100144    chr14 69318236-69318909      * |             5.09810</span>
<span class="co">#&gt;    peak_98635    chr14 41124351-41126121      * |             5.07972</span>
<span class="co">#&gt;    peak_42826     chr6 29700952-29701570      * |             5.02324</span>
<span class="co">#&gt;    peak_44599     chr6 67186496-67187415      * |             4.97407</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 19 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="session-info-and-logo">Session info and logo<a class="anchor" aria-label="anchor" href="#session-info-and-logo"></a>
</h1>
<p>The monaLisa logo uses a drawing that was obtained from <a href="http://vectorish.com/lisa-simpson.html" class="external-link uri">http://vectorish.com/lisa-simpson.html</a> under the Creative Commons attribution - non-commercial 3.0 license: <a href="https://creativecommons.org/licenses/by-nc/3.0/" class="external-link uri">https://creativecommons.org/licenses/by-nc/3.0/</a>.</p>
<p>This vignette was built using:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2021-11-11 r81177)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] circlize_0.4.13                    ComplexHeatmap_2.11.0             </span>
<span class="co">#&gt;  [3] SummarizedExperiment_1.25.2        Biobase_2.55.0                    </span>
<span class="co">#&gt;  [5] MatrixGenerics_1.7.0               matrixStats_0.61.0                </span>
<span class="co">#&gt;  [7] BSgenome.Mmusculus.UCSC.mm10_1.4.3 BSgenome_1.63.2                   </span>
<span class="co">#&gt;  [9] rtracklayer_1.55.0                 Biostrings_2.63.0                 </span>
<span class="co">#&gt; [11] XVector_0.35.0                     GenomicRanges_1.47.3              </span>
<span class="co">#&gt; [13] GenomeInfoDb_1.31.1                IRanges_2.29.0                    </span>
<span class="co">#&gt; [15] S4Vectors_0.33.2                   BiocGenerics_0.41.1               </span>
<span class="co">#&gt; [17] TFBSTools_1.33.0                   JASPAR2020_0.99.10                </span>
<span class="co">#&gt; [19] monaLisa_1.1.0                     BiocStyle_2.23.0                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] vioplot_0.3.7               colorspace_2.0-2           </span>
<span class="co">#&gt;   [3] rjson_0.2.20                ellipsis_0.3.2             </span>
<span class="co">#&gt;   [5] rprojroot_2.0.2             GlobalOptions_0.1.2        </span>
<span class="co">#&gt;   [7] fs_1.5.0                    clue_0.3-60                </span>
<span class="co">#&gt;   [9] bit64_4.0.5                 AnnotationDbi_1.57.1       </span>
<span class="co">#&gt;  [11] fansi_0.5.0                 splines_4.2.0              </span>
<span class="co">#&gt;  [13] R.methodsS3_1.8.1           codetools_0.2-18           </span>
<span class="co">#&gt;  [15] doParallel_1.0.16           cachem_1.0.6               </span>
<span class="co">#&gt;  [17] knitr_1.36                  jsonlite_1.7.2             </span>
<span class="co">#&gt;  [19] Rsamtools_2.11.0            seqLogo_1.61.0             </span>
<span class="co">#&gt;  [21] annotate_1.73.0             GO.db_3.14.0               </span>
<span class="co">#&gt;  [23] cluster_2.1.2               R.oo_1.24.0                </span>
<span class="co">#&gt;  [25] png_0.1-7                   httr_1.4.2                 </span>
<span class="co">#&gt;  [27] BiocManager_1.30.16         readr_2.1.0                </span>
<span class="co">#&gt;  [29] compiler_4.2.0              assertthat_0.2.1           </span>
<span class="co">#&gt;  [31] Matrix_1.3-4                fastmap_1.1.0              </span>
<span class="co">#&gt;  [33] htmltools_0.5.2             tools_4.2.0                </span>
<span class="co">#&gt;  [35] gtable_0.3.0                glue_1.5.0                 </span>
<span class="co">#&gt;  [37] TFMPvalue_0.0.8             GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt;  [39] reshape2_1.4.4              dplyr_1.0.7                </span>
<span class="co">#&gt;  [41] Rcpp_1.0.7                  jquerylib_0.1.4            </span>
<span class="co">#&gt;  [43] pkgdown_1.9000.9000.9000    vctrs_0.3.8                </span>
<span class="co">#&gt;  [45] iterators_1.0.13            xfun_0.28                  </span>
<span class="co">#&gt;  [47] CNEr_1.31.0                 stringr_1.4.0              </span>
<span class="co">#&gt;  [49] lifecycle_1.0.1             restfulr_0.0.13            </span>
<span class="co">#&gt;  [51] poweRlaw_0.70.6             gtools_3.9.2               </span>
<span class="co">#&gt;  [53] XML_3.99-0.8                zoo_1.8-9                  </span>
<span class="co">#&gt;  [55] zlibbioc_1.41.0             scales_1.1.1               </span>
<span class="co">#&gt;  [57] ragg_1.2.0                  hms_1.1.1                  </span>
<span class="co">#&gt;  [59] parallel_4.2.0              RColorBrewer_1.1-2         </span>
<span class="co">#&gt;  [61] yaml_2.2.1                  memoise_2.0.0              </span>
<span class="co">#&gt;  [63] ggplot2_3.3.5               stabs_0.6-4                </span>
<span class="co">#&gt;  [65] sass_0.4.0                  stringi_1.7.5              </span>
<span class="co">#&gt;  [67] RSQLite_2.2.8               highr_0.9                  </span>
<span class="co">#&gt;  [69] BiocIO_1.5.0                desc_1.4.0                 </span>
<span class="co">#&gt;  [71] foreach_1.5.1               caTools_1.18.2             </span>
<span class="co">#&gt;  [73] BiocParallel_1.29.0         shape_1.4.6                </span>
<span class="co">#&gt;  [75] rlang_0.4.12                pkgconfig_2.0.3            </span>
<span class="co">#&gt;  [77] systemfonts_1.0.3           bitops_1.0-7               </span>
<span class="co">#&gt;  [79] pracma_2.3.3                evaluate_0.14              </span>
<span class="co">#&gt;  [81] lattice_0.20-45             purrr_0.3.4                </span>
<span class="co">#&gt;  [83] GenomicAlignments_1.31.2    bit_4.0.4                  </span>
<span class="co">#&gt;  [85] tidyselect_1.1.1            plyr_1.8.6                 </span>
<span class="co">#&gt;  [87] magrittr_2.0.1              bookdown_0.24              </span>
<span class="co">#&gt;  [89] R6_2.5.1                    generics_0.1.1             </span>
<span class="co">#&gt;  [91] sm_2.2-5.7                  DelayedArray_0.21.1        </span>
<span class="co">#&gt;  [93] DBI_1.1.1                   pillar_1.6.4               </span>
<span class="co">#&gt;  [95] survival_3.2-13             KEGGREST_1.35.0            </span>
<span class="co">#&gt;  [97] RCurl_1.98-1.5              tibble_3.1.6               </span>
<span class="co">#&gt;  [99] crayon_1.4.2                utf8_1.2.2                 </span>
<span class="co">#&gt; [101] tzdb_0.2.0                  rmarkdown_2.11             </span>
<span class="co">#&gt; [103] GetoptLong_1.0.5            blob_1.2.2                 </span>
<span class="co">#&gt; [105] digest_0.6.28               xtable_1.8-4               </span>
<span class="co">#&gt; [107] R.utils_2.11.0              textshaping_0.3.6          </span>
<span class="co">#&gt; [109] glmnet_4.1-3                munsell_0.5.0              </span>
<span class="co">#&gt; [111] DirichletMultinomial_1.37.0 bslib_0.3.1</span></code></pre></div>
</div>
<div class="section level1 unnumbered">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references">
<div id="ref-StabSel">
<p>Meinshausen, Nicolai, and Peter Bühlmann. 2010. “Stability Selection.” <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 72 (4): 417–73. <a href="https://doi.org/doi:10.1111/j.1467-9868.2010.00740.x" class="external-link">https://doi.org/doi:10.1111/j.1467-9868.2010.00740.x</a>.</p>
</div>
<div id="ref-compStabs">
<p>Shah, Rajen D., and Richard J. Samworth. 2013. “Variable Selection with Error Control: Another Look at Stability Selection.” <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 75 (1): 55–80. <a href="https://doi.org/doi:10.1111/j.1467-9868.2011.01034.x" class="external-link">https://doi.org/doi:10.1111/j.1467-9868.2011.01034.x</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
