<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Regression Based Approach for Motif Selection • monaLisa</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Regression Based Approach for Motif Selection">
<meta property="og:description" content="monaLisa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">monaLisa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/monaLisa.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/selecting_motifs_with_randLassoStabSel.html">Regression Based Approach for Motif Selection</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fmicompbio/monaLisa" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Regression Based Approach for Motif
Selection</h1>
                        <h4 data-toc-skip class="author">Dania Machlab,
Lukas Burger, Charlotte Soneson, Michael Stadler</h4>
            
            <h4 data-toc-skip class="date">2024-05-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/monaLisa/blob/HEAD/vignettes/selecting_motifs_with_randLassoStabSel.Rmd" class="external-link"><code>vignettes/selecting_motifs_with_randLassoStabSel.Rmd</code></a></small>
      <div class="hidden name"><code>selecting_motifs_with_randLassoStabSel.Rmd</code></div>

    </div>

    
    
<p><img src="monaLisa_logo_v1.png" style="display: block; margin-left: auto;
margin-right: auto; width: 50%; border: 0"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Identifying important transcription factor (TF) motifs, as shown in
<a href="https://bioconductor.org/packages/3.19/monaLisa/vignettes/monaLisa.html" class="external-link">the
main vignette</a>, could also be done using a regression-based approach,
where motifs have to compete against each other for selection. In this
framework, the response vector can be the observed experimental measure
of interest, e.g. log-fold changes of accessibility for a set of
regions, and the predictors consist of the TF motif hits across those
regions. In <em><a href="https://bioconductor.org/packages/3.19/monaLisa" class="external-link">monaLisa</a></em>,
we implement the randomized lasso stability selection proposed by <span class="citation">Meinshausen and Bühlmann (2010)</span> with the
improved error bounds introduced by <span class="citation">Shah and
Samworth (2013)</span>. We have modified the
<code><a href="https://rdrr.io/pkg/stabs/man/fitfuns.html" class="external-link">stabs::glmnet.lasso</a></code> function used by
<code><a href="https://rdrr.io/pkg/stabs/man/stabsel.html" class="external-link">stabs::stabsel</a></code> from the <em><a href="https://CRAN.R-project.org/package=stabs" class="external-link">stabs</a></em> package
to implement the randomized lasso.</p>
<p>Lasso stability selection performs the lasso regression multiple
times on subsamples of the data, and returns a selection probability for
each predictor (number of times selected divided by number of
regressions done). With the randomized lasso, a weakness parameter is
additionally used to vary the lasso penalty term <span class="math inline">\(\lambda\)</span> to a randomly chosen value
between [<span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\lambda\)</span>/weakness] for each predictor.
This type of regularization has advantages in cases where the number of
predictors exceeds the number of observations, in selecting variables
consistently, demonstrating better error control and not depending
strongly on the penalization parameter <span class="citation">(Meinshausen and Bühlmann 2010)</span>.</p>
<p>With this approach, TF motifs compete against each other to explain
the response vector, and we can also include additional predictors like
GC content to compete against the TF motifs for selection. This is
especially useful if the response is biased by sequence composition, for
example if regions with higher GC content tend to have higher response
values.</p>
<p>It is worth noting that, as with any regression analysis, the
interpretability of the results depends strongly on the quality of the
predictors. Hence, increasing the size of the motif database is not, in
itself, a guarantee for more interpretable results, since the added
motifs may be unrelated to the signal of interest. In addition, as
discussed in section @ref(collinearity), a high level of redundancy,
resulting in strong correlations among the motifs, may result in more
ambiguous selection probabilities in the regression approach. In fact,
also for the binned approach, although the motifs are evaluated
independently for association with the outcome, a high degree of
redundancy can lead to large collections of very similar motifs showing
significant enrichments, complicating interpretability of the
results.</p>
</div>
<div class="section level2">
<h2 id="stabsel">Motif selection with Randomized Lasso Stability Selection<a class="anchor" aria-label="anchor" href="#stabsel"></a>
</h2>
<p>In the example below, we select for TF motifs explaining log-fold
changes in chromatin accessibility (ATAC-seq) across the enhancers
between mouse liver and lung tissue at P0, but this can be applied to
other data types as well (ChIP-seq, RNA-seq, methylation etc.). Positive
log2-fold changes indicate more accessibility in the liver tissue,
whereas negative values indicate more accessibility in the lung
tissue.</p>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>We start by loading the needed packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/monaLisa" class="external-link">monaLisa</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-dataset">Load dataset<a class="anchor" aria-label="anchor" href="#load-dataset"></a>
</h3>
<p>In this example dataset from ENCODE <span class="citation">(The
ENCODE Project Consortium 2012)</span>, and available in <em><a href="https://bioconductor.org/packages/3.19/monaLisa" class="external-link">monaLisa</a></em>,
we have quantified ATAC-seq reads on enhancers in mouse P0 lung and
liver tissues. The log2-fold change (our response vector in this
example) is for liver vs lung chromatin accessibility. We are using a
set of 10,000 randomly sampled enhancers to illustrate how randomized
lasso stability selection can be used to select TF motifs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load GRanges object with logFC and peaks</span></span>
<span><span class="va">gr_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"atac_liver_vs_lung.rds"</span>, </span>
<span>                       package <span class="op">=</span> <span class="st">"monaLisa"</span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">gr_path</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-tfbs-per-motif-and-peak">Get TFBS per motif and peak<a class="anchor" aria-label="anchor" href="#get-tfbs-per-motif-and-peak"></a>
</h3>
<p>We will now construct the transcription factor binding site (TFBS)
matrix for known motifs (from a database like <em><a href="https://bioconductor.org/packages/3.19/JASPAR2020" class="external-link">JASPAR2020</a></em>)
in the given peak regions. We use the <code>findMotifHits</code>
function to scan for TF motif hits. This matrix will be the predictor
matrix in our regression. This step may take a while, and it may be
useful to parallelize it using the <code>BPPARAM</code> argument (e.g.
to run on <code>n</code> parallel threads using the multi-core backend,
you can use:
<code>findMotifHits(..., BPPARAM = BiocParallel::MulticoreParam(n))</code>).</p>
<p>As mentioned, this framework offers the flexibility to add additional
predictors to compete against the TF motifs for selection. Here, we add
the fraction of G+C and CpG observed/expected ratio as predictors, to
ensure that selected TF motifs are not just detecting a simple trend in
GC or CpG composition.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get PFMs (vertebrate TFs from Jaspar)</span></span>
<span><span class="va">pfms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span><span class="va">JASPAR2020</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>matrixtype <span class="op">=</span> <span class="st">"PFM"</span>, </span>
<span>                                      tax_group <span class="op">=</span> <span class="st">"vertebrates"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># randomly sample 300 PFMs for illustration purposes (for quick runtime)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">4563</span><span class="op">)</span></span>
<span><span class="va">pfms</span> <span class="op">&lt;-</span> <span class="va">pfms</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">300</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># convert PFMs to PWMs</span></span>
<span><span class="va">pwms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/toPWM-methods.html" class="external-link">toPWM</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get TFBS on given GRanges (peaks)</span></span>
<span><span class="co"># suppress warnings generated by matchPWM due to the presence of Ns </span></span>
<span><span class="co"># in the sequences</span></span>
<span><span class="va">peakSeq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Mmusculus.UCSC.mm10</span>, <span class="va">gr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">hits</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findMotifHits-methods.html">findMotifHits</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">pwms</span>, subject <span class="op">=</span> <span class="va">peakSeq</span>, min.score <span class="op">=</span> <span class="fl">10.0</span>,</span>
<span>                        BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get TFBS matrix</span></span>
<span><span class="va">TFBSmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">unclass</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">hits</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">hits</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                            <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">hits</span><span class="op">$</span><span class="va">pwmname</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">name</a></span><span class="op">(</span><span class="va">pwms</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">TFBSmatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt;               NR3C2 Arnt LHX1 SNAI1 MAFG ZSCAN4</span></span>
<span><span class="co">#&gt;   peak_51663      0    0    0     0    0      0</span></span>
<span><span class="co">#&gt;   peak_57870      0    0    0     0    0      0</span></span>
<span><span class="co">#&gt;   peak_2986       2    0    0     0    0      0</span></span>
<span><span class="co">#&gt;   peak_124022     0    0    0     0    0      0</span></span>
<span><span class="co">#&gt;   peak_29925      0    2    0     1    0      0</span></span>
<span><span class="co">#&gt;   peak_95246      0    0    0     0    0      0</span></span>
<span></span>
<span><span class="co"># remove TF motifs with 0 binding sites in all regions</span></span>
<span><span class="va">zero_TF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">TFBSmatrix</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">zero_TF</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span><span class="va">TFBSmatrix</span> <span class="op">&lt;-</span> <span class="va">TFBSmatrix</span><span class="op">[</span>, <span class="op">!</span><span class="va">zero_TF</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># calculate G+C and CpG obs/expected</span></span>
<span><span class="va">fMono</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/nucleotideFrequency.html" class="external-link">oligonucleotideFrequency</a></span><span class="op">(</span><span class="va">peakSeq</span>, width <span class="op">=</span> <span class="fl">1L</span>, as.prob <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">fDi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/nucleotideFrequency.html" class="external-link">oligonucleotideFrequency</a></span><span class="op">(</span><span class="va">peakSeq</span>, width <span class="op">=</span> <span class="fl">2L</span>, as.prob <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">fracGC</span> <span class="op">&lt;-</span> <span class="va">fMono</span><span class="op">[</span>, <span class="st">"G"</span><span class="op">]</span> <span class="op">+</span> <span class="va">fMono</span><span class="op">[</span>, <span class="st">"C"</span><span class="op">]</span></span>
<span><span class="va">oeCpG</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">fDi</span><span class="op">[</span>, <span class="st">"CG"</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">fMono</span><span class="op">[</span>, <span class="st">"G"</span><span class="op">]</span> <span class="op">*</span> <span class="va">fMono</span><span class="op">[</span>, <span class="st">"C"</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add GC and oeCpG to predictor matrix</span></span>
<span><span class="va">TFBSmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">fracGC</span>, <span class="va">oeCpG</span>, <span class="va">TFBSmatrix</span><span class="op">)</span></span>
<span><span class="va">TFBSmatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;                fracGC     oeCpG NR3C2 Arnt LHX1 SNAI1</span></span>
<span><span class="co">#&gt; peak_51663  0.5155709 0.4079115     0    0    0     0</span></span>
<span><span class="co">#&gt; peak_57870  0.4963235 0.3048298     0    0    0     0</span></span>
<span><span class="co">#&gt; peak_2986   0.4008264 0.3103806     2    0    0     0</span></span>
<span><span class="co">#&gt; peak_124022 0.4572650 0.4429813     0    0    0     0</span></span>
<span><span class="co">#&gt; peak_29925  0.4675000 0.3495939     0    2    0     1</span></span>
<span><span class="co">#&gt; peak_95246  0.5144509 0.4020976     0    0    0     0</span></span></code></pre></div>
<div class="section level4">
<h4 id="collinearity">A note on collinearity<a class="anchor" aria-label="anchor" href="#collinearity"></a>
</h4>
<p>At this point it is useful for the user to get an overall feeling of
the collinearity structure in the TFBS matrix. Motifs that share a lot
of similar binding sites across the peaks will be highly correlated.
High collinearity between predictors is a well known problem in linear
regression. It particularly manifests itself in the lasso regression for
example, where if variables are equally highly correlated with the
response, not all are co-selected as predictors (if they are signal
variables). Instead, one is arbitrarily chosen while the others’
coefficients are set to zero. The rationale is that the non-selected
correlated predictors do not provide much additional information to
explain the response. It is good to be aware of these properties of
regression, and to place more weight on the meaning of the selected
motif itself, rather than the specific TF name when interpreting the
results.</p>
<p>If many cases of high correlations exist and this is a concern, one
may consider selecting a representative set of predictors to use. This
may for example be achieved by clustering the weight matrices beforehand
and using only one representative motif per cluster for running the
regression, using tools such as for example RSAT <span class="citation">(Castro-Mondragon et al. 2017)</span>. RSAT-derived
clusters of Jaspar weight matrices can be found at <a href="https://jaspar.genereg.net/matrix-clusters/vertebrates/" class="external-link uri">https://jaspar.genereg.net/matrix-clusters/vertebrates/</a>.</p>
<p>If the user is interested in working with all correlated motifs, the
binned approach is preferable as the motifs are independently tested for
significance (see the <a href="https://bioconductor.org/packages/3.19/monaLisa/vignettes/monaLisa.html" class="external-link">binned
enrichment vignette</a>). In the regression-based approach on the other
hand, we can more clearly understand the relative contributions of TF
motifs to the response in the context of each other.</p>
</div>
</div>
<div class="section level3">
<h3 id="identify-important-tfs">Identify important TFs<a class="anchor" aria-label="anchor" href="#identify-important-tfs"></a>
</h3>
<p>We can now run randomized lasso stability selection to identify TFs
that are likely to explain the log-fold changes in accessibility. The
exact choice of parameter values for this approach will depend largely
on how stringent the user wishes to be and how much signal there is in
the data. For example, for more stringent selections, one may decrease
the value of the <code>weakness</code> parameter which will make it
harder for a variable to get selected. The user is in control of false
discoveries with the <code>PFER</code> parameter, which indicates the
number of falsely selected variables. As for the selection probability
cutoff, <span class="citation">Meinshausen and Bühlmann (2010)</span>
argue that values in the range of [0.6, 0.9] should give similar
results. See the <code>randLassoStabSel</code> function for more details
and default parameter values as well as the <code><a href="https://rdrr.io/pkg/stabs/man/stabsel.html" class="external-link">stabs::stabsel</a></code>
function for the default assumptions on the stability selection
implementation by <span class="citation">Shah and Samworth
(2013)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## randLassoStabSel() is stochastic, so if we run with parallelization</span></span>
<span><span class="co">##   (`mc.cores` argument), we must select a random number generator that can</span></span>
<span><span class="co">##   provide multiple streams of random numbers used in the `parallel` package</span></span>
<span><span class="co">##   and set its seed for reproducible results</span></span>
<span><span class="co"># RNGkind("L'Ecuyer-CMRG")</span></span>
<span><span class="co"># set.seed(123)</span></span>
<span><span class="co"># se &lt;- randLassoStabSel(x = TFBSmatrix, y = gr$logFC_liver_vs_lung, </span></span>
<span><span class="co">#                        cutoff = 0.8, mc.preschedule = TRUE, </span></span>
<span><span class="co">#                        mc.set.seed = TRUE, mc.cores = 2L)</span></span>
<span></span>
<span><span class="co"># if not running in parallel mode, it is enough to use set.seed() before </span></span>
<span><span class="co">#   the call to ensure reproducibility (`mc.sores = 1`)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/randLassoStabSel.html">randLassoStabSel</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TFBSmatrix</span>, y <span class="op">=</span> <span class="va">gr</span><span class="op">$</span><span class="va">logFC_liver_vs_lung</span>, </span>
<span>                       cutoff <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="va">se</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 10000 300 </span></span>
<span><span class="co">#&gt; metadata(12): stabsel.params.cutoff stabsel.params.selected ...</span></span>
<span><span class="co">#&gt;   stabsel.params.call randStabsel.params.weakness</span></span>
<span><span class="co">#&gt; assays(1): x</span></span>
<span><span class="co">#&gt; rownames(10000): peak_51663 peak_57870 ... peak_98880 peak_67984</span></span>
<span><span class="co">#&gt; rowData names(1): y</span></span>
<span><span class="co">#&gt; colnames(300): fracGC oeCpG ... CLOCK OLIG2</span></span>
<span><span class="co">#&gt; colData names(20): selProb selected ... regStep16 regStep17</span></span>
<span></span>
<span><span class="co"># selected TFs</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="va">se</span><span class="op">$</span><span class="va">selected</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] "NKX2-5"       "GATA1::TAL1"  "HNF1B"        "HNF4A(var.2)" "Nr2f6"       </span></span>
<span><span class="co">#&gt;  [6] "ONECUT3"      "MYF5"         "THRB"         "ISL2"         "NR2C2"       </span></span>
<span><span class="co">#&gt; [11] "TEAD3"        "TEAD4"        "GATA3"        "RORA"         "NFIC"        </span></span>
<span><span class="co">#&gt; [16] "ZEB1"</span></span></code></pre></div>
<p>The stability paths visualize how predictors get selected over
decreasing regularization stringency (from left to right):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotStabilityPaths.html">plotStabilityPaths</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/plotStabilityPaths-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>Each line corresponds to a predictor, and we can see the selection
probabilities as a function of the regularization steps, corresponding
to decreasing values for the lambda regularization parameter in lasso.
The predictor (TF motif) selection happens at the last step, given the
specified minimum probability.</p>
<p>We can also visualize the selection probabilities of the selected TF
motifs, optionally multiplied by the sign of the correlation to the
response vector, to know how the TF relates to the change of
accessibility (<code>directional</code> parameter). In our example,
positive correlations to the response vector indicate a correlation with
enhancers more accessible in the liver, whereas negative ones indicate a
correlation with enhancers more accessible in the lung. Note that
although one can vary the <code>selProbMinPlot</code> argument which
sets the selection probability cutoff, it is best to re-run randomized
lasso stability selection with the new cutoff, as this influences other
parameter choices the model uses internally. See <span class="citation">Meinshausen and Bühlmann (2010)</span> for more
details.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSelectionProb.html">plotSelectionProb</a></span><span class="op">(</span><span class="va">se</span>, directional <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/plotSelProbs-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>Next, we visualize the correlation structure of the selected TF
motifs within the TFBS matrix. While the collinearity of predictors is a
general issue in regression-based approaches, randomized lasso stability
selection normally does better at co-selecting intermediately correlated
predictors. In practice, we see it select predictors with correlations
as high as 0.9. However, it is good to keep in mind that this can be an
issue, and that predictors that are highly correlated with each other
might not end up being co-selected (see section @ref(collinearity)).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset the selected TFs</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="va">se</span><span class="op">$</span><span class="va">selected</span><span class="op">]</span></span>
<span><span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="va">sel</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># exclude oeCpG and fracGC</span></span>
<span><span class="va">excl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"oeCpG"</span>, <span class="st">"fracGC"</span><span class="op">)</span></span>
<span><span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se_sub</span><span class="op">[</span>, <span class="op">!</span><span class="va">excl</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># correlation matrix </span></span>
<span><span class="va">TFBSmatrixCorSel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">TFBSmatrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># heatmap</span></span>
<span><span class="va">pfmsSel</span> <span class="op">&lt;-</span> <span class="va">pfms</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">TFBSmatrixCorSel</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">name</a></span><span class="op">(</span><span class="va">pfms</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">maxwidth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="va">pfmsSel</span><span class="op">)</span>, <span class="va">ncol</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seqlogoGrobs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pfmsSel</span>, <span class="va">seqLogoGrob</span>, xmax <span class="op">=</span> <span class="va">maxwidth</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hmSeqlogo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html" class="external-link">rowAnnotation</a></span><span class="op">(</span>logo <span class="op">=</span> <span class="fu"><a href="../reference/annoSeqlogo.html">annoSeqlogo</a></span><span class="op">(</span><span class="va">seqlogoGrobs</span>, which <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                           annotation_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"inch"</span><span class="op">)</span>, </span>
<span>                           show_annotation_name <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">colAnn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>AUC <span class="op">=</span> <span class="va">se_sub</span><span class="op">$</span><span class="va">selAUC</span>, selProb <span class="op">=</span> <span class="va">se_sub</span><span class="op">$</span><span class="va">selProb</span>,</span>
<span>                            show_legend <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                            show_annotation_name <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                              AUC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                                               <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"brown"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                              selProb <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                                                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"steelblue"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">TFBSmatrixCorSel</span>, </span>
<span>        show_row_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        name <span class="op">=</span> <span class="st">"Pear. Cor."</span>, column_title <span class="op">=</span> <span class="st">"Selected TFs"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>        right_annotation <span class="op">=</span> <span class="va">hmSeqlogo</span>,</span>
<span>        top_annotation <span class="op">=</span> <span class="va">colAnn</span><span class="op">)</span></span></code></pre></div>
<p><img src="selecting_motifs_with_randLassoStabSel_files/figure-html/TFBScor_selected-1.png" width="80%" style="display: block; margin: auto;"></p>
<p> </p>
<p>We can examine the peaks that have hits for a selected TF motif of
interest, ordered by absolute accessibility changes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TF</span> <span class="op">&lt;-</span> <span class="va">sel</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">TF</span></span>
<span><span class="co">#&gt; [1] "GATA1::TAL1"</span></span>
<span></span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"x"</span><span class="op">)</span><span class="op">[</span>, <span class="va">TF</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># peaks that contain TF hits...</span></span>
<span><span class="va">nm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">gr</span><span class="op">$</span><span class="va">logFC_liver_vs_lung</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="co"># ... order by |logFC|</span></span>
<span></span>
<span><span class="va">gr</span><span class="op">[</span><span class="va">nm</span><span class="op">]</span></span>
<span><span class="co">#&gt; GRanges object with 1102 ranges and 1 metadata column:</span></span>
<span><span class="co">#&gt;               seqnames              ranges strand | logFC_liver_vs_lung</span></span>
<span><span class="co">#&gt;                  &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;    peak_59642     chr8   80499941-80501128      * |             5.31505</span></span>
<span><span class="co">#&gt;    peak_58095     chr8   40658907-40659590      * |             5.13536</span></span>
<span><span class="co">#&gt;   peak_100144    chr14   69318236-69318909      * |             5.09810</span></span>
<span><span class="co">#&gt;    peak_98635    chr14   41124351-41126121      * |             5.07972</span></span>
<span><span class="co">#&gt;    peak_42826     chr6   29700952-29701570      * |             5.02324</span></span>
<span><span class="co">#&gt;           ...      ...                 ...    ... .                 ...</span></span>
<span><span class="co">#&gt;    peak_45104     chr6   84095148-84095495      * |        -0.020828315</span></span>
<span><span class="co">#&gt;    peak_16029     chr2 132269993-132270884      * |         0.007621169</span></span>
<span><span class="co">#&gt;    peak_55106     chr7 129526059-129526291      * |         0.004120784</span></span>
<span><span class="co">#&gt;    peak_97436    chr14   20873345-20873654      * |         0.000765027</span></span>
<span><span class="co">#&gt;    peak_67068     chr9   75050089-75050395      * |        -0.000217812</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 19 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-info-and-logo">Session info and logo<a class="anchor" aria-label="anchor" href="#session-info-and-logo"></a>
</h2>
<p>The monaLisa logo uses a drawing that was obtained from <a href="http://vectorish.com/lisa-simpson.html" class="external-link uri">http://vectorish.com/lisa-simpson.html</a> under the
Creative Commons attribution - non-commercial 3.0 license: <a href="https://creativecommons.org/licenses/by-nc/3.0/" class="external-link uri">https://creativecommons.org/licenses/by-nc/3.0/</a>.</p>
<p>This vignette was built using:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.4.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] circlize_0.4.16                    ComplexHeatmap_2.20.0             </span></span>
<span><span class="co">#&gt;  [3] SummarizedExperiment_1.34.0        Biobase_2.64.0                    </span></span>
<span><span class="co">#&gt;  [5] MatrixGenerics_1.16.0              matrixStats_1.3.0                 </span></span>
<span><span class="co">#&gt;  [7] BSgenome.Mmusculus.UCSC.mm10_1.4.3 BSgenome_1.72.0                   </span></span>
<span><span class="co">#&gt;  [9] rtracklayer_1.64.0                 BiocIO_1.14.0                     </span></span>
<span><span class="co">#&gt; [11] Biostrings_2.72.0                  XVector_0.44.0                    </span></span>
<span><span class="co">#&gt; [13] GenomicRanges_1.56.0               GenomeInfoDb_1.40.0               </span></span>
<span><span class="co">#&gt; [15] IRanges_2.38.0                     S4Vectors_0.42.0                  </span></span>
<span><span class="co">#&gt; [17] BiocGenerics_0.50.0                TFBSTools_1.42.0                  </span></span>
<span><span class="co">#&gt; [19] JASPAR2020_0.99.10                 monaLisa_1.11.1                   </span></span>
<span><span class="co">#&gt; [21] BiocStyle_2.32.0                  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3          jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;   [3] shape_1.4.6.1               magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [5] rmarkdown_2.26.2            GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.4                    zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt;   [9] ragg_1.3.1                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [11] memoise_2.0.1               Rsamtools_2.20.0           </span></span>
<span><span class="co">#&gt;  [13] RCurl_1.98-1.14             htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt;  [15] S4Arrays_1.4.0              curl_5.2.1                 </span></span>
<span><span class="co">#&gt;  [17] CNEr_1.40.0                 SparseArray_1.4.3          </span></span>
<span><span class="co">#&gt;  [19] sass_0.4.9                  pracma_2.4.4               </span></span>
<span><span class="co">#&gt;  [21] bslib_0.7.0                 desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [23] plyr_1.8.9                  zoo_1.8-12                 </span></span>
<span><span class="co">#&gt;  [25] cachem_1.0.8                GenomicAlignments_1.40.0   </span></span>
<span><span class="co">#&gt;  [27] lifecycle_1.0.4             iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [29] pkgconfig_2.0.3             Matrix_1.7-0               </span></span>
<span><span class="co">#&gt;  [31] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [33] GenomeInfoDbData_1.2.12     clue_0.3-65                </span></span>
<span><span class="co">#&gt;  [35] digest_0.6.35               colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [37] TFMPvalue_0.0.9             AnnotationDbi_1.66.0       </span></span>
<span><span class="co">#&gt;  [39] stabs_0.6-4                 textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [41] RSQLite_2.3.6               seqLogo_1.70.0             </span></span>
<span><span class="co">#&gt;  [43] fansi_1.0.6                 httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [45] abind_1.4-5                 compiler_4.4.0             </span></span>
<span><span class="co">#&gt;  [47] bit64_4.0.5                 doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [49] BiocParallel_1.38.0         DBI_1.2.2                  </span></span>
<span><span class="co">#&gt;  [51] highr_0.10                  R.utils_2.12.3             </span></span>
<span><span class="co">#&gt;  [53] poweRlaw_0.80.0             DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [55] rjson_0.2.21                gtools_3.9.5               </span></span>
<span><span class="co">#&gt;  [57] caTools_1.18.2              tools_4.4.0                </span></span>
<span><span class="co">#&gt;  [59] R.oo_1.26.0                 glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [61] restfulr_0.0.15             cluster_2.1.6              </span></span>
<span><span class="co">#&gt;  [63] reshape2_1.4.4              generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [65] gtable_0.3.5                tzdb_0.4.0                 </span></span>
<span><span class="co">#&gt;  [67] R.methodsS3_1.8.2           sm_2.2-6.0                 </span></span>
<span><span class="co">#&gt;  [69] hms_1.1.3                   utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [71] foreach_1.5.2               pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [73] stringr_1.5.1               splines_4.4.0              </span></span>
<span><span class="co">#&gt;  [75] dplyr_1.1.4                 lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [77] survival_3.6-4              bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [79] annotate_1.82.0             tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [81] DirichletMultinomial_1.46.0 GO.db_3.19.1               </span></span>
<span><span class="co">#&gt;  [83] vioplot_0.4.0               knitr_1.46                 </span></span>
<span><span class="co">#&gt;  [85] bookdown_0.39               xfun_0.43                  </span></span>
<span><span class="co">#&gt;  [87] stringi_1.8.4               UCSC.utils_1.0.0           </span></span>
<span><span class="co">#&gt;  [89] yaml_2.3.8                  evaluate_0.23              </span></span>
<span><span class="co">#&gt;  [91] codetools_0.2-20            tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [93] BiocManager_1.30.23         cli_3.6.2                  </span></span>
<span><span class="co">#&gt;  [95] xtable_1.8-4                systemfonts_1.0.6          </span></span>
<span><span class="co">#&gt;  [97] munsell_0.5.1               jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [99] Rcpp_1.0.12                 png_0.1-8                  </span></span>
<span><span class="co">#&gt; [101] XML_3.99-0.16.1             parallel_4.4.0             </span></span>
<span><span class="co">#&gt; [103] pkgdown_2.0.9.9000          ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt; [105] readr_2.1.5                 blob_1.2.4                 </span></span>
<span><span class="co">#&gt; [107] bitops_1.0-7                glmnet_4.1-8               </span></span>
<span><span class="co">#&gt; [109] pwalign_1.0.0               scales_1.3.0               </span></span>
<span><span class="co">#&gt; [111] purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">#&gt; [113] GetoptLong_1.0.5            rlang_1.1.3                </span></span>
<span><span class="co">#&gt; [115] KEGGREST_1.44.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-RSAT" class="csl-entry">
Castro-Mondragon, J. A., S. Jaeger, D. Thieffry, M. Thomas-Chollier, and
J. van Helden. 2017. <span>“<span class="nocase"><span>R</span><span>S</span><span>A</span><span>T</span>
matrix-clustering: dynamic exploration and redundancy reduction of
transcription factor binding motif collections</span>.”</span>
<em>Nucleic Acids Res</em> 45 (13): e119. <a href="https://doi.org/10.1093/nar/gkx314" class="external-link">https://doi.org/10.1093/nar/gkx314</a>.
</div>
<div id="ref-StabSel" class="csl-entry">
Meinshausen, Nicolai, and Peter Bühlmann. 2010. <span>“Stability
Selection.”</span> <em>Journal of the Royal Statistical Society: Series
B (Statistical Methodology)</em> 72 (4): 417–73. <a href="https://doi.org/doi:10.1111/j.1467-9868.2010.00740.x" class="external-link">https://doi.org/doi:10.1111/j.1467-9868.2010.00740.x</a>.
</div>
<div id="ref-compStabs" class="csl-entry">
Shah, Rajen D., and Richard J. Samworth. 2013. <span>“Variable Selection
with Error Control: Another Look at Stability Selection.”</span>
<em>Journal of the Royal Statistical Society: Series B (Statistical
Methodology)</em> 75 (1): 55–80. <a href="https://doi.org/doi:10.1111/j.1467-9868.2011.01034.x" class="external-link">https://doi.org/doi:10.1111/j.1467-9868.2011.01034.x</a>.
</div>
<div id="ref-encode2012" class="csl-entry">
The ENCODE Project Consortium. 2012. <span>“An Integrated Encyclopedia
of <span>DNA</span> Elements in the Human Genome.”</span>
<em>Nature</em> 489 (7414): 57–74.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/machlabd" class="external-link">Dania Machlab</a>, <a href="https://github.com/LukasBurger" class="external-link">Lukas Burger</a>, <a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a>, <a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
